{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "I need ideally data from 5 to 17 for all samples and I need to be able to correct for FACS sorting when I do compositional analysis (need to look more into this) and make sure I have all the metadata info to be able to do this for all samples \n",
    "\n",
    "So I have data from panfetal paper, yolk sac paper, skin paper, liver paper, thymus paper which I want to include for the cross-organ comparison \n",
    "\n",
    "From the panfetal paper metadata, I will need to remove the bone marrow samples \n",
    "I need to first look at all the Source Name samples from these papers. Make a list of them.  Concatenate all the harmonised datadrames into one. Done (lots of repeated rows) \n",
    "\n",
    "Then I need to check how many of these are present in the panfetal metadata and then need to transfer that metadata. Then I need to make a list of samples \n",
    "1)\tShared with panfetal (167)\n",
    "2)\tUnique to panfetal (28) \n",
    "3)\tUnique to skin, yolksac, and thymus studies (91) \n",
    "\n",
    "Combine these samples. Then create an integrated metadata. Then check how many of these samples are present in the list alex gave, and give him the list of samples that need to be remapped\n",
    "\n",
    "So total samples 167 + 28 + 91 = 286 samples for cross-organ \n",
    "\n",
    "I combined the samples and except number 2 samples, I had 256 samples \n",
    "From Thymus paper, I removed 11 samples as they were VDJ libraries\n",
    "FCAImmP7607591\n",
    "FCAImmP7607599\n",
    "FCAImmP7607600\n",
    "FCAImmP7607608\n",
    "FCAImmP7607609\n",
    "FCAImmP7851882\n",
    "FCAImmP7851883\n",
    "FCAImmP7851884\n",
    "FCAImmP7851885\n",
    "FCAImmP7851886\n",
    "FCAImmP7851887\n",
    "\n",
    "Also removed four thymus samples from panfetal paper \n",
    "WSSS_F_Imm11192515\tCCR10high\n",
    "WSSS_F_Imm11192516\tCCR10lowCD27+CD43+\n",
    "WSSS_F_Imm11192517\tCCR10lowCD27-CD43+\n",
    "WSSS_F_Imm11192518\tCCR10lowCD27-CD43-\n",
    "\n",
    "So I was left with 256-15= 241 samples\n",
    "So I added 28 panfetal, so in total I had 269 samples for cross organ comparison. From this I removed 2 samples from MLN because the sample size is too small for meaninnful comparison. Then I added 12 meninges samples to this \n",
    "so the final cross organ comparison contains 279 samples in total from the liver, yolksac, meninges, kidney, skin, \n",
    "spleen and thymus. \n",
    "\n",
    "For the bone atlas: 221 samples (including meninges)l after removing the meninges samples, there are in total 209 samples from the skull, long bones, spine, hip, knee and shoulder. \n",
    "\n",
    "So I combined all the samples together . This is 488 samples in total. 10X scRNA-seq (cellular RNA 5'prime and 3'prime) and multiome (nuclear RNA) samples. \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "cannot assign to operator (1030814685.py, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[1], line 3\u001b[0;36m\u001b[0m\n\u001b[0;31m    for GSM/SRR samples, here's the relationship between the sample and run:\u001b[0m\n\u001b[0m        ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m cannot assign to operator\n"
     ]
    }
   ],
   "source": [
    "# Finally I need to change the Sample_IDs for the following \n",
    "\n",
    "for GSM/SRR samples, here is  the relationship between the sample and run: so I need to replace the Sample_IDs in the second column \n",
    "with the ones in the first column. \n",
    "\n",
    "GSM4274188      SRR10901267\n",
    "GSM4274189      SRR10901268\n",
    "GSM4274190      SRR10901269\n",
    "GSM4274191      SRR10901270\n",
    "GSM4274192      SRR10901271,SRR10901272,SRR10901273,SRR10901274\n",
    "GSM4274193      SRR10901275\n",
    "GSM4274194      SRR10901276\n",
    "GSM4609136      SRR11977450,SRR11977451,SRR11977452,SRR11977453\n",
    "4:05\n",
    "What is done usually, is separate sequencing lanes are submitted as separate lanes.\n",
    "So processing them separately would not make sense,\n",
    "because they came from the same 10x run, and have same barcodes\n",
    "I can double-check if e.g. SRR10901271,SRR10901272,SRR10901273,SRR10901274 all have matching top barcodes, \n",
    "but I am 99% sure it is  how I describe it. We always reprocess public data based on the sample, \n",
    "and that is how vast majority of people submit them to GEO>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "379"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "214 + 165"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "488"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "279+ 209"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "bone_atlas_metadata= pd.read_csv('/nfs/team298/sm54/metadata/sample_metadata_combined.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "bone_atlas_metadata.to_csv('/nfs/team298/sm54/metadata/sample_metadata_combined.csv', index=False )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "bone_atlas_metadata['Multiplexed'] =bone_atlas_metadata['Multiplexed'].str.strip()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['No', 'Yes'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bone_atlas_metadata[\"Multiplexed\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiplexed_samples= bone_atlas_metadata[bone_atlas_metadata['Multiplexed']==\"Yes\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "110"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "22 * 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organ</th>\n",
       "      <th>Run_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ.1</th>\n",
       "      <th>Age</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Technology</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Multiplexed</th>\n",
       "      <th>Expected_Donors</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F13177434_and_HCA_BN_F13177207</td>\n",
       "      <td>HCA_BN_F13177434</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2356/2372</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F13177435_and_HCA_BN_F13177208</td>\n",
       "      <td>HCA_BN_F13177435</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2356/2372</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F13177436_and_HCA_BN_F13177209</td>\n",
       "      <td>HCA_BN_F13177436</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2356/2372</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12808058_and_HCA_BN_F12803234</td>\n",
       "      <td>HCA_BN_F12808058</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2370/2369</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12808059_and_HCA_BN_F12803235</td>\n",
       "      <td>HCA_BN_F12808059</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2370/2369</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12808060_and_HCA_BN_F12803236</td>\n",
       "      <td>HCA_BN_F12808060</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>skullbase</td>\n",
       "      <td>2370/2369</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Hip</td>\n",
       "      <td>HCA_BN_F12627471_and_HCA_BN_F12605340</td>\n",
       "      <td>HCA_BN_F12627471</td>\n",
       "      <td>Hip</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>hip</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Hip/Knee</td>\n",
       "      <td>HCA_BN_F12922485_and_HCA_BN_F12918723</td>\n",
       "      <td>HCA_BN_F12922485</td>\n",
       "      <td>Hip/Knee</td>\n",
       "      <td>7.3/8.6</td>\n",
       "      <td>hip/knee</td>\n",
       "      <td>2392/2394</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Hip/Shoulder</td>\n",
       "      <td>HCA_BN_F12627472_and_HCA_BN_F12605341</td>\n",
       "      <td>HCA_BN_F12627472</td>\n",
       "      <td>Hip/Shoulder</td>\n",
       "      <td>5.8, 8.3</td>\n",
       "      <td>hip/shoulder</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Knee</td>\n",
       "      <td>HCA_BN_F12627470_and_HCA_BN_F12605339</td>\n",
       "      <td>HCA_BN_F12627470</td>\n",
       "      <td>Knee</td>\n",
       "      <td>5.8, 7.6, 8.3</td>\n",
       "      <td>knee</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Shoulder</td>\n",
       "      <td>HCA_BN_F12922487_and_HCA_BN_F12918725</td>\n",
       "      <td>HCA_BN_F12922487</td>\n",
       "      <td>Shoulder</td>\n",
       "      <td>7.3/8.6</td>\n",
       "      <td>shoulder</td>\n",
       "      <td>2392/2394</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Shoulder</td>\n",
       "      <td>HCA_BN_F12627473_and_HCA_BN_F12605342</td>\n",
       "      <td>HCA_BN_F12627473</td>\n",
       "      <td>Shoulder</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>shoulder</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Skull base</td>\n",
       "      <td>HCA_BN_F13109714_and_HCA_BN_F13128942</td>\n",
       "      <td>HCA_BN_F13109714</td>\n",
       "      <td>Skull base</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>HCA_BN_F13109715_and_HCA_BN_F13128943</td>\n",
       "      <td>HCA_BN_F13109715</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>HCA_BN_F13109718_and_HCA_BN_F13128946</td>\n",
       "      <td>HCA_BN_F13109718</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>HCA_BN_F13109719_and_HCA_BN_F13128947</td>\n",
       "      <td>HCA_BN_F13109719</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13168990_and_HCA_BRA_F13168606</td>\n",
       "      <td>HCA_BRA_F13168990</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>cerebral</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13168991_and_HCA_BRA_F13168607</td>\n",
       "      <td>HCA_BRA_F13168991</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>falcine</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13168992_and_HCA_BRA_F13168608</td>\n",
       "      <td>HCA_BRA_F13168992</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>skullbase</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13168993_and_HCA_BRA_F13168609</td>\n",
       "      <td>HCA_BRA_F13168993</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>skullbase</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13774551_and_HCA_BRA_F13768421</td>\n",
       "      <td>HCA_BRA_F13774551</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>HCA_BRA_F13774552</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13774553_and_HCA_BRA_F13768423</td>\n",
       "      <td>HCA_BRA_F13774553</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13774554_and_HCA_BRA_F13852281</td>\n",
       "      <td>HCA_BRA_F13774554</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13774555_and_HCA_BRA_F13852282</td>\n",
       "      <td>HCA_BRA_F13774555</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Meninges</td>\n",
       "      <td>HCA_BRA_F13774556_and_HCA_BRA_F13852283</td>\n",
       "      <td>HCA_BRA_F13774556</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3</td>\n",
       "      <td>10x_multiome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Organ                                   Run_ID  \\\n",
       "11             Calvaria     HCA_BN_F13177434_and_HCA_BN_F13177207   \n",
       "12             Calvaria     HCA_BN_F13177435_and_HCA_BN_F13177208   \n",
       "13             Calvaria     HCA_BN_F13177436_and_HCA_BN_F13177209   \n",
       "14             Calvaria     HCA_BN_F12808058_and_HCA_BN_F12803234   \n",
       "15             Calvaria     HCA_BN_F12808059_and_HCA_BN_F12803235   \n",
       "16             Calvaria     HCA_BN_F12808060_and_HCA_BN_F12803236   \n",
       "22                  Hip     HCA_BN_F12627471_and_HCA_BN_F12605340   \n",
       "23              Hip/Knee    HCA_BN_F12922485_and_HCA_BN_F12918723   \n",
       "24          Hip/Shoulder    HCA_BN_F12627472_and_HCA_BN_F12605341   \n",
       "32                  Knee    HCA_BN_F12627470_and_HCA_BN_F12605339   \n",
       "37             Shoulder     HCA_BN_F12922487_and_HCA_BN_F12918725   \n",
       "38             Shoulder     HCA_BN_F12627473_and_HCA_BN_F12605342   \n",
       "50           Skull base     HCA_BN_F13109714_and_HCA_BN_F13128942   \n",
       "51   Skull base/Calvaria    HCA_BN_F13109715_and_HCA_BN_F13128943   \n",
       "52   Skull base/Calvaria    HCA_BN_F13109718_and_HCA_BN_F13128946   \n",
       "53   Skull base/Calvaria    HCA_BN_F13109719_and_HCA_BN_F13128947   \n",
       "118             Meninges  HCA_BRA_F13168990_and_HCA_BRA_F13168606   \n",
       "119             Meninges  HCA_BRA_F13168991_and_HCA_BRA_F13168607   \n",
       "120             Meninges  HCA_BRA_F13168992_and_HCA_BRA_F13168608   \n",
       "121             Meninges  HCA_BRA_F13168993_and_HCA_BRA_F13168609   \n",
       "122             Meninges  HCA_BRA_F13774551_and_HCA_BRA_F13768421   \n",
       "123             Meninges  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "124             Meninges  HCA_BRA_F13774553_and_HCA_BRA_F13768423   \n",
       "125             Meninges  HCA_BRA_F13774554_and_HCA_BRA_F13852281   \n",
       "126             Meninges  HCA_BRA_F13774555_and_HCA_BRA_F13852282   \n",
       "127             Meninges  HCA_BRA_F13774556_and_HCA_BRA_F13852283   \n",
       "\n",
       "             Sample_ID              Organ.1            Age  \\\n",
       "11    HCA_BN_F13177434            Calvaria       10.3,11.2   \n",
       "12    HCA_BN_F13177435            Calvaria       10.3,11.2   \n",
       "13    HCA_BN_F13177436            Calvaria       10.3,11.2   \n",
       "14    HCA_BN_F12808058            Calvaria         5.8/7.6   \n",
       "15    HCA_BN_F12808059            Calvaria         5.8/7.6   \n",
       "16    HCA_BN_F12808060            Calvaria         5.8/7.6   \n",
       "22    HCA_BN_F12627471                 Hip        7.6, 8.3   \n",
       "23    HCA_BN_F12922485             Hip/Knee        7.3/8.6   \n",
       "24    HCA_BN_F12627472         Hip/Shoulder       5.8, 8.3   \n",
       "32    HCA_BN_F12627470                 Knee  5.8, 7.6, 8.3   \n",
       "37    HCA_BN_F12922487            Shoulder         7.3/8.6   \n",
       "38    HCA_BN_F12627473            Shoulder        7.6, 8.3   \n",
       "50    HCA_BN_F13109714          Skull base         11.2,13   \n",
       "51    HCA_BN_F13109715  Skull base/Calvaria        11.2,13   \n",
       "52    HCA_BN_F13109718  Skull base/Calvaria        11.2,13   \n",
       "53    HCA_BN_F13109719  Skull base/Calvaria        11.2,13   \n",
       "118  HCA_BRA_F13168990             Meninges        6.7,8.4   \n",
       "119  HCA_BRA_F13168991             Meninges        6.7,8.4   \n",
       "120  HCA_BRA_F13168992             Meninges        6.7,8.4   \n",
       "121  HCA_BRA_F13168993             Meninges        6.7,8.4   \n",
       "122  HCA_BRA_F13774551             Meninges   9.2,9.9,12.5   \n",
       "123  HCA_BRA_F13774552             Meninges   9.2,9.9,12.5   \n",
       "124  HCA_BRA_F13774553             Meninges   9.2,9.9,12.5   \n",
       "125  HCA_BRA_F13774554             Meninges   9.2,9.9,12.5   \n",
       "126  HCA_BRA_F13774555             Meninges   9.2,9.9,12.5   \n",
       "127  HCA_BRA_F13774556             Meninges   9.2,9.9,12.5   \n",
       "\n",
       "        Anatomical_Site     Donor_ID   Technology  FACS_Sorting  \\\n",
       "11             calvaria    2356/2372  10X_multiome        Total   \n",
       "12             calvaria    2356/2372  10X_multiome        Total   \n",
       "13             calvaria    2356/2372  10X_multiome        Total   \n",
       "14             calvaria    2370/2369  10X_multiome        Total   \n",
       "15             calvaria    2370/2369  10X_multiome        Total   \n",
       "16            skullbase    2370/2369  10X_multiome        Total   \n",
       "22                  hip  multiplexed  10X_multiome        Total   \n",
       "23             hip/knee    2392/2394  10X_multiome        Total   \n",
       "24         hip/shoulder  multiplexed  10X_multiome        Total   \n",
       "32                 knee  multiplexed  10X_multiome        Total   \n",
       "37             shoulder    2392/2394  10X_multiome        Total   \n",
       "38             shoulder  multiplexed  10X_multiome        Total   \n",
       "50   calvaria/skullbase   2372/16403  10X_multiome        Total   \n",
       "51   calvaria/skullbase   2372/16403  10X_multiome        Total   \n",
       "52   calvaria/skullbase   2372/16403  10X_multiome        Total   \n",
       "53   calvaria/skullbase   2372/16403  10X_multiome        Total   \n",
       "118            cerebral  multiplexed  10X_multiome        Total   \n",
       "119             falcine  multiplexed  10X_multiome        Total   \n",
       "120           skullbase  multiplexed  10X_multiome        Total   \n",
       "121           skullbase  multiplexed  10X_multiome        Total   \n",
       "122         multiplexed  multiplexed  10X_multiome        Total   \n",
       "123         multiplexed  multiplexed  10X_multiome        Total   \n",
       "124         multiplexed  multiplexed  10X_multiome        Total   \n",
       "125              sb-ant  multiplexed  10X_multiome        Total   \n",
       "126             sb-post  multiplexed  10X_multiome        Total   \n",
       "127            sagittal  multiplexed  10X_multiome        Total   \n",
       "\n",
       "      Accession_ID         Study  Multiplexed  Expected_Donors  10X_Chemistry  \n",
       "11   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "12   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "13   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "14   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "15   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "16   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "22   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "23   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "24   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "32   E-MTAB-14385   To_et_al_2024         Yes                3  10x_multiome   \n",
       "37   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "38   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "50   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "51   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "52   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "53   E-MTAB-14385   To_et_al_2024         Yes                2  10x_multiome   \n",
       "118         Sanger         Sanger         Yes                2   10x_multiome  \n",
       "119         Sanger         Sanger         Yes                2   10x_multiome  \n",
       "120         Sanger         Sanger         Yes                2   10x_multiome  \n",
       "121         Sanger         Sanger         Yes                2   10x_multiome  \n",
       "122         Sanger         Sanger         Yes                3   10x_multiome  \n",
       "123         Sanger         Sanger         Yes                3   10x_multiome  \n",
       "124         Sanger         Sanger         Yes                3   10x_multiome  \n",
       "125         Sanger         Sanger         Yes                3   10x_multiome  \n",
       "126         Sanger         Sanger         Yes                3   10x_multiome  \n",
       "127         Sanger         Sanger         Yes                3   10x_multiome  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiplexed_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiplexed_samples = multiplexed_samples[['Age', 'Anatomical_Site', 'Donor_ID', \"Multiplexed\",\"Expected_Donors\"]]\n",
    "len(multiplexed_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiplexed_samples.to_csv('/nfs/team298/sm54/metadata/multiplexed_samples.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cross Organ Comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Panfetal Samples\n",
    "panfetal_md_paper= pd.read_csv('/nfs/team298/sm54/Pan_fetal_immune/metadata/sample_metadata.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pfm= pd.read_csv('/nfs/team298/sm54/metadata/missing_panfetal_samples.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10XChemistry/Method</th>\n",
       "      <th>10X_Chemistry/Method</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age.1</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Characteristics[developmental stage]</th>\n",
       "      <th>Characteristics[protocol]</th>\n",
       "      <th>Clinical_Info</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>...</th>\n",
       "      <th>Term Accession Number</th>\n",
       "      <th>Term Accession Number.1</th>\n",
       "      <th>Term Source REF</th>\n",
       "      <th>Term Source REF.1</th>\n",
       "      <th>Time_unit</th>\n",
       "      <th>Time_unit.1</th>\n",
       "      <th>Unit[time unit].1</th>\n",
       "      <th>Unit_time</th>\n",
       "      <th>Unit_time.1</th>\n",
       "      <th>input_Molecule</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  10XChemistry/Method 10X_Chemistry/Method Accession_ID                Age  \\\n",
       "0                 NaN            10x 3' v2  E-MTAB-7407  8 weeks gestation   \n",
       "1                 NaN            10x 3' v2  E-MTAB-7407  8 weeks gestation   \n",
       "2                 NaN            10x 3' v2  E-MTAB-7407  8 weeks gestation   \n",
       "3                 NaN            10x 3' v2  E-MTAB-7407  8 weeks gestation   \n",
       "4                 NaN            10x 3' v2  E-MTAB-7407  8 weeks gestation   \n",
       "\n",
       "   Age.1       Assay Name Characteristics[developmental stage]  \\\n",
       "0    NaN  4834STDY7002875                                  NaN   \n",
       "1    NaN  4834STDY7002876                                  NaN   \n",
       "2    NaN  4834STDY7002877                                  NaN   \n",
       "3    NaN  4834STDY7002878                                  NaN   \n",
       "4    NaN  4834STDY7002879                                  NaN   \n",
       "\n",
       "  Characteristics[protocol]    Clinical_Info Comment[BioSD_SAMPLE]  ...  \\\n",
       "0                       NaN  First trimister                   NaN  ...   \n",
       "1                       NaN  First trimister                   NaN  ...   \n",
       "2                       NaN  First trimister                   NaN  ...   \n",
       "3                       NaN  First trimister                   NaN  ...   \n",
       "4                       NaN  First trimister                   NaN  ...   \n",
       "\n",
       "  Term Accession Number Term Accession Number.1 Term Source REF  \\\n",
       "0                   NaN                     NaN             NaN   \n",
       "1                   NaN                     NaN             NaN   \n",
       "2                   NaN                     NaN             NaN   \n",
       "3                   NaN                     NaN             NaN   \n",
       "4                   NaN                     NaN             NaN   \n",
       "\n",
       "  Term Source REF.1 Time_unit Time_unit.1 Unit[time unit].1 Unit_time  \\\n",
       "0               NaN       NaN         NaN               NaN      week   \n",
       "1               NaN       NaN         NaN               NaN      week   \n",
       "2               NaN       NaN         NaN               NaN      week   \n",
       "3               NaN       NaN         NaN               NaN      week   \n",
       "4               NaN       NaN         NaN               NaN      week   \n",
       "\n",
       "  Unit_time.1 input_Molecule  \n",
       "0         NaN            NaN  \n",
       "1         NaN            NaN  \n",
       "2         NaN            NaN  \n",
       "3         NaN            NaN  \n",
       "4         NaN            NaN  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "# Set the directory path\n",
    "directory = '/nfs/team298/sm54//metadata/cross_organ_comparison/'\n",
    "\n",
    "# Use glob to find all CSV files in the directory\n",
    "csv_files = glob.glob(os.path.join(directory, '*.csv'))\n",
    "\n",
    "# Read and store all dataframes\n",
    "dfs = [pd.read_csv(f) for f in csv_files]\n",
    "\n",
    "# Concatenate dataframes, aligning columns and filling missing values with NaN\n",
    "combined_df = pd.concat(dfs, axis=0, ignore_index=True, sort=True)\n",
    "\n",
    "# Now `combined_df` contains the merged result\n",
    "(combined_df.head())  # Optional: view the first few rows\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "S= (set(panfetal_md_paper[\"File\"].unique()) - set(combined_df[\"Sample_ID\"]))\n",
    "Z=  set(combined_df[\"Sample_ID\"]) - set(panfetal_md_paper[\"File\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(S)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10XChemistry/Method</th>\n",
       "      <th>10X_Chemistry/Method</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age.1</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Characteristics[developmental stage]</th>\n",
       "      <th>Characteristics[protocol]</th>\n",
       "      <th>Clinical_Info</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>...</th>\n",
       "      <th>Term Accession Number.1</th>\n",
       "      <th>Term Source REF</th>\n",
       "      <th>Term Source REF.1</th>\n",
       "      <th>Time_unit</th>\n",
       "      <th>Time_unit.1</th>\n",
       "      <th>Unit[time unit].1</th>\n",
       "      <th>Unit_time</th>\n",
       "      <th>Unit_time.1</th>\n",
       "      <th>input_Molecule</th>\n",
       "      <th>duplicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>First trimister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-10552</td>\n",
       "      <td>carnegie stage 18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FCAImmP8096693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA9196809</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-10552</td>\n",
       "      <td>carnegie stage 18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FCAImmP8096694</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA9196811</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-10552</td>\n",
       "      <td>carnegie stage 18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FCAImmP8096695</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA9196814</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-10552</td>\n",
       "      <td>carnegie stage 18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FCAImmP8096696</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA9196813</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-10552</td>\n",
       "      <td>carnegie stage 18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FCAImmP8102304</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA9196812</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    10XChemistry/Method 10X_Chemistry/Method  Accession_ID                Age  \\\n",
       "0                   NaN            10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "1                   NaN            10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "2                   NaN            10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "3                   NaN            10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "4                   NaN            10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "..                  ...                  ...           ...                ...   \n",
       "251                 NaN                 5GEX  E-MTAB-10552  carnegie stage 18   \n",
       "252                 NaN                 5GEX  E-MTAB-10552  carnegie stage 18   \n",
       "253                 NaN                 5GEX  E-MTAB-10552  carnegie stage 18   \n",
       "254                 NaN                 5GEX  E-MTAB-10552  carnegie stage 18   \n",
       "255                 NaN                 5GEX  E-MTAB-10552  carnegie stage 18   \n",
       "\n",
       "     Age.1       Assay Name Characteristics[developmental stage]  \\\n",
       "0      NaN  4834STDY7002875                                  NaN   \n",
       "1      NaN  4834STDY7002876                                  NaN   \n",
       "2      NaN  4834STDY7002877                                  NaN   \n",
       "3      NaN  4834STDY7002878                                  NaN   \n",
       "4      NaN  4834STDY7002879                                  NaN   \n",
       "..     ...              ...                                  ...   \n",
       "251    NaN   FCAImmP8096693                                  NaN   \n",
       "252    NaN   FCAImmP8096694                                  NaN   \n",
       "253    NaN   FCAImmP8096695                                  NaN   \n",
       "254    NaN   FCAImmP8096696                                  NaN   \n",
       "255    NaN   FCAImmP8102304                                  NaN   \n",
       "\n",
       "    Characteristics[protocol]    Clinical_Info Comment[BioSD_SAMPLE]  ...  \\\n",
       "0                         NaN  First trimister                   NaN  ...   \n",
       "1                         NaN  First trimister                   NaN  ...   \n",
       "2                         NaN  First trimister                   NaN  ...   \n",
       "3                         NaN  First trimister                   NaN  ...   \n",
       "4                         NaN  First trimister                   NaN  ...   \n",
       "..                        ...              ...                   ...  ...   \n",
       "251                       NaN              NaN          SAMEA9196809  ...   \n",
       "252                       NaN              NaN          SAMEA9196811  ...   \n",
       "253                       NaN              NaN          SAMEA9196814  ...   \n",
       "254                       NaN              NaN          SAMEA9196813  ...   \n",
       "255                       NaN              NaN          SAMEA9196812  ...   \n",
       "\n",
       "    Term Accession Number.1 Term Source REF Term Source REF.1 Time_unit  \\\n",
       "0                       NaN             NaN               NaN       NaN   \n",
       "1                       NaN             NaN               NaN       NaN   \n",
       "2                       NaN             NaN               NaN       NaN   \n",
       "3                       NaN             NaN               NaN       NaN   \n",
       "4                       NaN             NaN               NaN       NaN   \n",
       "..                      ...             ...               ...       ...   \n",
       "251                     NaN             NaN               NaN       NaN   \n",
       "252                     NaN             NaN               NaN       NaN   \n",
       "253                     NaN             NaN               NaN       NaN   \n",
       "254                     NaN             NaN               NaN       NaN   \n",
       "255                     NaN             NaN               NaN       NaN   \n",
       "\n",
       "    Time_unit.1 Unit[time unit].1 Unit_time Unit_time.1 input_Molecule  \\\n",
       "0           NaN               NaN      week         NaN            NaN   \n",
       "1           NaN               NaN      week         NaN            NaN   \n",
       "2           NaN               NaN      week         NaN            NaN   \n",
       "3           NaN               NaN      week         NaN            NaN   \n",
       "4           NaN               NaN      week         NaN            NaN   \n",
       "..          ...               ...       ...         ...            ...   \n",
       "251         NaN               NaN       NaN         NaN            NaN   \n",
       "252         NaN               NaN       NaN         NaN            NaN   \n",
       "253         NaN               NaN       NaN         NaN            NaN   \n",
       "254         NaN               NaN       NaN         NaN            NaN   \n",
       "255         NaN               NaN       NaN         NaN            NaN   \n",
       "\n",
       "    duplicated  \n",
       "0           no  \n",
       "1           no  \n",
       "2           no  \n",
       "3           no  \n",
       "4           no  \n",
       "..         ...  \n",
       "251         no  \n",
       "252         no  \n",
       "253         no  \n",
       "254         no  \n",
       "255         no  \n",
       "\n",
       "[256 rows x 77 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify duplicated Sample_IDs (all occurrences)\n",
    "is_duplicated = combined_df['Sample_ID'].duplicated(keep=False)\n",
    "\n",
    "# Identify first occurrence among duplicates\n",
    "is_first = combined_df['Sample_ID'].duplicated(keep='first')\n",
    "\n",
    "# Keep only the first occurrence of each duplicated Sample_ID\n",
    "subset_df = combined_df[is_duplicated & ~is_first | (combined_df['Sample_ID'].duplicated(keep='first') == False)]\n",
    "\n",
    "# Add the 'duplicated' column: 'yes' for duplicated Sample_IDs\n",
    "subset_df = subset_df.copy()  # Avoid SettingWithCopyWarning\n",
    "subset_df['duplicated'] = subset_df['Sample_ID'].duplicated(keep=False).map({True: 'yes', False: \"no\"})\n",
    "\n",
    "# Reset index if needed\n",
    "subset_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# View the result\n",
    "(subset_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "95 + 17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_df.to_csv('cross_organ_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "256"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(subset_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "O= (set(panfetal_md_paper[\"File\"].unique()) - set(subset_df[\"Sample_ID\"]))\n",
    "Z=  set(subset_df[\"Sample_ID\"]) - set(panfetal_md_paper[\"File\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I downloaded the cross organ metadata, removed unnecessary columns and tidied up info the new metadata has \n",
    "# 256-15 samples. I will now add the missing panfetal samples to this list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "S= (set(panfetal_md_paper[\"File\"].unique()) - set(combined_df[\"Sample_ID\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_pf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the tidied metadata and add the 28 samples from the panfetal paper `"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "panfetal_md_paper= pd.read_csv('/nfs/team298/sm54/Pan_fetal_immune/metadata/sample_metadata.csv', index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "panfetal_md_paper= pd.read_csv('/nfs/team298/sm54/Pan_fetal_immune/metadata/sample_metadata.csv', index_col=0)\n",
    "combined_metadata= pd.read_csv('/nfs/team298/sm54/metadata/cross_organ_metadata_with_columns_removed.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "S= (set(panfetal_md_paper[\"File\"].unique()) - set(combined_metadata[\"Sample_ID\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_pf= panfetal_md_paper[panfetal_md_paper[\"File\"].isin(S)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove bone marrow samples \n",
    "missing_pf_final= missing_pf[missing_pf[\"Organ\"]!=\"BM\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-30-c4f2651437dd>:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  missing_pf_final[\"Accession_ID\"]='E-MTAB-11343'\n",
      "<ipython-input-30-c4f2651437dd>:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  missing_pf_final[\"Study \"]='Suo_Dann_et_al_2022'\n",
      "<ipython-input-30-c4f2651437dd>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  missing_pf_final[\"Technology\"]='RNA-Seq'\n"
     ]
    }
   ],
   "source": [
    "missing_pf_final[\"Accession_ID\"]='E-MTAB-11343'\n",
    "missing_pf_final[\"Study \"]='Suo_Dann_et_al_2022'\n",
    "missing_pf_final[\"Technology\"]='RNA-Seq'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_pf_final = missing_pf_final.rename(columns={\n",
    "    'File': 'Sample_ID',\n",
    "    'Donor': 'Donor_ID',\n",
    "    'Sort_id': \"FACS_Sorting\",\n",
    "    'Method': \"10X_Chemistry\",\n",
    "    'Anatomical_region':\"Anatomical_Site\"\n",
    "    \n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df_final = pd.concat([combined_metadata, missing_pf_final], ignore_index=True, sort=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "269"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(combined_df_final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "269"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(combined_df_final[\"Accession_ID\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df_final.to_csv('/nfs/team298/sm54/metadata/cross_organ_metadata_with_panfetal.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Samples from E-MTAB-10552 (YS) have 3 prime in the EMTAB metadata and 5 prime in YS metadata csv from paper. I went with what's available on EMTAB-; reading the csv back in after correction "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now for samples which are present in panfetal paper, I want to add the harmonized metadata from that paper to the metadata frame using Sample_ID from combined metadata and \n",
    "# file from panfetal paper "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata= pd.read_csv('/nfs/team298/sm54/metadata/cross_organ_metadata_with_panfetal_edited.csv', index_col=0)\n",
    "pf= pd.read_csv('/nfs/team298/sm54/Pan_fetal_immune/metadata/sample_metadata.csv', index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "269"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Sort_id</th>\n",
       "      <th>Age</th>\n",
       "      <th>Method</th>\n",
       "      <th>Donor</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Processing_method</th>\n",
       "      <th>Anatomical_region</th>\n",
       "      <th>N_cells</th>\n",
       "      <th>Termination_method</th>\n",
       "      <th>abTCR_library_id</th>\n",
       "      <th>gdTCR_library_id</th>\n",
       "      <th>BCR_library_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>FCAImmP7803020</td>\n",
       "      <td>SP</td>\n",
       "      <td>CD45P</td>\n",
       "      <td>15</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>F50</td>\n",
       "      <td>F50_SP_CD45P_FCAImmP7803020</td>\n",
       "      <td>female</td>\n",
       "      <td>Collagnase</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5632</td>\n",
       "      <td>Medical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSSS_F_Imm10014575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>FCAImmP7803020</td>\n",
       "      <td>SP</td>\n",
       "      <td>CD45P</td>\n",
       "      <td>15</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>F50</td>\n",
       "      <td>F50_SP_CD45P_FCAImmP7803020</td>\n",
       "      <td>female</td>\n",
       "      <td>Collagnase</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5632</td>\n",
       "      <td>Medical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FCAImmP7812649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>FCAImmP7803021</td>\n",
       "      <td>SP</td>\n",
       "      <td>CD45P</td>\n",
       "      <td>15</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>F50</td>\n",
       "      <td>F50_SP_CD45P_FCAImmP7803021</td>\n",
       "      <td>female</td>\n",
       "      <td>Collagnase</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5880</td>\n",
       "      <td>Medical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSSS_F_Imm10424060</td>\n",
       "      <td>WSSS_F_Imm10014576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>FCAImmP7803021</td>\n",
       "      <td>SP</td>\n",
       "      <td>CD45P</td>\n",
       "      <td>15</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>F50</td>\n",
       "      <td>F50_SP_CD45P_FCAImmP7803021</td>\n",
       "      <td>female</td>\n",
       "      <td>Collagnase</td>\n",
       "      <td>unknown</td>\n",
       "      <td>5880</td>\n",
       "      <td>Medical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSSS_F_Imm10424060</td>\n",
       "      <td>FCAImmP7812650</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               File Organ Sort_id  Age Method Donor  \\\n",
       "167  FCAImmP7803020    SP   CD45P   15   5GEX   F50   \n",
       "168  FCAImmP7803020    SP   CD45P   15   5GEX   F50   \n",
       "165  FCAImmP7803021    SP   CD45P   15   5GEX   F50   \n",
       "166  FCAImmP7803021    SP   CD45P   15   5GEX   F50   \n",
       "\n",
       "                          Sample     Sex Processing_method Anatomical_region  \\\n",
       "167  F50_SP_CD45P_FCAImmP7803020  female        Collagnase           unknown   \n",
       "168  F50_SP_CD45P_FCAImmP7803020  female        Collagnase           unknown   \n",
       "165  F50_SP_CD45P_FCAImmP7803021  female        Collagnase           unknown   \n",
       "166  F50_SP_CD45P_FCAImmP7803021  female        Collagnase           unknown   \n",
       "\n",
       "     N_cells Termination_method abTCR_library_id    gdTCR_library_id  \\\n",
       "167     5632            Medical              NaN                 NaN   \n",
       "168     5632            Medical              NaN                 NaN   \n",
       "165     5880            Medical              NaN  WSSS_F_Imm10424060   \n",
       "166     5880            Medical              NaN  WSSS_F_Imm10424060   \n",
       "\n",
       "         BCR_library_id  \n",
       "167  WSSS_F_Imm10014575  \n",
       "168      FCAImmP7812649  \n",
       "165  WSSS_F_Imm10014576  \n",
       "166      FCAImmP7812650  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show duplicated File values in pf\n",
    "duplicated_pf = pf[pf['File'].duplicated(keep=False)]\n",
    "\n",
    "# Print how many and inspect\n",
    "(duplicated_pf.sort_values('File'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "pf = pf.drop_duplicates(subset='File', keep='first')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Sort_id</th>\n",
       "      <th>Age</th>\n",
       "      <th>Method</th>\n",
       "      <th>Donor</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Processing_method</th>\n",
       "      <th>Anatomical_region</th>\n",
       "      <th>N_cells</th>\n",
       "      <th>Termination_method</th>\n",
       "      <th>abTCR_library_id</th>\n",
       "      <th>gdTCR_library_id</th>\n",
       "      <th>BCR_library_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [File, Organ, Sort_id, Age, Method, Donor, Sample, Sex, Processing_method, Anatomical_region, N_cells, Termination_method, abTCR_library_id, gdTCR_library_id, BCR_library_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show duplicated File values in pf\n",
    "duplicated_pf = pf[pf['File'].duplicated(keep=False)]\n",
    "\n",
    "# Print how many and inspect\n",
    "(duplicated_pf.sort_values('File'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-81-e6c1be5ee01c>:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pf['File'] = pf['File'].astype(str).str.strip()\n"
     ]
    }
   ],
   "source": [
    "metadata['Sample_ID'] = metadata['Sample_ID'].astype(str).str.strip()\n",
    "pf['File'] = pf['File'].astype(str).str.strip()\n",
    "pf_renamed = pf.rename(columns={col: f\"{col}_pf\" for col in pf.columns if col != 'File'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10XChemistry/Method</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>BCR_library_id</th>\n",
       "      <th>Comment[library construction]</th>\n",
       "      <th>Comment[technical replicate group]</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>End_Bias</th>\n",
       "      <th>...</th>\n",
       "      <th>Run_ID</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Study</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Termination_method</th>\n",
       "      <th>abTCR_library_id</th>\n",
       "      <th>gdTCR_library_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>kidney</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F4</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>...</td>\n",
       "      <td>4834STDY7002875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Popescue_et_al_2019</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>kidney</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F4</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>...</td>\n",
       "      <td>4834STDY7002876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Popescue_et_al_2019</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>liver</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F4</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>...</td>\n",
       "      <td>4834STDY7002877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Popescue_et_al_2019</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>liver</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F4</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>...</td>\n",
       "      <td>4834STDY7002878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Popescue_et_al_2019</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>E-MTAB-7407</td>\n",
       "      <td>8 weeks gestation</td>\n",
       "      <td>skin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F4</td>\n",
       "      <td>3 prime tag</td>\n",
       "      <td>...</td>\n",
       "      <td>4834STDY7002879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4834STDY7002879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Popescue_et_al_2019</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-11343</td>\n",
       "      <td>17</td>\n",
       "      <td>LargeInt</td>\n",
       "      <td>FCA_gut8090119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78_GU_CD45P_CD45N_FCA_gut8015060</td>\n",
       "      <td>FCA_gut8015060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Suo_Dann_et_al_2022</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Medical</td>\n",
       "      <td>FCA_gut8090114</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-11343</td>\n",
       "      <td>17</td>\n",
       "      <td>SmallInt</td>\n",
       "      <td>FCA_gut8090118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78_GU_CD45P_CD45N_FCA_gut8015059</td>\n",
       "      <td>FCA_gut8015059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Suo_Dann_et_al_2022</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Medical</td>\n",
       "      <td>FCA_gut8090113</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-11343</td>\n",
       "      <td>17</td>\n",
       "      <td>SmallInt</td>\n",
       "      <td>FCA_gut8090117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78_GU_CD45P_CD45N_FCA_gut8015058</td>\n",
       "      <td>FCA_gut8015058</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Suo_Dann_et_al_2022</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Medical</td>\n",
       "      <td>FCA_gut8090112</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-11343</td>\n",
       "      <td>17</td>\n",
       "      <td>MLN</td>\n",
       "      <td>FCA_gut8090120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78_MLN_CD45P_CD45N_FCA_gut8015061</td>\n",
       "      <td>FCA_gut8015061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Suo_Dann_et_al_2022</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Medical</td>\n",
       "      <td>FCA_gut8090115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>E-MTAB-11343</td>\n",
       "      <td>17</td>\n",
       "      <td>SmallInt</td>\n",
       "      <td>FCA_gut8090116</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F78_GU_CD45P_CD45N_FCA_gut8015057</td>\n",
       "      <td>FCA_gut8015057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>Suo_Dann_et_al_2022</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Medical</td>\n",
       "      <td>FCA_gut8090111</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>269 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    10XChemistry/Method 10X_Chemistry  Accession_ID                Age  \\\n",
       "0                   NaN     10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "1                   NaN     10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "2                   NaN     10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "3                   NaN     10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "4                   NaN     10x 3' v2   E-MTAB-7407  8 weeks gestation   \n",
       "..                  ...           ...           ...                ...   \n",
       "264                 NaN          5GEX  E-MTAB-11343                 17   \n",
       "265                 NaN          5GEX  E-MTAB-11343                 17   \n",
       "266                 NaN          5GEX  E-MTAB-11343                 17   \n",
       "267                 NaN          5GEX  E-MTAB-11343                 17   \n",
       "268                 NaN          5GEX  E-MTAB-11343                 17   \n",
       "\n",
       "    Anatomical_Site  BCR_library_id Comment[library construction]  \\\n",
       "0            kidney             NaN                           NaN   \n",
       "1            kidney             NaN                           NaN   \n",
       "2             liver             NaN                           NaN   \n",
       "3             liver             NaN                           NaN   \n",
       "4              skin             NaN                           NaN   \n",
       "..              ...             ...                           ...   \n",
       "264        LargeInt  FCA_gut8090119                           NaN   \n",
       "265        SmallInt  FCA_gut8090118                           NaN   \n",
       "266        SmallInt  FCA_gut8090117                           NaN   \n",
       "267             MLN  FCA_gut8090120                           NaN   \n",
       "268        SmallInt  FCA_gut8090116                           NaN   \n",
       "\n",
       "    Comment[technical replicate group] Donor_ID     End_Bias  ...  \\\n",
       "0                                  NaN       F4  3 prime tag  ...   \n",
       "1                                  NaN       F4  3 prime tag  ...   \n",
       "2                                  NaN       F4  3 prime tag  ...   \n",
       "3                                  NaN       F4  3 prime tag  ...   \n",
       "4                                  NaN       F4  3 prime tag  ...   \n",
       "..                                 ...      ...          ...  ...   \n",
       "264                                NaN      F78          NaN  ...   \n",
       "265                                NaN      F78          NaN  ...   \n",
       "266                                NaN      F78          NaN  ...   \n",
       "267                                NaN      F78          NaN  ...   \n",
       "268                                NaN      F78          NaN  ...   \n",
       "\n",
       "              Run_ID                              Sample        Sample_ID  \\\n",
       "0    4834STDY7002875                                 NaN  4834STDY7002875   \n",
       "1    4834STDY7002876                                 NaN  4834STDY7002876   \n",
       "2    4834STDY7002877                                 NaN  4834STDY7002877   \n",
       "3    4834STDY7002878                                 NaN  4834STDY7002878   \n",
       "4    4834STDY7002879                                 NaN  4834STDY7002879   \n",
       "..               ...                                 ...              ...   \n",
       "264              NaN   F78_GU_CD45P_CD45N_FCA_gut8015060   FCA_gut8015060   \n",
       "265              NaN   F78_GU_CD45P_CD45N_FCA_gut8015059   FCA_gut8015059   \n",
       "266              NaN   F78_GU_CD45P_CD45N_FCA_gut8015058   FCA_gut8015058   \n",
       "267              NaN  F78_MLN_CD45P_CD45N_FCA_gut8015061   FCA_gut8015061   \n",
       "268              NaN   F78_GU_CD45P_CD45N_FCA_gut8015057   FCA_gut8015057   \n",
       "\n",
       "    Sampling_Site   Sex               Study  Technology Termination_method  \\\n",
       "0             NaN  male  Popescue_et_al_2019    RNA-Seq                NaN   \n",
       "1             NaN  male  Popescue_et_al_2019    RNA-Seq                NaN   \n",
       "2             NaN  male  Popescue_et_al_2019    RNA-Seq                NaN   \n",
       "3             NaN  male  Popescue_et_al_2019    RNA-Seq                NaN   \n",
       "4             NaN  male  Popescue_et_al_2019    RNA-Seq                NaN   \n",
       "..            ...   ...                  ...        ...                ...   \n",
       "264           NaN  male  Suo_Dann_et_al_2022    RNA-Seq            Medical   \n",
       "265           NaN  male  Suo_Dann_et_al_2022    RNA-Seq            Medical   \n",
       "266           NaN  male  Suo_Dann_et_al_2022    RNA-Seq            Medical   \n",
       "267           NaN  male  Suo_Dann_et_al_2022    RNA-Seq            Medical   \n",
       "268           NaN  male  Suo_Dann_et_al_2022    RNA-Seq            Medical   \n",
       "\n",
       "    abTCR_library_id gdTCR_library_id  \n",
       "0                NaN              NaN  \n",
       "1                NaN              NaN  \n",
       "2                NaN              NaN  \n",
       "3                NaN              NaN  \n",
       "4                NaN              NaN  \n",
       "..               ...              ...  \n",
       "264   FCA_gut8090114              NaN  \n",
       "265   FCA_gut8090113              NaN  \n",
       "266   FCA_gut8090112              NaN  \n",
       "267   FCA_gut8090115              NaN  \n",
       "268   FCA_gut8090111              NaN  \n",
       "\n",
       "[269 rows x 29 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df = metadata.merge(pf_renamed, left_on='Sample_ID', right_on='File', how='left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.to_csv('merged.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/nfs/team298/sm54'"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (197865849.py, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Input \u001b[0;32mIn [23]\u001b[0;36m\u001b[0m\n\u001b[0;31m    FCAImmP7528295 from liver paper have 10x3'vs in EMAT7407 but 5 prime tag.. so I corrected these\u001b[0m\n\u001b[0m                   ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "FCAImmP7292028\n",
    "FCAImmP7292032\n",
    "FCAImmP7528295 from liver paper have 10x3'vs in EMAT7407 but 5 prime tag.. so I corrected these \n",
    "\n",
    "# YS are 5'prime but EMAT said 3'prime bias so I also corrected these based on the info below\n",
    "# Also removed 2 samples from MLN(2) as they are not enough  for a meaningful comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata= sc.read_h5ad('/nfs/team298/sm54/published_datasets/ys_portal_object_20221208.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sequencing.type</th>\n",
       "      <th>3earlye</th>\n",
       "      <th>5GEX</th>\n",
       "      <th>TBC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lanes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FCAImmP7316900</th>\n",
       "      <td>1001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7462244</th>\n",
       "      <td>1383</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7504909</th>\n",
       "      <td>2566</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7504910</th>\n",
       "      <td>2197</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7504911</th>\n",
       "      <td>2019</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7504912</th>\n",
       "      <td>1985</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7504913</th>\n",
       "      <td>1962</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7862088</th>\n",
       "      <td>0</td>\n",
       "      <td>6317</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7862089</th>\n",
       "      <td>0</td>\n",
       "      <td>5611</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7862090</th>\n",
       "      <td>0</td>\n",
       "      <td>3705</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7862091</th>\n",
       "      <td>0</td>\n",
       "      <td>10429</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7862092</th>\n",
       "      <td>0</td>\n",
       "      <td>9709</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP7862093</th>\n",
       "      <td>0</td>\n",
       "      <td>10174</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096686</th>\n",
       "      <td>0</td>\n",
       "      <td>1337</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096687</th>\n",
       "      <td>0</td>\n",
       "      <td>10112</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096688</th>\n",
       "      <td>0</td>\n",
       "      <td>5028</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096689</th>\n",
       "      <td>0</td>\n",
       "      <td>4073</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096690</th>\n",
       "      <td>0</td>\n",
       "      <td>7434</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096691</th>\n",
       "      <td>0</td>\n",
       "      <td>8013</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096692</th>\n",
       "      <td>0</td>\n",
       "      <td>1752</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096693</th>\n",
       "      <td>0</td>\n",
       "      <td>5324</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096694</th>\n",
       "      <td>0</td>\n",
       "      <td>5055</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096695</th>\n",
       "      <td>0</td>\n",
       "      <td>4680</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8096696</th>\n",
       "      <td>0</td>\n",
       "      <td>2114</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCAImmP8102304</th>\n",
       "      <td>0</td>\n",
       "      <td>6909</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vitelline V-CS14/4.5w</th>\n",
       "      <td>0</td>\n",
       "      <td>1233</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WS_wEMB12142157</th>\n",
       "      <td>0</td>\n",
       "      <td>4700</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WS_wEMB12142158</th>\n",
       "      <td>0</td>\n",
       "      <td>5558</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WS_wEMB12142159</th>\n",
       "      <td>0</td>\n",
       "      <td>4405</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WS_wEMB12142160</th>\n",
       "      <td>0</td>\n",
       "      <td>9321</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WS_wEMB12142161</th>\n",
       "      <td>0</td>\n",
       "      <td>7696</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wang_CS10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wang_CS11</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YolkSac-CS14/4.5w</th>\n",
       "      <td>0</td>\n",
       "      <td>5023</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "sequencing.type        3earlye   5GEX   TBC\n",
       "lanes                                      \n",
       "FCAImmP7316900            1001      0     0\n",
       "FCAImmP7462244            1383      0     0\n",
       "FCAImmP7504909            2566      0     0\n",
       "FCAImmP7504910            2197      0     0\n",
       "FCAImmP7504911            2019      0     0\n",
       "FCAImmP7504912            1985      0     0\n",
       "FCAImmP7504913            1962      0     0\n",
       "FCAImmP7862088               0   6317     0\n",
       "FCAImmP7862089               0   5611     0\n",
       "FCAImmP7862090               0   3705     0\n",
       "FCAImmP7862091               0  10429     0\n",
       "FCAImmP7862092               0   9709     0\n",
       "FCAImmP7862093               0  10174     0\n",
       "FCAImmP8096686               0   1337     0\n",
       "FCAImmP8096687               0  10112     0\n",
       "FCAImmP8096688               0   5028     0\n",
       "FCAImmP8096689               0   4073     0\n",
       "FCAImmP8096690               0   7434     0\n",
       "FCAImmP8096691               0   8013     0\n",
       "FCAImmP8096692               0   1752     0\n",
       "FCAImmP8096693               0   5324     0\n",
       "FCAImmP8096694               0   5055     0\n",
       "FCAImmP8096695               0   4680     0\n",
       "FCAImmP8096696               0   2114     0\n",
       "FCAImmP8102304               0   6909     0\n",
       "Vitelline V-CS14/4.5w        0   1233     0\n",
       "WS_wEMB12142157              0   4700     0\n",
       "WS_wEMB12142158              0   5558     0\n",
       "WS_wEMB12142159              0   4405     0\n",
       "WS_wEMB12142160              0   9321     0\n",
       "WS_wEMB12142161              0   7696     0\n",
       "Wang_CS10                    0      0  3615\n",
       "Wang_CS11                    0      0  7054\n",
       "YolkSac-CS14/4.5w            0   5023     0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.crosstab(adata.obs['lanes'], adata.obs['sequencing.type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>component</th>\n",
       "      <th>stage</th>\n",
       "      <th>sex</th>\n",
       "      <th>sort.ids</th>\n",
       "      <th>fetal.ids</th>\n",
       "      <th>orig.dataset</th>\n",
       "      <th>sequencing.type</th>\n",
       "      <th>lanes</th>\n",
       "      <th>LVL1</th>\n",
       "      <th>LVL2</th>\n",
       "      <th>LVL3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Wang_et_al_AAACCTGAGACCTTTG_1-0</th>\n",
       "      <td>Membrane</td>\n",
       "      <td>CS10</td>\n",
       "      <td>Male</td>\n",
       "      <td>Total</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>Wang_et_al_Cell_Stem Cell_2021</td>\n",
       "      <td>TBC</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>STROMA</td>\n",
       "      <td>ENDOTHELIUM</td>\n",
       "      <td>ENDOTHELIUM_SINUSOIDAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wang_et_al_AAACCTGCACTTAAGC_1-0</th>\n",
       "      <td>Membrane</td>\n",
       "      <td>CS10</td>\n",
       "      <td>Male</td>\n",
       "      <td>Total</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>Wang_et_al_Cell_Stem Cell_2021</td>\n",
       "      <td>TBC</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>PROGENITOR</td>\n",
       "      <td>HSPC</td>\n",
       "      <td>PRIM_HSPC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wang_et_al_AAACCTGGTAACGACG_1-0</th>\n",
       "      <td>Membrane</td>\n",
       "      <td>CS10</td>\n",
       "      <td>Male</td>\n",
       "      <td>Total</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>Wang_et_al_Cell_Stem Cell_2021</td>\n",
       "      <td>TBC</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wang_et_al_AAACCTGGTCAAAGAT_1-0</th>\n",
       "      <td>Membrane</td>\n",
       "      <td>CS10</td>\n",
       "      <td>Male</td>\n",
       "      <td>Total</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>Wang_et_al_Cell_Stem Cell_2021</td>\n",
       "      <td>TBC</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wang_et_al_AAACCTGGTCCGAAGA_1-0</th>\n",
       "      <td>Membrane</td>\n",
       "      <td>CS10</td>\n",
       "      <td>Male</td>\n",
       "      <td>Total</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>Wang_et_al_Cell_Stem Cell_2021</td>\n",
       "      <td>TBC</td>\n",
       "      <td>Wang_CS10</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCTCTAGC-1-5-3</th>\n",
       "      <td>VA</td>\n",
       "      <td>CS17</td>\n",
       "      <td>Female</td>\n",
       "      <td>Total</td>\n",
       "      <td>F138</td>\n",
       "      <td>WE_YS</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>WS_wEMB12142161</td>\n",
       "      <td>STROMA</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGCCGTGA-1-5-3</th>\n",
       "      <td>VA</td>\n",
       "      <td>CS17</td>\n",
       "      <td>Female</td>\n",
       "      <td>Total</td>\n",
       "      <td>F138</td>\n",
       "      <td>WE_YS</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>WS_wEMB12142161</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGTTTAGG-1-5-3</th>\n",
       "      <td>VA</td>\n",
       "      <td>CS17</td>\n",
       "      <td>Female</td>\n",
       "      <td>Total</td>\n",
       "      <td>F138</td>\n",
       "      <td>WE_YS</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>WS_wEMB12142161</td>\n",
       "      <td>MYELOID</td>\n",
       "      <td>MACROPHAGE</td>\n",
       "      <td>MACROPHAGE_IRON_RECYCLING</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGACCTC-1-5-3</th>\n",
       "      <td>VA</td>\n",
       "      <td>CS17</td>\n",
       "      <td>Female</td>\n",
       "      <td>Total</td>\n",
       "      <td>F138</td>\n",
       "      <td>WE_YS</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>WS_wEMB12142161</td>\n",
       "      <td>STROMA</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGCTTGC-1-5-3</th>\n",
       "      <td>VA</td>\n",
       "      <td>CS17</td>\n",
       "      <td>Female</td>\n",
       "      <td>Total</td>\n",
       "      <td>F138</td>\n",
       "      <td>WE_YS</td>\n",
       "      <td>5GEX</td>\n",
       "      <td>WS_wEMB12142161</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "      <td>ERYTHROID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>169494 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                component stage     sex sort.ids  fetal.ids  \\\n",
       "Wang_et_al_AAACCTGAGACCTTTG_1-0  Membrane  CS10    Male    Total  Wang_CS10   \n",
       "Wang_et_al_AAACCTGCACTTAAGC_1-0  Membrane  CS10    Male    Total  Wang_CS10   \n",
       "Wang_et_al_AAACCTGGTAACGACG_1-0  Membrane  CS10    Male    Total  Wang_CS10   \n",
       "Wang_et_al_AAACCTGGTCAAAGAT_1-0  Membrane  CS10    Male    Total  Wang_CS10   \n",
       "Wang_et_al_AAACCTGGTCCGAAGA_1-0  Membrane  CS10    Male    Total  Wang_CS10   \n",
       "...                                   ...   ...     ...      ...        ...   \n",
       "TTTGTCATCCTCTAGC-1-5-3                 VA  CS17  Female    Total       F138   \n",
       "TTTGTCATCGCCGTGA-1-5-3                 VA  CS17  Female    Total       F138   \n",
       "TTTGTCATCGTTTAGG-1-5-3                 VA  CS17  Female    Total       F138   \n",
       "TTTGTCATCTGACCTC-1-5-3                 VA  CS17  Female    Total       F138   \n",
       "TTTGTCATCTGCTTGC-1-5-3                 VA  CS17  Female    Total       F138   \n",
       "\n",
       "                                                   orig.dataset  \\\n",
       "Wang_et_al_AAACCTGAGACCTTTG_1-0  Wang_et_al_Cell_Stem Cell_2021   \n",
       "Wang_et_al_AAACCTGCACTTAAGC_1-0  Wang_et_al_Cell_Stem Cell_2021   \n",
       "Wang_et_al_AAACCTGGTAACGACG_1-0  Wang_et_al_Cell_Stem Cell_2021   \n",
       "Wang_et_al_AAACCTGGTCAAAGAT_1-0  Wang_et_al_Cell_Stem Cell_2021   \n",
       "Wang_et_al_AAACCTGGTCCGAAGA_1-0  Wang_et_al_Cell_Stem Cell_2021   \n",
       "...                                                         ...   \n",
       "TTTGTCATCCTCTAGC-1-5-3                                    WE_YS   \n",
       "TTTGTCATCGCCGTGA-1-5-3                                    WE_YS   \n",
       "TTTGTCATCGTTTAGG-1-5-3                                    WE_YS   \n",
       "TTTGTCATCTGACCTC-1-5-3                                    WE_YS   \n",
       "TTTGTCATCTGCTTGC-1-5-3                                    WE_YS   \n",
       "\n",
       "                                sequencing.type            lanes        LVL1  \\\n",
       "Wang_et_al_AAACCTGAGACCTTTG_1-0             TBC        Wang_CS10      STROMA   \n",
       "Wang_et_al_AAACCTGCACTTAAGC_1-0             TBC        Wang_CS10  PROGENITOR   \n",
       "Wang_et_al_AAACCTGGTAACGACG_1-0             TBC        Wang_CS10   ERYTHROID   \n",
       "Wang_et_al_AAACCTGGTCAAAGAT_1-0             TBC        Wang_CS10   ERYTHROID   \n",
       "Wang_et_al_AAACCTGGTCCGAAGA_1-0             TBC        Wang_CS10   ERYTHROID   \n",
       "...                                         ...              ...         ...   \n",
       "TTTGTCATCCTCTAGC-1-5-3                     5GEX  WS_wEMB12142161      STROMA   \n",
       "TTTGTCATCGCCGTGA-1-5-3                     5GEX  WS_wEMB12142161   ERYTHROID   \n",
       "TTTGTCATCGTTTAGG-1-5-3                     5GEX  WS_wEMB12142161     MYELOID   \n",
       "TTTGTCATCTGACCTC-1-5-3                     5GEX  WS_wEMB12142161      STROMA   \n",
       "TTTGTCATCTGCTTGC-1-5-3                     5GEX  WS_wEMB12142161   ERYTHROID   \n",
       "\n",
       "                                        LVL2                       LVL3  \n",
       "Wang_et_al_AAACCTGAGACCTTTG_1-0  ENDOTHELIUM     ENDOTHELIUM_SINUSOIDAL  \n",
       "Wang_et_al_AAACCTGCACTTAAGC_1-0         HSPC                PRIM_HSPC_1  \n",
       "Wang_et_al_AAACCTGGTAACGACG_1-0    ERYTHROID                  ERYTHROID  \n",
       "Wang_et_al_AAACCTGGTCAAAGAT_1-0    ERYTHROID                  ERYTHROID  \n",
       "Wang_et_al_AAACCTGGTCCGAAGA_1-0    ERYTHROID                  ERYTHROID  \n",
       "...                                      ...                        ...  \n",
       "TTTGTCATCCTCTAGC-1-5-3            FIBROBLAST                 FIBROBLAST  \n",
       "TTTGTCATCGCCGTGA-1-5-3             ERYTHROID                  ERYTHROID  \n",
       "TTTGTCATCGTTTAGG-1-5-3            MACROPHAGE  MACROPHAGE_IRON_RECYCLING  \n",
       "TTTGTCATCTGACCTC-1-5-3            FIBROBLAST                 FIBROBLAST  \n",
       "TTTGTCATCTGCTTGC-1-5-3             ERYTHROID                  ERYTHROID  \n",
       "\n",
       "[169494 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 169494 × 37318\n",
       "    obs: 'component', 'stage', 'sex', 'sort.ids', 'fetal.ids', 'orig.dataset', 'sequencing.type', 'lanes', 'LVL1', 'LVL2', 'LVL3'\n",
       "    obsm: 'X_umap'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = [\n",
    "    \"FCAImmP8096686\",\n",
    "    \"FCAImmP8096687\",\n",
    "    \"FCAImmP8096688\",\n",
    "    \"FCAImmP8096689\",\n",
    "    \"FCAImmP8096690\",\n",
    "    \"FCAImmP8096691\",\n",
    "    \"FCAImmP8096692\",\n",
    "    \"FCAImmP8096693\",\n",
    "    \"FCAImmP8096694\",\n",
    "    \"FCAImmP8096695\",\n",
    "    \"FCAImmP8096696\",\n",
    "    \"FCAImmP8102304\",\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sequencing.type\n",
       "5GEX    61831\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[adata.obs['lanes'].isin(samples)].obs['sequencing.type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 61831 × 37318\n",
       "    obs: 'component', 'stage', 'sex', 'sort.ids', 'fetal.ids', 'orig.dataset', 'sequencing.type', 'lanes', 'LVL1', 'LVL2', 'LVL3'\n",
       "    obsm: 'X_umap'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[adata.obs['lanes'].isin(samples)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "pf= pd.read_csv('/nfs/team298/sm54/Pan_fetal_immune/metadata/sample_metadata.csv', index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Age\n",
       "12    14\n",
       "16     6\n",
       "11     4\n",
       "14     3\n",
       "7      2\n",
       "17     2\n",
       "9      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pf[pf['Organ']==\"TH\"][\"Age\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we have 279 samples for cross organ comparison with organs liver, yolksac, kidney, gut, skin, meninges  \n",
    "# 209 samples for bone atlas skull, spine, long bones (femur, fore limb, hind limb), hip, knee, shoulder\n",
    "# For the skull-brain-meninges, study, I will use the samples from the skull samples from bone atlas, meninges samples from cross organ and \n",
    "# brain data from linnarson lab (still waiting to hear back about that data).\n",
    "#For this, I will also need the whole embryo scrna-seq, visium and xenium data and head visium data from Ken "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading in the metadata "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "cross_organ= pd.read_csv('/nfs/team298/sm54/metadata/CROSS_ORGAN_METADATA.csv')\n",
    "bone_metadata= pd.read_csv('/nfs/team298/sm54/metadata/BONE_ATLAS_METADATA.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample_ID\n",
       "Human_colon_16S7985393    1\n",
       "FCAImmP7277563            1\n",
       "FCAImmP7803040            1\n",
       "FCAImmP7803039            1\n",
       "FCAImmP7803038            1\n",
       "                         ..\n",
       "FCAImmP7528291            1\n",
       "FCAImmP7528296            1\n",
       "FCAImmP7579212            1\n",
       "FCAImmP7579213            1\n",
       "HCA_BRA_F13774556         1\n",
       "Name: count, Length: 279, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cross_organ[\"Sample_ID\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(279, 44)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bone_metadata.shape\n",
    "cross_organ.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "cross_organ.columns = cross_organ.columns.str.strip()\n",
    "bone_metadata.columns = bone_metadata.columns.str.strip()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Combine the two DataFrames, aligning columns by name\n",
    "combined_df = pd.concat([bone_metadata, cross_organ], ignore_index=True, sort=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Age_pf</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Anatomical_region_pf</th>\n",
       "      <th>BCR_library_id</th>\n",
       "      <th>Comment[library construction]</th>\n",
       "      <th>Comment[technical replicate group]</th>\n",
       "      <th>...</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Study</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>9.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>488 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    10X_Chemistry  Accession_ID           Age      Age(PCW)  Age_pf  \\\n",
       "0    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "1    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "2    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "3    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "4    10x_multiome  E-MTAB-14385           9.4           NaN     NaN   \n",
       "..            ...           ...           ...           ...     ...   \n",
       "483  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "484  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "485  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "486  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "487  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "\n",
       "    Anatomical_Site Anatomical_region_pf BCR_library_id  \\\n",
       "0          calvaria                  NaN            NaN   \n",
       "1          calvaria                  NaN            NaN   \n",
       "2          calvaria                  NaN            NaN   \n",
       "3          calvaria                  NaN            NaN   \n",
       "4          calvaria                  NaN            NaN   \n",
       "..              ...                  ...            ...   \n",
       "483     multiplexed                  NaN            NaN   \n",
       "484     multiplexed                  NaN            NaN   \n",
       "485          sb-ant                  NaN            NaN   \n",
       "486         sb-post                  NaN            NaN   \n",
       "487        sagittal                  NaN            NaN   \n",
       "\n",
       "    Comment[library construction] Comment[technical replicate group]  ...  \\\n",
       "0                             NaN                                NaN  ...   \n",
       "1                             NaN                                NaN  ...   \n",
       "2                             NaN                                NaN  ...   \n",
       "3                             NaN                                NaN  ...   \n",
       "4                             NaN                                NaN  ...   \n",
       "..                            ...                                ...  ...   \n",
       "483                           NaN                                NaN  ...   \n",
       "484                           NaN                                NaN  ...   \n",
       "485                           NaN                                NaN  ...   \n",
       "486                           NaN                                NaN  ...   \n",
       "487                           NaN                                NaN  ...   \n",
       "\n",
       "    Sampling_Site      Sex Sex_pf Sort_id_pf          Study    Technology  \\\n",
       "0             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "1             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "2             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "3             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "4             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "..            ...      ...    ...        ...            ...           ...   \n",
       "483           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "484           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "485           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "486           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "487           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "\n",
       "    Termination_method_pf Unnamed: 14 abTCR_library_id_pf gdTCR_library_id_pf  \n",
       "0                     NaN         NaN                 NaN                 NaN  \n",
       "1                     NaN         NaN                 NaN                 NaN  \n",
       "2                     NaN         NaN                 NaN                 NaN  \n",
       "3                     NaN         NaN                 NaN                 NaN  \n",
       "4                     NaN         NaN                 NaN                 NaN  \n",
       "..                    ...         ...                 ...                 ...  \n",
       "483                   NaN         NaN                 NaN                 NaN  \n",
       "484                   NaN         NaN                 NaN                 NaN  \n",
       "485                   NaN         NaN                 NaN                 NaN  \n",
       "486                   NaN         NaN                 NaN                 NaN  \n",
       "487                   NaN         NaN                 NaN                 NaN  \n",
       "\n",
       "[488 rows x 45 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "488"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "209+ 279"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique to df1: {'Unnamed: 14'}\n",
      "Unique to df2: {'Anatomical_region_pf', 'Organ_pf', 'End_Bias', 'Comment[technical replicate group]', 'N_cells_pf', 'Donor_pf', 'Sampling_Site', 'Age(PCW)', 'Comment[library construction]', 'BCR_library_id', 'Age_pf', 'Organ_Orig', 'Sex', 'gdTCR_library_id_pf', 'Processing_method', 'Processing_method_pf', 'abTCR_library_id_pf', 'N_cells', 'Sample_ID_Copy', 'Method_pf', 'Sex_pf', 'Factor Value[facs sorting]', 'Organism', 'Sort_id_pf', 'Termination_method_pf', 'File_pf', 'Sample_pf', 'Sample'}\n"
     ]
    }
   ],
   "source": [
    "unique_to_df1 = set(bone_metadata.columns) - set(cross_organ.columns)\n",
    "unique_to_df2 = set(cross_organ.columns) - set(bone_metadata.columns)\n",
    "common = set(cross_organ.columns) & set(bone_metadata.columns)\n",
    "\n",
    "print(\"Unique to df1:\", unique_to_df1)\n",
    "print(\"Unique to df2:\", unique_to_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_to_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_to_df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(common)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv('/nfs/team298/sm54/metadata/combined_metadata_cross_organ_bone_atlas.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Age_pf</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Anatomical_region_pf</th>\n",
       "      <th>BCR_library_id</th>\n",
       "      <th>Comment[library construction]</th>\n",
       "      <th>Comment[technical replicate group]</th>\n",
       "      <th>...</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Study</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>9.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>488 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    10X_Chemistry  Accession_ID           Age      Age(PCW)  Age_pf  \\\n",
       "0    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "1    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "2    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "3    10x_multiome  E-MTAB-14385           8.5           NaN     NaN   \n",
       "4    10x_multiome  E-MTAB-14385           9.4           NaN     NaN   \n",
       "..            ...           ...           ...           ...     ...   \n",
       "483  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "484  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "485  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "486  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "487  10x_multiome        Sanger  9.2,9.9,12.5  9.2,9.9,12.5     NaN   \n",
       "\n",
       "    Anatomical_Site Anatomical_region_pf BCR_library_id  \\\n",
       "0          calvaria                  NaN            NaN   \n",
       "1          calvaria                  NaN            NaN   \n",
       "2          calvaria                  NaN            NaN   \n",
       "3          calvaria                  NaN            NaN   \n",
       "4          calvaria                  NaN            NaN   \n",
       "..              ...                  ...            ...   \n",
       "483     multiplexed                  NaN            NaN   \n",
       "484     multiplexed                  NaN            NaN   \n",
       "485          sb-ant                  NaN            NaN   \n",
       "486         sb-post                  NaN            NaN   \n",
       "487        sagittal                  NaN            NaN   \n",
       "\n",
       "    Comment[library construction] Comment[technical replicate group]  ...  \\\n",
       "0                             NaN                                NaN  ...   \n",
       "1                             NaN                                NaN  ...   \n",
       "2                             NaN                                NaN  ...   \n",
       "3                             NaN                                NaN  ...   \n",
       "4                             NaN                                NaN  ...   \n",
       "..                            ...                                ...  ...   \n",
       "483                           NaN                                NaN  ...   \n",
       "484                           NaN                                NaN  ...   \n",
       "485                           NaN                                NaN  ...   \n",
       "486                           NaN                                NaN  ...   \n",
       "487                           NaN                                NaN  ...   \n",
       "\n",
       "    Sampling_Site      Sex Sex_pf Sort_id_pf          Study    Technology  \\\n",
       "0             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "1             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "2             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "3             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "4             NaN      NaN    NaN        NaN  To_et_al_2024  10X_multiome   \n",
       "..            ...      ...    ...        ...            ...           ...   \n",
       "483           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "484           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "485           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "486           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "487           NaN  Unknown    NaN        NaN         Sanger  10X_multiome   \n",
       "\n",
       "    Termination_method_pf Unnamed: 14 abTCR_library_id_pf gdTCR_library_id_pf  \n",
       "0                     NaN         NaN                 NaN                 NaN  \n",
       "1                     NaN         NaN                 NaN                 NaN  \n",
       "2                     NaN         NaN                 NaN                 NaN  \n",
       "3                     NaN         NaN                 NaN                 NaN  \n",
       "4                     NaN         NaN                 NaN                 NaN  \n",
       "..                    ...         ...                 ...                 ...  \n",
       "483                   NaN         NaN                 NaN                 NaN  \n",
       "484                   NaN         NaN                 NaN                 NaN  \n",
       "485                   NaN         NaN                 NaN                 NaN  \n",
       "486                   NaN         NaN                 NaN                 NaN  \n",
       "487                   NaN         NaN                 NaN                 NaN  \n",
       "\n",
       "[488 rows x 45 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tidy Up Final Combined Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_METADATA.csv', index_col=0)\n",
    "cross_organ= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/CROSS_ORGAN_METADATA.csv')\n",
    "bone_metadata= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/BONE_ATLAS_METADATA.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample_pf</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "      <th>Organ.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HCA_BN_F13177434</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2356/2372</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HCA_BN_F13177435</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2356/2372</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HCA_BN_F13177436</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>10.3,11.2</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2356/2372</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HCA_BN_F12808058</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2370/2369</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HCA_BN_F12808059</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2370/2369</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HCA_BN_F12808060</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>5.8/7.6</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2370/2369</td>\n",
       "      <td>Total</td>\n",
       "      <td>skullbase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HCA_BN_F12627471</td>\n",
       "      <td>Hip</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>hip</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HCA_BN_F12922485</td>\n",
       "      <td>Hip/Knee</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>7.3/8.6</td>\n",
       "      <td>7.3/8.6</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2392/2394</td>\n",
       "      <td>Total</td>\n",
       "      <td>hip/knee</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hip/Knee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HCA_BN_F12627472</td>\n",
       "      <td>Hip/Shoulder</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>5.8, 8.3</td>\n",
       "      <td>5.8, 8.3</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>hip/shoulder</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hip/Shoulder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HCA_BN_F12627470</td>\n",
       "      <td>Knee</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>5.8, 7.6, 8.3</td>\n",
       "      <td>5.8, 7.6, 8.3</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>knee</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Knee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>HCA_BN_F12922487</td>\n",
       "      <td>Shoulder</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>7.3/8.6</td>\n",
       "      <td>7.3/8.6</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2392/2394</td>\n",
       "      <td>Total</td>\n",
       "      <td>shoulder</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shoulder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>HCA_BN_F12627473</td>\n",
       "      <td>Shoulder</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>shoulder</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shoulder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>HCA_BN_F13109714</td>\n",
       "      <td>Skull base</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull base</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>HCA_BN_F13109715</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>HCA_BN_F13109718</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>HCA_BN_F13109719</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>11.2,13</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2372/16403</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria/skullbase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull base/Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>HCA_BRA_F13168990</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>cerebral</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>HCA_BRA_F13168991</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>falcine</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>HCA_BRA_F13168992</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>skullbase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>HCA_BRA_F13168993</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>6.7,8.4</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>skullbase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>HCA_BRA_F13774551</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>HCA_BRA_F13774552</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>HCA_BRA_F13774553</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>HCA_BRA_F13774554</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>HCA_BRA_F13774555</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>HCA_BRA_F13774556</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample_ID                Organ 10X_Chemistry  Accession_ID  \\\n",
       "11    HCA_BN_F13177434            Calvaria   10x_multiome  E-MTAB-14385   \n",
       "12    HCA_BN_F13177435            Calvaria   10x_multiome  E-MTAB-14385   \n",
       "13    HCA_BN_F13177436            Calvaria   10x_multiome  E-MTAB-14385   \n",
       "14    HCA_BN_F12808058            Calvaria   10x_multiome  E-MTAB-14385   \n",
       "15    HCA_BN_F12808059            Calvaria   10x_multiome  E-MTAB-14385   \n",
       "16    HCA_BN_F12808060            Calvaria   10x_multiome  E-MTAB-14385   \n",
       "22    HCA_BN_F12627471                 Hip   10x_multiome  E-MTAB-14385   \n",
       "23    HCA_BN_F12922485             Hip/Knee  10x_multiome  E-MTAB-14385   \n",
       "24    HCA_BN_F12627472         Hip/Shoulder  10x_multiome  E-MTAB-14385   \n",
       "32    HCA_BN_F12627470                 Knee  10x_multiome  E-MTAB-14385   \n",
       "37    HCA_BN_F12922487            Shoulder   10x_multiome  E-MTAB-14385   \n",
       "38    HCA_BN_F12627473            Shoulder   10x_multiome  E-MTAB-14385   \n",
       "50    HCA_BN_F13109714          Skull base   10x_multiome  E-MTAB-14385   \n",
       "51    HCA_BN_F13109715  Skull base/Calvaria  10x_multiome  E-MTAB-14385   \n",
       "52    HCA_BN_F13109718  Skull base/Calvaria  10x_multiome  E-MTAB-14385   \n",
       "53    HCA_BN_F13109719  Skull base/Calvaria  10x_multiome  E-MTAB-14385   \n",
       "478  HCA_BRA_F13168990             Meninges  10x_multiome        Sanger   \n",
       "479  HCA_BRA_F13168991             Meninges  10x_multiome        Sanger   \n",
       "480  HCA_BRA_F13168992             Meninges  10x_multiome        Sanger   \n",
       "481  HCA_BRA_F13168993             Meninges  10x_multiome        Sanger   \n",
       "482  HCA_BRA_F13774551             Meninges  10x_multiome        Sanger   \n",
       "483  HCA_BRA_F13774552             Meninges  10x_multiome        Sanger   \n",
       "484  HCA_BRA_F13774553             Meninges  10x_multiome        Sanger   \n",
       "485  HCA_BRA_F13774554             Meninges  10x_multiome        Sanger   \n",
       "486  HCA_BRA_F13774555             Meninges  10x_multiome        Sanger   \n",
       "487  HCA_BRA_F13774556             Meninges  10x_multiome        Sanger   \n",
       "\n",
       "               Age       Age(PCW)      Dataset     Donor_ID FACS_Sorting  \\\n",
       "11       10.3,11.2      10.3,11.2   Bone_Atlas    2356/2372        Total   \n",
       "12       10.3,11.2      10.3,11.2   Bone_Atlas    2356/2372        Total   \n",
       "13       10.3,11.2      10.3,11.2   Bone_Atlas    2356/2372        Total   \n",
       "14         5.8/7.6        5.8/7.6   Bone_Atlas    2370/2369        Total   \n",
       "15         5.8/7.6        5.8/7.6   Bone_Atlas    2370/2369        Total   \n",
       "16         5.8/7.6        5.8/7.6   Bone_Atlas    2370/2369        Total   \n",
       "22        7.6, 8.3       7.6, 8.3   Bone_Atlas  multiplexed        Total   \n",
       "23         7.3/8.6        7.3/8.6   Bone_Atlas    2392/2394        Total   \n",
       "24        5.8, 8.3       5.8, 8.3   Bone_Atlas  multiplexed        Total   \n",
       "32   5.8, 7.6, 8.3  5.8, 7.6, 8.3   Bone_Atlas  multiplexed        Total   \n",
       "37         7.3/8.6        7.3/8.6   Bone_Atlas    2392/2394        Total   \n",
       "38        7.6, 8.3       7.6, 8.3   Bone_Atlas  multiplexed        Total   \n",
       "50         11.2,13        11.2,13   Bone_Atlas   2372/16403        Total   \n",
       "51         11.2,13        11.2,13   Bone_Atlas   2372/16403        Total   \n",
       "52         11.2,13        11.2,13   Bone_Atlas   2372/16403        Total   \n",
       "53         11.2,13        11.2,13   Bone_Atlas   2372/16403        Total   \n",
       "478        6.7,8.4        6.7,8.4  Cross_Organ  multiplexed       Total    \n",
       "479        6.7,8.4        6.7,8.4  Cross_Organ  multiplexed       Total    \n",
       "480        6.7,8.4        6.7,8.4  Cross_Organ  multiplexed       Total    \n",
       "481        6.7,8.4        6.7,8.4  Cross_Organ  multiplexed       Total    \n",
       "482   9.2,9.9,12.5   9.2,9.9,12.5  Cross_Organ  multiplexed       Total    \n",
       "483   9.2,9.9,12.5   9.2,9.9,12.5  Cross_Organ  multiplexed       Total    \n",
       "484   9.2,9.9,12.5   9.2,9.9,12.5  Cross_Organ  multiplexed       Total    \n",
       "485   9.2,9.9,12.5   9.2,9.9,12.5  Cross_Organ  multiplexed       Total    \n",
       "486   9.2,9.9,12.5   9.2,9.9,12.5  Cross_Organ  multiplexed       Total    \n",
       "487   9.2,9.9,12.5   9.2,9.9,12.5  Cross_Organ  multiplexed       Total    \n",
       "\n",
       "        Anatomical_Site  ... Sample_pf Sampling_Site      Sex Sex_pf  \\\n",
       "11             calvaria  ...       NaN           NaN      NaN    NaN   \n",
       "12             calvaria  ...       NaN           NaN      NaN    NaN   \n",
       "13             calvaria  ...       NaN           NaN      NaN    NaN   \n",
       "14             calvaria  ...       NaN           NaN      NaN    NaN   \n",
       "15             calvaria  ...       NaN           NaN      NaN    NaN   \n",
       "16            skullbase  ...       NaN           NaN      NaN    NaN   \n",
       "22                  hip  ...       NaN           NaN      NaN    NaN   \n",
       "23             hip/knee  ...       NaN           NaN      NaN    NaN   \n",
       "24         hip/shoulder  ...       NaN           NaN      NaN    NaN   \n",
       "32                 knee  ...       NaN           NaN      NaN    NaN   \n",
       "37             shoulder  ...       NaN           NaN      NaN    NaN   \n",
       "38             shoulder  ...       NaN           NaN      NaN    NaN   \n",
       "50   calvaria/skullbase  ...       NaN           NaN      NaN    NaN   \n",
       "51   calvaria/skullbase  ...       NaN           NaN      NaN    NaN   \n",
       "52   calvaria/skullbase  ...       NaN           NaN      NaN    NaN   \n",
       "53   calvaria/skullbase  ...       NaN           NaN      NaN    NaN   \n",
       "478            cerebral  ...       NaN           NaN  Unknown    NaN   \n",
       "479             falcine  ...       NaN           NaN  Unknown    NaN   \n",
       "480           skullbase  ...       NaN           NaN  Unknown    NaN   \n",
       "481           skullbase  ...       NaN           NaN  Unknown    NaN   \n",
       "482         multiplexed  ...       NaN           NaN  Unknown    NaN   \n",
       "483         multiplexed  ...       NaN           NaN  Unknown    NaN   \n",
       "484         multiplexed  ...       NaN           NaN  Unknown    NaN   \n",
       "485              sb-ant  ...       NaN           NaN  Unknown    NaN   \n",
       "486             sb-post  ...       NaN           NaN  Unknown    NaN   \n",
       "487            sagittal  ...       NaN           NaN  Unknown    NaN   \n",
       "\n",
       "    Sort_id_pf Termination_method_pf Unnamed: 14  abTCR_library_id_pf  \\\n",
       "11         NaN                   NaN         NaN                  NaN   \n",
       "12         NaN                   NaN         NaN                  NaN   \n",
       "13         NaN                   NaN         NaN                  NaN   \n",
       "14         NaN                   NaN         NaN                  NaN   \n",
       "15         NaN                   NaN         NaN                  NaN   \n",
       "16         NaN                   NaN         NaN                  NaN   \n",
       "22         NaN                   NaN         NaN                  NaN   \n",
       "23         NaN                   NaN         NaN                  NaN   \n",
       "24         NaN                   NaN         NaN                  NaN   \n",
       "32         NaN                   NaN         NaN                  NaN   \n",
       "37         NaN                   NaN         NaN                  NaN   \n",
       "38         NaN                   NaN         NaN                  NaN   \n",
       "50         NaN                   NaN         NaN                  NaN   \n",
       "51         NaN                   NaN         NaN                  NaN   \n",
       "52         NaN                   NaN         NaN                  NaN   \n",
       "53         NaN                   NaN         NaN                  NaN   \n",
       "478        NaN                   NaN         NaN                  NaN   \n",
       "479        NaN                   NaN         NaN                  NaN   \n",
       "480        NaN                   NaN         NaN                  NaN   \n",
       "481        NaN                   NaN         NaN                  NaN   \n",
       "482        NaN                   NaN         NaN                  NaN   \n",
       "483        NaN                   NaN         NaN                  NaN   \n",
       "484        NaN                   NaN         NaN                  NaN   \n",
       "485        NaN                   NaN         NaN                  NaN   \n",
       "486        NaN                   NaN         NaN                  NaN   \n",
       "487        NaN                   NaN         NaN                  NaN   \n",
       "\n",
       "    gdTCR_library_id_pf              Organ.1  \n",
       "11                  NaN            Calvaria   \n",
       "12                  NaN            Calvaria   \n",
       "13                  NaN            Calvaria   \n",
       "14                  NaN            Calvaria   \n",
       "15                  NaN            Calvaria   \n",
       "16                  NaN            Calvaria   \n",
       "22                  NaN                 Hip   \n",
       "23                  NaN             Hip/Knee  \n",
       "24                  NaN         Hip/Shoulder  \n",
       "32                  NaN                 Knee  \n",
       "37                  NaN            Shoulder   \n",
       "38                  NaN            Shoulder   \n",
       "50                  NaN          Skull base   \n",
       "51                  NaN  Skull base/Calvaria  \n",
       "52                  NaN  Skull base/Calvaria  \n",
       "53                  NaN  Skull base/Calvaria  \n",
       "478                 NaN             Meninges  \n",
       "479                 NaN             Meninges  \n",
       "480                 NaN             Meninges  \n",
       "481                 NaN             Meninges  \n",
       "482                 NaN             Meninges  \n",
       "483                 NaN             Meninges  \n",
       "484                 NaN             Meninges  \n",
       "485                 NaN             Meninges  \n",
       "486                 NaN             Meninges  \n",
       "487                 NaN             Meninges  \n",
       "\n",
       "[26 rows x 45 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[metadata[\"Multiplexed\"]==\"Yes\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "don't need these which have been mapped \n",
    "\n",
    "['6180STDY9448808',\n",
    " '6180STDY9448811',\n",
    " '6180STDY9448813',\n",
    " 'WSSS_F_Imm11192515',\n",
    " 'WSSS_F_Imm11192516',\n",
    " 'WSSS_F_Imm11192517',\n",
    " 'WSSS_F_Imm11192518']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (218506739.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[53], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    488  samples\u001b[0m\n\u001b[0m         ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "488  samples\n",
    "26 samples are multiplexed \n",
    "66 samples are multiome \n",
    "\n",
    "# Need to rename some samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "482"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Define the SRR → GSM mapping\n",
    "mapping = {\n",
    "    \"SRR10901267\": \"GSM4274188\",\n",
    "    \"SRR10901268\": \"GSM4274189\",\n",
    "    \"SRR10901269\": \"GSM4274190\",\n",
    "    \"SRR10901270\": \"GSM4274191\",\n",
    "    # all runs for GSM4274192\n",
    "    **{run: \"GSM4274192\" for run in [\"SRR10901271\", \"SRR10901272\", \"SRR10901273\", \"SRR10901274\"]},\n",
    "    \"SRR10901275\": \"GSM4274193\",\n",
    "    \"SRR10901276\": \"GSM4274194\",\n",
    "    # all runs for GSM4609136\n",
    "    **{run: \"GSM4609136\" for run in [\"SRR11977450\", \"SRR11977451\", \"SRR11977452\", \"SRR11977453\"]},\n",
    "}\n",
    "\n",
    "# 3. Apply mapping: replace any Sample_ID matching an SRR with its GSM\n",
    "metadata[\"Sample_ID\"] = metadata[\"Sample_ID\"].map(mapping).fillna(metadata[\"Sample_ID\"])\n",
    "bone_metadata[\"Sample_ID\"] = bone_metadata[\"Sample_ID\"].map(mapping).fillna(bone_metadata[\"Sample_ID\"])\n",
    "\n",
    "# 4. (Optional) Check the result\n",
    "len((metadata[\"Sample_ID\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# m\n",
    "\n",
    "total_samples= list(metadata[\"Sample_ID\"]).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/cellgeni/tickets/tic-3721/Cellbender/cellbender-results\n"
     ]
    }
   ],
   "source": [
    "cd /nfs/cellgeni/tickets/tic-3721/Cellbender/cellbender-results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Define the path to the directory containing the subfolders\n",
    "dir_path = \"/nfs/cellgeni/tickets/tic-3721/Cellbender/cellbender-results\"\n",
    "\n",
    "# List all entries in the directory and filter to include only directories\n",
    "dir_names = [name for name in os.listdir(dir_path) \n",
    "             if os.path.isdir(os.path.join(dir_path, name))]\n",
    "\n",
    "# Get the set of Sample_ID values from the metadata DataFrame\n",
    "sample_ids = set(metadata[\"Sample_ID\"])\n",
    "\n",
    "# Find the intersection of directory names and Sample_ID values\n",
    "matches = sorted(set(dir_names) & sample_ids)  # using set intersection\n",
    "non_matches = sorted(set(dir_names) - sample_ids)  # using set intersection\n",
    "non_matches2 = sorted(set(sample_ids) - set(dir_names))  # using set intersection\n",
    "\n",
    "num_matches = len(matches)\n",
    "\n",
    "# # Print the results in a clean format\n",
    "# print(f\"Number of matching directories: {num_matches}\")\n",
    "# if num_matches > 0:\n",
    "#     print(\"Matching directory names:\")\n",
    "#     for m in matches:\n",
    "#         print(\"-\", m)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "482"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sample_ids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "481"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['F83_TH_EPCAM']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_matches2 # this is a thymus organpoid that's missing and I don't really need it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['6180STDY9448808',\n",
       " '6180STDY9448811',\n",
       " '6180STDY9448813',\n",
       " 'WSSS8062676',\n",
       " 'WSSS_F_Imm11192515',\n",
       " 'WSSS_F_Imm11192516',\n",
       " 'WSSS_F_Imm11192517',\n",
       " 'WSSS_F_Imm11192518',\n",
       " 'qc_report']"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_matches # I am not using these files from the cellbencer outputs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hence all samples I need are present "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata[\"10X_Chemistry\"] = metadata[\"10X_Chemistry\"].str.replace(\" \", \"\", regex=False)\n",
    "metadata[\"10X_Chemistry\"] = metadata[\"10X_Chemistry\"].str.replace(\"''\", \"'\", regex=False)\n",
    "bone_metadata[\"10X_Chemistry\"] = bone_metadata[\"10X_Chemistry\"].str.replace(\" \", \"\", regex=False)\n",
    "bone_metadata[\"10X_Chemistry\"] = bone_metadata[\"10X_Chemistry\"].str.replace(\"''\", \"'\", regex=False)\n",
    "cross_organ[\"10X_Chemistry\"] = cross_organ[\"10X_Chemistry\"].str.replace(\" \", \"\", regex=False)\n",
    "cross_organ[\"10X_Chemistry\"] = cross_organ[\"10X_Chemistry\"].str.replace(\"''\", \"'\", regex=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample_pf</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "      <th>Organ.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>FCAImmP7579230</td>\n",
       "      <td>thymus</td>\n",
       "      <td>10x5'v2</td>\n",
       "      <td>E-MTAB-8581</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>F45</td>\n",
       "      <td>CD45-positive</td>\n",
       "      <td>thymus</td>\n",
       "      <td>...</td>\n",
       "      <td>F45_TH_CD45P_FCAImmP7579230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>CD45P</td>\n",
       "      <td>Medical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FCAImmP7607608</td>\n",
       "      <td>WSSS_F_Imm10424046</td>\n",
       "      <td>thymus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample_ID   Organ 10X_Chemistry Accession_ID Age Age(PCW)  \\\n",
       "419  FCAImmP7579230  thymus       10x5'v2  E-MTAB-8581  12       12   \n",
       "\n",
       "         Dataset Donor_ID   FACS_Sorting Anatomical_Site  ...  \\\n",
       "419  Cross_Organ      F45  CD45-positive          thymus  ...   \n",
       "\n",
       "                       Sample_pf Sampling_Site     Sex  Sex_pf Sort_id_pf  \\\n",
       "419  F45_TH_CD45P_FCAImmP7579230           NaN  female  female      CD45P   \n",
       "\n",
       "    Termination_method_pf Unnamed: 14  abTCR_library_id_pf  \\\n",
       "419               Medical         NaN       FCAImmP7607608   \n",
       "\n",
       "    gdTCR_library_id_pf Organ.1  \n",
       "419  WSSS_F_Imm10424046  thymus  \n",
       "\n",
       "[1 rows x 45 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[metadata[\"Method\"]==\"10xV2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "209"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(bone_metadata[\"Method\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata[\"Method\"] = metadata[\"Method\"].str.replace(\"10xV2\", \"5GEX\", regex=False)\n",
    "cross_organ[\"Method\"] = cross_organ[\"Method\"].str.replace(\"10xV2\", \"5GEX\", regex=False)\n",
    "bone_metadata.columns = bone_metadata.columns.str.strip()\n",
    "metadata[\"Method\"] = metadata[\"Method\"].str.replace(\"10x_multiome\", \"10X_multiome\", regex=False)\n",
    "bone_metadata[\"Method\"] = bone_metadata[\"Method\"].str.replace(\"10x_multiome\", \"10X_multiome\", regex=False)\n",
    "cross_organ[\"Method\"] = cross_organ[\"Method\"].str.replace(\"10x_multiome\", \"10X_multiome\", regex=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Method\n",
       "5GEX            142\n",
       "3GEX            125\n",
       "10X_multiome     12\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cross_organ[\"Method\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Technology\n",
       "10X_scRNA-seq    422\n",
       "10X_multiome      66\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[\"Technology\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_METADATA.csv')\n",
    "cross_organ.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/CROSS_ORGAN_METADATA.csv')\n",
    "bone_metadata.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/BONE_ATLAS_METADATA.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "metadata= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_METADATA.csv', index_col=0)\n",
    "cross_organ= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/CROSS_ORGAN_METADATA.csv',index_col=0)\n",
    "bone_metadata= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/BONE_ATLAS_METADATA.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample_pf</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "      <th>Organ.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143757</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Total</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long bone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143758</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Total</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long bone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143759</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Total</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long bone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143760</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Total</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long bone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143763</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Total</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143764</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Total</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143765</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Total</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>GSE143766</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Total</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Sample_ID      Organ 10X_Chemistry Accession_ID Age Age(PCW)  \\\n",
       "120  GSM4274192  Long bone       10x3'v2    GSE143757   8        8   \n",
       "121  GSM4274192  Long bone       10x3'v2    GSE143758   8        8   \n",
       "122  GSM4274192  Long bone       10x3'v2    GSE143759   8        8   \n",
       "123  GSM4274192  Long bone       10x3'v2    GSE143760   8        8   \n",
       "126  GSM4609136  Calvaria        10x3'v2    GSE143763   8        8   \n",
       "127  GSM4609136  Calvaria        10x3'v2    GSE143764   8        8   \n",
       "128  GSM4609136  Calvaria        10x3'v2    GSE143765   8        8   \n",
       "129  GSM4609136  Calvaria        10x3'v2    GSE143766   8        8   \n",
       "\n",
       "        Dataset    Donor_ID FACS_Sorting Anatomical_Site  ... Sample_pf  \\\n",
       "120  Bone_Atlas  GSM4274192        Total       Long bone  ...       NaN   \n",
       "121  Bone_Atlas  GSM4274192        Total       Long bone  ...       NaN   \n",
       "122  Bone_Atlas  GSM4274192        Total       Long bone  ...       NaN   \n",
       "123  Bone_Atlas  GSM4274192        Total       Long bone  ...       NaN   \n",
       "126  Bone_Atlas  GSM4609136        Total  Calvarial bone  ...       NaN   \n",
       "127  Bone_Atlas  GSM4609136        Total  Calvarial bone  ...       NaN   \n",
       "128  Bone_Atlas  GSM4609136        Total  Calvarial bone  ...       NaN   \n",
       "129  Bone_Atlas  GSM4609136        Total  Calvarial bone  ...       NaN   \n",
       "\n",
       "    Sampling_Site  Sex Sex_pf Sort_id_pf Termination_method_pf Unnamed: 14  \\\n",
       "120           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "121           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "122           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "123           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "126           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "127           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "128           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "129           NaN  NaN    NaN        NaN                   NaN         NaN   \n",
       "\n",
       "     abTCR_library_id_pf gdTCR_library_id_pf    Organ.1  \n",
       "120                  NaN                 NaN  Long bone  \n",
       "121                  NaN                 NaN  Long bone  \n",
       "122                  NaN                 NaN  Long bone  \n",
       "123                  NaN                 NaN  Long bone  \n",
       "126                  NaN                 NaN  Calvaria   \n",
       "127                  NaN                 NaN  Calvaria   \n",
       "128                  NaN                 NaN  Calvaria   \n",
       "129                  NaN                 NaN  Calvaria   \n",
       "\n",
       "[8 rows x 45 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mark all rows where Sample_ID is a duplicate\n",
    "dup_mask = metadata['Sample_ID'].duplicated(keep=False)\n",
    "\n",
    "# select just those rows\n",
    "duplicate_rows = metadata[dup_mask]\n",
    "\n",
    "# inspect\n",
    "(duplicate_rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Run_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ.1</th>\n",
       "      <th>Age</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Technology</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Multiplexed</th>\n",
       "      <th>Expected_Donors</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>Method</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>120</td>\n",
       "      <td>120</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>SRR10901271</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>8</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143757</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>121</td>\n",
       "      <td>121</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>SRR10901272</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>8</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143758</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>122</td>\n",
       "      <td>122</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>SRR10901273</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>8</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143759</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>SRR10901274</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>8</td>\n",
       "      <td>Long bone</td>\n",
       "      <td>GSM4274192</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143760</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>SRR11977450</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143763</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>SRR11977451</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143764</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>128</td>\n",
       "      <td>128</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>SRR11977452</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143765</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>129</td>\n",
       "      <td>129</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>SRR11977453</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8</td>\n",
       "      <td>Calvarial bone</td>\n",
       "      <td>GSM4609136</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>GSE143766</td>\n",
       "      <td>He_et_al_2021</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0.1  Unnamed: 0      Organ       Run_ID   Sample_ID    Organ.1  \\\n",
       "120           120         120  Long bone  SRR10901271  GSM4274192  Long bone   \n",
       "121           121         121  Long bone  SRR10901272  GSM4274192  Long bone   \n",
       "122           122         122  Long bone  SRR10901273  GSM4274192  Long bone   \n",
       "123           123         123  Long bone  SRR10901274  GSM4274192  Long bone   \n",
       "126           126         126  Calvaria   SRR11977450  GSM4609136  Calvaria    \n",
       "127           127         127  Calvaria   SRR11977451  GSM4609136  Calvaria    \n",
       "128           128         128  Calvaria   SRR11977452  GSM4609136  Calvaria    \n",
       "129           129         129  Calvaria   SRR11977453  GSM4609136  Calvaria    \n",
       "\n",
       "    Age Anatomical_Site    Donor_ID     Technology FACS_Sorting Accession_ID  \\\n",
       "120   8       Long bone  GSM4274192  10X_scRNA-seq        Total    GSE143757   \n",
       "121   8       Long bone  GSM4274192  10X_scRNA-seq        Total    GSE143758   \n",
       "122   8       Long bone  GSM4274192  10X_scRNA-seq        Total    GSE143759   \n",
       "123   8       Long bone  GSM4274192  10X_scRNA-seq        Total    GSE143760   \n",
       "126   8  Calvarial bone  GSM4609136  10X_scRNA-seq        Total    GSE143763   \n",
       "127   8  Calvarial bone  GSM4609136  10X_scRNA-seq        Total    GSE143764   \n",
       "128   8  Calvarial bone  GSM4609136  10X_scRNA-seq        Total    GSE143765   \n",
       "129   8  Calvarial bone  GSM4609136  10X_scRNA-seq        Total    GSE143766   \n",
       "\n",
       "             Study Multiplexed  Expected_Donors 10X_Chemistry  Unnamed: 14  \\\n",
       "120  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "121  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "122  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "123  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "126  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "127  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "128  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "129  He_et_al_2021          No                1       10x3'v2          NaN   \n",
       "\n",
       "    Method     Dataset  \n",
       "120   3GEX  Bone_Atlas  \n",
       "121   3GEX  Bone_Atlas  \n",
       "122   3GEX  Bone_Atlas  \n",
       "123   3GEX  Bone_Atlas  \n",
       "126   3GEX  Bone_Atlas  \n",
       "127   3GEX  Bone_Atlas  \n",
       "128   3GEX  Bone_Atlas  \n",
       "129   3GEX  Bone_Atlas  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dup_mask = bone_metadata['Sample_ID'].duplicated(keep=False)\n",
    "\n",
    "# select just those rows\n",
    "duplicate_rows = bone_metadata[dup_mask]\n",
    "\n",
    "# inspect\n",
    "(duplicate_rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = metadata[~metadata['Sample_ID'].duplicated(keep='first')]\n",
    "bone_metadata = bone_metadata[~bone_metadata['Sample_ID'].duplicated(keep='first')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample_pf</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "      <th>Organ.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCA_BN_F12482363</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCA_BN_F12482364</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCA_BN_F12482365</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCA_BN_F12482366</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCA_BN_F12966455</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>9.4</td>\n",
       "      <td>9.4</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2384</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>HCA_BRA_F13774552</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>HCA_BRA_F13774553</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>HCA_BRA_F13774554</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>HCA_BRA_F13774555</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>HCA_BRA_F13774556</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample_ID      Organ 10X_Chemistry  Accession_ID           Age  \\\n",
       "0     HCA_BN_F12482363  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "1     HCA_BN_F12482364  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "2     HCA_BN_F12482365  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "3     HCA_BN_F12482366  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "4     HCA_BN_F12966455  Calvaria   10x_multiome  E-MTAB-14385           9.4   \n",
       "..                 ...        ...           ...           ...           ...   \n",
       "483  HCA_BRA_F13774552   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "484  HCA_BRA_F13774553   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "485  HCA_BRA_F13774554   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "486  HCA_BRA_F13774555   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "487  HCA_BRA_F13774556   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "\n",
       "         Age(PCW)      Dataset     Donor_ID FACS_Sorting Anatomical_Site  ...  \\\n",
       "0             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "1             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "2             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "3             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "4             9.4   Bone_Atlas         2384        Total        calvaria  ...   \n",
       "..            ...          ...          ...          ...             ...  ...   \n",
       "483  9.2,9.9,12.5  Cross_Organ  multiplexed       Total      multiplexed  ...   \n",
       "484  9.2,9.9,12.5  Cross_Organ  multiplexed       Total      multiplexed  ...   \n",
       "485  9.2,9.9,12.5  Cross_Organ  multiplexed       Total           sb-ant  ...   \n",
       "486  9.2,9.9,12.5  Cross_Organ  multiplexed       Total          sb-post  ...   \n",
       "487  9.2,9.9,12.5  Cross_Organ  multiplexed       Total         sagittal  ...   \n",
       "\n",
       "    Sample_pf Sampling_Site      Sex Sex_pf Sort_id_pf Termination_method_pf  \\\n",
       "0         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "1         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "2         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "3         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "4         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "..        ...           ...      ...    ...        ...                   ...   \n",
       "483       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "484       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "485       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "486       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "487       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "\n",
       "    Unnamed: 14  abTCR_library_id_pf gdTCR_library_id_pf    Organ.1  \n",
       "0           NaN                  NaN                 NaN  Calvaria   \n",
       "1           NaN                  NaN                 NaN  Calvaria   \n",
       "2           NaN                  NaN                 NaN  Calvaria   \n",
       "3           NaN                  NaN                 NaN  Calvaria   \n",
       "4           NaN                  NaN                 NaN  Calvaria   \n",
       "..          ...                  ...                 ...        ...  \n",
       "483         NaN                  NaN                 NaN   Meninges  \n",
       "484         NaN                  NaN                 NaN   Meninges  \n",
       "485         NaN                  NaN                 NaN   Meninges  \n",
       "486         NaN                  NaN                 NaN   Meninges  \n",
       "487         NaN                  NaN                 NaN   Meninges  \n",
       "\n",
       "[482 rows x 45 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Run_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ.1</th>\n",
       "      <th>Age</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Technology</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Multiplexed</th>\n",
       "      <th>Expected_Donors</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>Method</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482363_and_HCA_BN_F12477275</td>\n",
       "      <td>HCA_BN_F12482363</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482364_and_HCA_BN_F12477276</td>\n",
       "      <td>HCA_BN_F12482364</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482365_and_HCA_BN_F12477277</td>\n",
       "      <td>HCA_BN_F12482365</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482366_and_HCA_BN_F12477278</td>\n",
       "      <td>HCA_BN_F12482366</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12966455_and_HCA_BN_F12951831</td>\n",
       "      <td>HCA_BN_F12966455</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>9.4</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2384</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>5478STDY7850145</td>\n",
       "      <td>5478STDY7850145</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>7</td>\n",
       "      <td>left fibuka</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>Femur</td>\n",
       "      <td>5478STDY7850146</td>\n",
       "      <td>5478STDY7850146</td>\n",
       "      <td>Femur</td>\n",
       "      <td>7</td>\n",
       "      <td>right proximal femur</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>Femur</td>\n",
       "      <td>5478STDY7850147</td>\n",
       "      <td>5478STDY7850147</td>\n",
       "      <td>Femur</td>\n",
       "      <td>7</td>\n",
       "      <td>right distal femur</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>207</td>\n",
       "      <td>207</td>\n",
       "      <td>Tibia</td>\n",
       "      <td>5478STDY7850148</td>\n",
       "      <td>5478STDY7850148</td>\n",
       "      <td>Tibia</td>\n",
       "      <td>7</td>\n",
       "      <td>right tibia</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>5478STDY7850149</td>\n",
       "      <td>5478STDY7850149</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>7</td>\n",
       "      <td>right fibula</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>203 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0.1  Unnamed: 0      Organ  \\\n",
       "0               0           0  Calvaria    \n",
       "1               1           1  Calvaria    \n",
       "2               2           2  Calvaria    \n",
       "3               3           3  Calvaria    \n",
       "4               4           4  Calvaria    \n",
       "..            ...         ...        ...   \n",
       "204           204         204     Fibula   \n",
       "205           205         205     Femur    \n",
       "206           206         206     Femur    \n",
       "207           207         207     Tibia    \n",
       "208           208         208     Fibula   \n",
       "\n",
       "                                    Run_ID         Sample_ID    Organ.1  Age  \\\n",
       "0    HCA_BN_F12482363_and_HCA_BN_F12477275  HCA_BN_F12482363  Calvaria   8.5   \n",
       "1    HCA_BN_F12482364_and_HCA_BN_F12477276  HCA_BN_F12482364  Calvaria   8.5   \n",
       "2    HCA_BN_F12482365_and_HCA_BN_F12477277  HCA_BN_F12482365  Calvaria   8.5   \n",
       "3    HCA_BN_F12482366_and_HCA_BN_F12477278  HCA_BN_F12482366  Calvaria   8.5   \n",
       "4    HCA_BN_F12966455_and_HCA_BN_F12951831  HCA_BN_F12966455  Calvaria   9.4   \n",
       "..                                     ...               ...        ...  ...   \n",
       "204                        5478STDY7850145   5478STDY7850145     Fibula    7   \n",
       "205                        5478STDY7850146   5478STDY7850146     Femur     7   \n",
       "206                        5478STDY7850147   5478STDY7850147     Femur     7   \n",
       "207                        5478STDY7850148   5478STDY7850148     Tibia     7   \n",
       "208                        5478STDY7850149   5478STDY7850149     Fibula    7   \n",
       "\n",
       "          Anatomical_Site Donor_ID     Technology FACS_Sorting  \\\n",
       "0                calvaria     2361   10X_multiome        Total   \n",
       "1                calvaria     2361   10X_multiome        Total   \n",
       "2                calvaria     2361   10X_multiome        Total   \n",
       "3                calvaria     2361   10X_multiome        Total   \n",
       "4                calvaria     2384   10X_multiome        Total   \n",
       "..                    ...      ...            ...          ...   \n",
       "204           left fibuka      C32  10X_scRNA-seq        Total   \n",
       "205  right proximal femur      C32  10X_scRNA-seq        Total   \n",
       "206    right distal femur      C32  10X_scRNA-seq        Total   \n",
       "207           right tibia      C32  10X_scRNA-seq        Total   \n",
       "208          right fibula      C32  10X_scRNA-seq        Total   \n",
       "\n",
       "        Accession_ID                Study Multiplexed  Expected_Donors  \\\n",
       "0       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "1       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "2       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "3       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "4       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "..               ...                  ...         ...              ...   \n",
       "204  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "205  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "206  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "207  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "208  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "\n",
       "    10X_Chemistry  Unnamed: 14        Method     Dataset  \n",
       "0    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "1    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "2    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "3    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "4    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "..            ...          ...           ...         ...  \n",
       "204       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "205       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "206       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "207       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "208       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "\n",
       "[203 rows x 19 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bone_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_METADATA.csv')\n",
    "cross_organ.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/CROSS_ORGAN_METADATA.csv')\n",
    "bone_metadata.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/BONE_ATLAS_METADATA.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample_pf</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "      <th>Organ.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCA_BN_F12482363</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCA_BN_F12482364</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCA_BN_F12482365</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCA_BN_F12482366</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCA_BN_F12966455</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>9.4</td>\n",
       "      <td>9.4</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2384</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>HCA_BRA_F13774552</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>HCA_BRA_F13774553</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>HCA_BRA_F13774554</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>HCA_BRA_F13774555</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>HCA_BRA_F13774556</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample_ID      Organ 10X_Chemistry  Accession_ID           Age  \\\n",
       "0     HCA_BN_F12482363  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "1     HCA_BN_F12482364  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "2     HCA_BN_F12482365  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "3     HCA_BN_F12482366  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "4     HCA_BN_F12966455  Calvaria   10x_multiome  E-MTAB-14385           9.4   \n",
       "..                 ...        ...           ...           ...           ...   \n",
       "483  HCA_BRA_F13774552   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "484  HCA_BRA_F13774553   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "485  HCA_BRA_F13774554   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "486  HCA_BRA_F13774555   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "487  HCA_BRA_F13774556   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "\n",
       "         Age(PCW)      Dataset     Donor_ID FACS_Sorting Anatomical_Site  ...  \\\n",
       "0             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "1             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "2             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "3             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "4             9.4   Bone_Atlas         2384        Total        calvaria  ...   \n",
       "..            ...          ...          ...          ...             ...  ...   \n",
       "483  9.2,9.9,12.5  Cross_Organ  multiplexed       Total      multiplexed  ...   \n",
       "484  9.2,9.9,12.5  Cross_Organ  multiplexed       Total      multiplexed  ...   \n",
       "485  9.2,9.9,12.5  Cross_Organ  multiplexed       Total           sb-ant  ...   \n",
       "486  9.2,9.9,12.5  Cross_Organ  multiplexed       Total          sb-post  ...   \n",
       "487  9.2,9.9,12.5  Cross_Organ  multiplexed       Total         sagittal  ...   \n",
       "\n",
       "    Sample_pf Sampling_Site      Sex Sex_pf Sort_id_pf Termination_method_pf  \\\n",
       "0         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "1         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "2         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "3         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "4         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "..        ...           ...      ...    ...        ...                   ...   \n",
       "483       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "484       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "485       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "486       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "487       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "\n",
       "    Unnamed: 14  abTCR_library_id_pf gdTCR_library_id_pf    Organ.1  \n",
       "0           NaN                  NaN                 NaN  Calvaria   \n",
       "1           NaN                  NaN                 NaN  Calvaria   \n",
       "2           NaN                  NaN                 NaN  Calvaria   \n",
       "3           NaN                  NaN                 NaN  Calvaria   \n",
       "4           NaN                  NaN                 NaN  Calvaria   \n",
       "..          ...                  ...                 ...        ...  \n",
       "483         NaN                  NaN                 NaN   Meninges  \n",
       "484         NaN                  NaN                 NaN   Meninges  \n",
       "485         NaN                  NaN                 NaN   Meninges  \n",
       "486         NaN                  NaN                 NaN   Meninges  \n",
       "487         NaN                  NaN                 NaN   Meninges  \n",
       "\n",
       "[482 rows x 45 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Run_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ.1</th>\n",
       "      <th>Age</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Technology</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Multiplexed</th>\n",
       "      <th>Expected_Donors</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>Method</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482363_and_HCA_BN_F12477275</td>\n",
       "      <td>HCA_BN_F12482363</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482364_and_HCA_BN_F12477276</td>\n",
       "      <td>HCA_BN_F12482364</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482365_and_HCA_BN_F12477277</td>\n",
       "      <td>HCA_BN_F12482365</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12482366_and_HCA_BN_F12477278</td>\n",
       "      <td>HCA_BN_F12482366</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>8.5</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2361</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>HCA_BN_F12966455_and_HCA_BN_F12951831</td>\n",
       "      <td>HCA_BN_F12966455</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>9.4</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>2384</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Total</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>To_et_al_2024</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>5478STDY7850145</td>\n",
       "      <td>5478STDY7850145</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>7</td>\n",
       "      <td>left fibuka</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>Femur</td>\n",
       "      <td>5478STDY7850146</td>\n",
       "      <td>5478STDY7850146</td>\n",
       "      <td>Femur</td>\n",
       "      <td>7</td>\n",
       "      <td>right proximal femur</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>Femur</td>\n",
       "      <td>5478STDY7850147</td>\n",
       "      <td>5478STDY7850147</td>\n",
       "      <td>Femur</td>\n",
       "      <td>7</td>\n",
       "      <td>right distal femur</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>207</td>\n",
       "      <td>207</td>\n",
       "      <td>Tibia</td>\n",
       "      <td>5478STDY7850148</td>\n",
       "      <td>5478STDY7850148</td>\n",
       "      <td>Tibia</td>\n",
       "      <td>7</td>\n",
       "      <td>right tibia</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>5478STDY7850149</td>\n",
       "      <td>5478STDY7850149</td>\n",
       "      <td>Fibula</td>\n",
       "      <td>7</td>\n",
       "      <td>right fibula</td>\n",
       "      <td>C32</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Total</td>\n",
       "      <td>EGAD00001011306</td>\n",
       "      <td>Lawrence_et_al_2023</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>203 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0.1  Unnamed: 0      Organ  \\\n",
       "0               0           0  Calvaria    \n",
       "1               1           1  Calvaria    \n",
       "2               2           2  Calvaria    \n",
       "3               3           3  Calvaria    \n",
       "4               4           4  Calvaria    \n",
       "..            ...         ...        ...   \n",
       "204           204         204     Fibula   \n",
       "205           205         205     Femur    \n",
       "206           206         206     Femur    \n",
       "207           207         207     Tibia    \n",
       "208           208         208     Fibula   \n",
       "\n",
       "                                    Run_ID         Sample_ID    Organ.1  Age  \\\n",
       "0    HCA_BN_F12482363_and_HCA_BN_F12477275  HCA_BN_F12482363  Calvaria   8.5   \n",
       "1    HCA_BN_F12482364_and_HCA_BN_F12477276  HCA_BN_F12482364  Calvaria   8.5   \n",
       "2    HCA_BN_F12482365_and_HCA_BN_F12477277  HCA_BN_F12482365  Calvaria   8.5   \n",
       "3    HCA_BN_F12482366_and_HCA_BN_F12477278  HCA_BN_F12482366  Calvaria   8.5   \n",
       "4    HCA_BN_F12966455_and_HCA_BN_F12951831  HCA_BN_F12966455  Calvaria   9.4   \n",
       "..                                     ...               ...        ...  ...   \n",
       "204                        5478STDY7850145   5478STDY7850145     Fibula    7   \n",
       "205                        5478STDY7850146   5478STDY7850146     Femur     7   \n",
       "206                        5478STDY7850147   5478STDY7850147     Femur     7   \n",
       "207                        5478STDY7850148   5478STDY7850148     Tibia     7   \n",
       "208                        5478STDY7850149   5478STDY7850149     Fibula    7   \n",
       "\n",
       "          Anatomical_Site Donor_ID     Technology FACS_Sorting  \\\n",
       "0                calvaria     2361   10X_multiome        Total   \n",
       "1                calvaria     2361   10X_multiome        Total   \n",
       "2                calvaria     2361   10X_multiome        Total   \n",
       "3                calvaria     2361   10X_multiome        Total   \n",
       "4                calvaria     2384   10X_multiome        Total   \n",
       "..                    ...      ...            ...          ...   \n",
       "204           left fibuka      C32  10X_scRNA-seq        Total   \n",
       "205  right proximal femur      C32  10X_scRNA-seq        Total   \n",
       "206    right distal femur      C32  10X_scRNA-seq        Total   \n",
       "207           right tibia      C32  10X_scRNA-seq        Total   \n",
       "208          right fibula      C32  10X_scRNA-seq        Total   \n",
       "\n",
       "        Accession_ID                Study Multiplexed  Expected_Donors  \\\n",
       "0       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "1       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "2       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "3       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "4       E-MTAB-14385        To_et_al_2024          No                1   \n",
       "..               ...                  ...         ...              ...   \n",
       "204  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "205  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "206  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "207  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "208  EGAD00001011306  Lawrence_et_al_2023          No                1   \n",
       "\n",
       "    10X_Chemistry  Unnamed: 14        Method     Dataset  \n",
       "0    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "1    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "2    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "3    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "4    10x_multiome          NaN  10X_multiome  Bone_Atlas  \n",
       "..            ...          ...           ...         ...  \n",
       "204       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "205       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "206       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "207       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "208       10x3'v2          NaN          3GEX  Bone_Atlas  \n",
       "\n",
       "[203 rows x 19 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bone_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Linnarson Data metatadata clean up "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "EGAD00001006049= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_EGAD00001006049.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_METADATA.csv', index_col=0)\n",
    "lm= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_Metadata_from_Paper.csv') # Linnarson data metadata from the paper supp table 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "EGAD50000001295= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_EGAD50000001295.csv') # Linnarson data metadata from EGA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>description</th>\n",
       "      <th>biological_sex</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>extra_attributes</th>\n",
       "      <th>submission_accession_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>edited_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGAN50000223063</td>\n",
       "      <td>10X188_4_ABC_1</td>\n",
       "      <td>10X188_4</td>\n",
       "      <td>XDD:351_Subcortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGAN50000223064</td>\n",
       "      <td>10X302_1_AB_1</td>\n",
       "      <td>10X302_1</td>\n",
       "      <td>XDD:398_Hindbrain_7w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:398</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGAN50000223065</td>\n",
       "      <td>10X252_4_AB_1</td>\n",
       "      <td>10X252_4</td>\n",
       "      <td>XDD:385_Hippocampus_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGAN50000223066</td>\n",
       "      <td>10X186_1_ABC_1</td>\n",
       "      <td>10X186_1</td>\n",
       "      <td>XDD:351_Cerebellum_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGAN50000223067</td>\n",
       "      <td>10X287_7_AB_1</td>\n",
       "      <td>10X287_7</td>\n",
       "      <td>XDD:395_Brain_6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:395</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>EGAN50000223127</td>\n",
       "      <td>10X254_8_AB_1</td>\n",
       "      <td>10X254_8</td>\n",
       "      <td>XDD:385_Thalamus_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>EGAN50000223128</td>\n",
       "      <td>10X254_4_AB_1</td>\n",
       "      <td>10X254_4</td>\n",
       "      <td>XDD:385_Cerebellum_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>EGAN50000223129</td>\n",
       "      <td>10X302_2_AB_1</td>\n",
       "      <td>10X302_2</td>\n",
       "      <td>XDD:398_Hindbrain_7w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:398</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>EGAN50000223130</td>\n",
       "      <td>10X298_4_AB_1</td>\n",
       "      <td>10X298_4</td>\n",
       "      <td>XDD:400_Hindbrain_5.5w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:400</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>EGAN50000223131</td>\n",
       "      <td>10X187_1_ABC_1</td>\n",
       "      <td>10X187_1</td>\n",
       "      <td>XDD:351_Cortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>2025-02-24 09:34:13.348897+01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession_id           alias Sample_ID              description  \\\n",
       "0   EGAN50000223063  10X188_4_ABC_1  10X188_4    XDD:351_Subcortex_12w   \n",
       "1   EGAN50000223064   10X302_1_AB_1  10X302_1     XDD:398_Hindbrain_7w   \n",
       "2   EGAN50000223065   10X252_4_AB_1  10X252_4  XDD:385_Hippocampus_14w   \n",
       "3   EGAN50000223066  10X186_1_ABC_1  10X186_1   XDD:351_Cerebellum_12w   \n",
       "4   EGAN50000223067   10X287_7_AB_1  10X287_7         XDD:395_Brain_6w   \n",
       "..              ...             ...       ...                      ...   \n",
       "64  EGAN50000223127   10X254_8_AB_1  10X254_8     XDD:385_Thalamus_14w   \n",
       "65  EGAN50000223128   10X254_4_AB_1  10X254_4   XDD:385_Cerebellum_14w   \n",
       "66  EGAN50000223129   10X302_2_AB_1  10X302_2     XDD:398_Hindbrain_7w   \n",
       "67  EGAN50000223130   10X298_4_AB_1  10X298_4   XDD:400_Hindbrain_5.5w   \n",
       "68  EGAN50000223131  10X187_1_ABC_1  10X187_1       XDD:351_Cortex_12w   \n",
       "\n",
       "   biological_sex subject_id phenotype extra_attributes  \\\n",
       "0         unknown    XDD:351    normal             [{}]   \n",
       "1         unknown    XDD:398    normal             [{}]   \n",
       "2         unknown    XDD:385    normal             [{}]   \n",
       "3         unknown    XDD:351    normal             [{}]   \n",
       "4         unknown    XDD:395    normal             [{}]   \n",
       "..            ...        ...       ...              ...   \n",
       "64        unknown    XDD:385    normal             [{}]   \n",
       "65        unknown    XDD:385    normal             [{}]   \n",
       "66        unknown    XDD:398    normal             [{}]   \n",
       "67        unknown    XDD:400    normal             [{}]   \n",
       "68        unknown    XDD:351    normal             [{}]   \n",
       "\n",
       "   submission_accession_id                     created_at  \\\n",
       "0           EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "1           EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "2           EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "3           EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "4           EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "..                     ...                            ...   \n",
       "64          EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "65          EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "66          EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "67          EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "68          EGA50000000829  2025-02-10 15:54:26.391641+01   \n",
       "\n",
       "                        edited_at  \n",
       "0   2025-02-24 09:34:13.348897+01  \n",
       "1   2025-02-24 09:34:13.348897+01  \n",
       "2   2025-02-24 09:34:13.348897+01  \n",
       "3   2025-02-24 09:34:13.348897+01  \n",
       "4   2025-02-24 09:34:13.348897+01  \n",
       "..                            ...  \n",
       "64  2025-02-24 09:34:13.348897+01  \n",
       "65  2025-02-24 09:34:13.348897+01  \n",
       "66  2025-02-24 09:34:13.348897+01  \n",
       "67  2025-02-24 09:34:13.348897+01  \n",
       "68  2025-02-24 09:34:13.348897+01  \n",
       "\n",
       "[69 rows x 11 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGAD50000001295"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#split into a list of tokens\n",
    "splits = EGAD00001006049['Donor_Anatomical_Site_Age'].str.split('_')\n",
    "EGAD00001006049['Donor_ID'] = splits.str[0]\n",
    "# 1a) AGE is always the last token, just strip the “w” off\n",
    "EGAD00001006049['Age'] = splits.str[-1].str.rstrip('w')\n",
    "\n",
    "# 1b) ANATOMICAL_SITE is everything *between* donor‐ID (first token) and age (last token).\n",
    "#     that may be one token (e.g. “Forebrain”) or several (e.g. “Spinal_cord_lumbar”).\n",
    "#    If you really only want the “broad” site (just Spinal_cord, not the lumbar region),\n",
    "#    you can tweak: take only the first two tokens after the donor ID.\n",
    "\n",
    "# approach (i): keep the full string between first and last _ \n",
    "EGAD00001006049['Anatomical_Site'] = splits.apply(lambda x: '_'.join(x[1:-1]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "EGAD00001006049[\"Method\"]= \"3GEX\"\n",
    "EGAD50000001295[\"Method\"]= \"3GEX\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Organ</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample_pf</th>\n",
       "      <th>Sampling_Site</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Sex_pf</th>\n",
       "      <th>Sort_id_pf</th>\n",
       "      <th>Termination_method_pf</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>abTCR_library_id_pf</th>\n",
       "      <th>gdTCR_library_id_pf</th>\n",
       "      <th>Organ.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCA_BN_F12482363</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCA_BN_F12482364</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCA_BN_F12482365</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCA_BN_F12482366</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>8.5</td>\n",
       "      <td>8.5</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2361</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCA_BN_F12966455</td>\n",
       "      <td>Calvaria</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>E-MTAB-14385</td>\n",
       "      <td>9.4</td>\n",
       "      <td>9.4</td>\n",
       "      <td>Bone_Atlas</td>\n",
       "      <td>2384</td>\n",
       "      <td>Total</td>\n",
       "      <td>calvaria</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Calvaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>HCA_BRA_F13774552</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>HCA_BRA_F13774553</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>HCA_BRA_F13774554</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>HCA_BRA_F13774555</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>HCA_BRA_F13774556</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Total</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Meninges</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample_ID      Organ 10X_Chemistry  Accession_ID           Age  \\\n",
       "0     HCA_BN_F12482363  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "1     HCA_BN_F12482364  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "2     HCA_BN_F12482365  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "3     HCA_BN_F12482366  Calvaria   10x_multiome  E-MTAB-14385           8.5   \n",
       "4     HCA_BN_F12966455  Calvaria   10x_multiome  E-MTAB-14385           9.4   \n",
       "..                 ...        ...           ...           ...           ...   \n",
       "483  HCA_BRA_F13774552   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "484  HCA_BRA_F13774553   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "485  HCA_BRA_F13774554   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "486  HCA_BRA_F13774555   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "487  HCA_BRA_F13774556   Meninges  10x_multiome        Sanger  9.2,9.9,12.5   \n",
       "\n",
       "         Age(PCW)      Dataset     Donor_ID FACS_Sorting Anatomical_Site  ...  \\\n",
       "0             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "1             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "2             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "3             8.5   Bone_Atlas         2361        Total        calvaria  ...   \n",
       "4             9.4   Bone_Atlas         2384        Total        calvaria  ...   \n",
       "..            ...          ...          ...          ...             ...  ...   \n",
       "483  9.2,9.9,12.5  Cross_Organ  multiplexed       Total      multiplexed  ...   \n",
       "484  9.2,9.9,12.5  Cross_Organ  multiplexed       Total      multiplexed  ...   \n",
       "485  9.2,9.9,12.5  Cross_Organ  multiplexed       Total           sb-ant  ...   \n",
       "486  9.2,9.9,12.5  Cross_Organ  multiplexed       Total          sb-post  ...   \n",
       "487  9.2,9.9,12.5  Cross_Organ  multiplexed       Total         sagittal  ...   \n",
       "\n",
       "    Sample_pf Sampling_Site      Sex Sex_pf Sort_id_pf Termination_method_pf  \\\n",
       "0         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "1         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "2         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "3         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "4         NaN           NaN      NaN    NaN        NaN                   NaN   \n",
       "..        ...           ...      ...    ...        ...                   ...   \n",
       "483       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "484       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "485       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "486       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "487       NaN           NaN  Unknown    NaN        NaN                   NaN   \n",
       "\n",
       "    Unnamed: 14  abTCR_library_id_pf gdTCR_library_id_pf    Organ.1  \n",
       "0           NaN                  NaN                 NaN  Calvaria   \n",
       "1           NaN                  NaN                 NaN  Calvaria   \n",
       "2           NaN                  NaN                 NaN  Calvaria   \n",
       "3           NaN                  NaN                 NaN  Calvaria   \n",
       "4           NaN                  NaN                 NaN  Calvaria   \n",
       "..          ...                  ...                 ...        ...  \n",
       "483         NaN                  NaN                 NaN   Meninges  \n",
       "484         NaN                  NaN                 NaN   Meninges  \n",
       "485         NaN                  NaN                 NaN   Meninges  \n",
       "486         NaN                  NaN                 NaN   Meninges  \n",
       "487         NaN                  NaN                 NaN   Meninges  \n",
       "\n",
       "[482 rows x 45 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      3GEX\n",
       "1      3GEX\n",
       "2      3GEX\n",
       "3      3GEX\n",
       "4      3GEX\n",
       "       ... \n",
       "374    3GEX\n",
       "375    3GEX\n",
       "376    3GEX\n",
       "377    3GEX\n",
       "378    3GEX\n",
       "Name: Method, Length: 379, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGAD00001006049[\"Method\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10X_Chemistry\n",
       "10x3'v2         237\n",
       "10x5'v1         117\n",
       "10x_multiome     66\n",
       "10x5'v2          37\n",
       "3GEX             13\n",
       "5GEX             12\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[\"10X_Chemistry\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(EGAD00001006049[\"Sample_ID\"]) - set(lm[\"Sample_ID\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# transfer 10X chemistry info from lm dataframe to EGAD00001006049\n",
    "\n",
    "# 1) define your mapping from the raw lm.Chemistry to the string you want\n",
    "chem_map = {\n",
    "    'v2': \"10x3'v2\",\n",
    "    'v3': \"10x3'v3\"\n",
    "}\n",
    "\n",
    "# 2) do a left‐merge so every sample in EGAD00001006049 gets its Chemistry (or NaN)\n",
    "merged = EGAD00001006049.merge(\n",
    "    lm[['Sample_ID','Chemistry']],\n",
    "    on='Sample_ID',\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# 3) map v2→\"10x3'v2\", v3→\"10x3'v3\", leave NaN for missing\n",
    "mapped = merged['Chemistry'].map(chem_map)\n",
    "\n",
    "# 4) fill missing (NaN) with the default \"10x3'v2\"\n",
    "EGAD00001006049['10X_Chemistry'] = mapped.fillna(\"10x3'v2\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "EGAD00001006049.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_EGAD00001006049.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>accession_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Donor_Anatomical_Site_Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Disease_Status</th>\n",
       "      <th>Further_Attributes</th>\n",
       "      <th>extra_attributes</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Study</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Method</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>EGAN00002408138</td>\n",
       "      <td>10X147_5_AB_1</td>\n",
       "      <td>10X147_5</td>\n",
       "      <td>6</td>\n",
       "      <td>XDD:326_Spinal_cord_lumbar_6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:326</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631142</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Spinal_cord_lumbar</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>EGAN00002408139</td>\n",
       "      <td>10X101_8_GRCh38-3_0_0</td>\n",
       "      <td>10X101_8</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631144</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>EGAN00002408140</td>\n",
       "      <td>10X101_7_GRCh38-3_0_0</td>\n",
       "      <td>10X101_7</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631145</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>EGAN00002408141</td>\n",
       "      <td>10X101_6_GRCh38-3_0_0</td>\n",
       "      <td>10X101_6</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631146</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>EGAN00002408142</td>\n",
       "      <td>10X101_5_GRCh38-3_0_0</td>\n",
       "      <td>10X101_5</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631147</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>374</td>\n",
       "      <td>EGAN00002408512</td>\n",
       "      <td>10X118_3_GRCh38-3_0_0</td>\n",
       "      <td>10X118_3</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Tel/diencephalon_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Tel/diencephalon</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>375</td>\n",
       "      <td>EGAN00002408513</td>\n",
       "      <td>10X118_4_GRCh38-3_0_0</td>\n",
       "      <td>10X118_4</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Tel/diencephalon_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Tel/diencephalon</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>376</td>\n",
       "      <td>EGAN00002408514</td>\n",
       "      <td>10X118_5_GRCh38-3_0_0</td>\n",
       "      <td>10X118_5</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Medulla_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Medulla</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>377</td>\n",
       "      <td>EGAN00002408515</td>\n",
       "      <td>10X118_6_GRCh38-3_0_0</td>\n",
       "      <td>10X118_6</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Medulla_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Medulla</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>378</td>\n",
       "      <td>EGAN00002408516</td>\n",
       "      <td>10X118_7_GRCh38-3_0_0</td>\n",
       "      <td>10X118_7</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Medulla_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Medulla</td>\n",
       "      <td>10x3'v2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>379 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0     accession_id                  alias Sample_ID  Age  \\\n",
       "0             0  EGAN00002408138          10X147_5_AB_1  10X147_5    6   \n",
       "1             1  EGAN00002408139  10X101_8_GRCh38-3_0_0  10X101_8   10   \n",
       "2             2  EGAN00002408140  10X101_7_GRCh38-3_0_0  10X101_7   10   \n",
       "3             3  EGAN00002408141  10X101_6_GRCh38-3_0_0  10X101_6   10   \n",
       "4             4  EGAN00002408142  10X101_5_GRCh38-3_0_0  10X101_5   10   \n",
       "..          ...              ...                    ...       ...  ...   \n",
       "374         374  EGAN00002408512  10X118_3_GRCh38-3_0_0  10X118_3  6.6   \n",
       "375         375  EGAN00002408513  10X118_4_GRCh38-3_0_0  10X118_4  6.6   \n",
       "376         376  EGAN00002408514  10X118_5_GRCh38-3_0_0  10X118_5  6.6   \n",
       "377         377  EGAN00002408515  10X118_6_GRCh38-3_0_0  10X118_6  6.6   \n",
       "378         378  EGAN00002408516  10X118_7_GRCh38-3_0_0  10X118_7  6.6   \n",
       "\n",
       "          Donor_Anatomical_Site_Age      Sex  Donor_ID Disease_Status  \\\n",
       "0     XDD:326_Spinal_cord_lumbar_6w  unknown   XDD:326         normal   \n",
       "1             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "2             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "3             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "4             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "..                              ...      ...       ...            ...   \n",
       "374  BRC2106A_Tel/diencephalon_6.6w  unknown  BRC2106A         normal   \n",
       "375  BRC2106A_Tel/diencephalon_6.6w  unknown  BRC2106A         normal   \n",
       "376           BRC2106A_Medulla_6.6w  unknown  BRC2106A         normal   \n",
       "377           BRC2106A_Medulla_6.6w  unknown  BRC2106A         normal   \n",
       "378           BRC2106A_Medulla_6.6w  unknown  BRC2106A         normal   \n",
       "\n",
       "                                    Further_Attributes extra_attributes  \\\n",
       "0    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631142   \n",
       "1    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631144   \n",
       "2    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631145   \n",
       "3    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631146   \n",
       "4    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631147   \n",
       "..                                                 ...              ...   \n",
       "374  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "375  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "376  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "377  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "378  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "\n",
       "     Organ     Technology             Study     Accession_ID Method  \\\n",
       "0    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "1    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "2    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "3    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "4    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "..     ...            ...               ...              ...    ...   \n",
       "374  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "375  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "376  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "377  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "378  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "\n",
       "        Anatomical_Site 10X_Chemistry  \n",
       "0    Spinal_cord_lumbar       10x3'v2  \n",
       "1             Forebrain       10x3'v2  \n",
       "2             Forebrain       10x3'v2  \n",
       "3             Forebrain       10x3'v2  \n",
       "4             Forebrain       10x3'v2  \n",
       "..                  ...           ...  \n",
       "374    Tel/diencephalon       10x3'v2  \n",
       "375    Tel/diencephalon       10x3'v2  \n",
       "376             Medulla       10x3'v2  \n",
       "377             Medulla       10x3'v2  \n",
       "378             Medulla       10x3'v2  \n",
       "\n",
       "[379 rows x 18 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGAD00001006049"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if linnarson metadata from paper covers all the samples from EGAD50000001295\n",
    "\n",
    "s=(set(EGAD50000001295[\"Sample_ID\"]) & set(lm[\"Sample_ID\"]))\n",
    "lmf= lm[lm[\"Sample_ID\"].isin(s)] # yes it does; lmf is the  linnarson metadata from paper with EGAD50000001295 samples only "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specimen</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Dissected_region</th>\n",
       "      <th>Target cell number</th>\n",
       "      <th>Retrieved cell number</th>\n",
       "      <th>Median genes/cell</th>\n",
       "      <th>Median UMIs/cell</th>\n",
       "      <th>QC</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 18</th>\n",
       "      <th>Unnamed: 19</th>\n",
       "      <th>Unnamed: 20</th>\n",
       "      <th>Unnamed: 21</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "      <th>Unnamed: 23</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "      <th>Unnamed: 25</th>\n",
       "      <th>Unnamed: 26</th>\n",
       "      <th>Unnamed: 27</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>XDD:400</td>\n",
       "      <td>10X298_1</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>5000</td>\n",
       "      <td>11183</td>\n",
       "      <td>5113.0</td>\n",
       "      <td>15327.0</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XDD:400</td>\n",
       "      <td>10X298_2</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>5000</td>\n",
       "      <td>10351</td>\n",
       "      <td>4602.0</td>\n",
       "      <td>13209.0</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>XDD:400</td>\n",
       "      <td>10X298_6</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>5000</td>\n",
       "      <td>10098</td>\n",
       "      <td>3065.0</td>\n",
       "      <td>7618.0</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>XDD:400</td>\n",
       "      <td>10X298_7</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>5000</td>\n",
       "      <td>10072</td>\n",
       "      <td>3195.0</td>\n",
       "      <td>7974.5</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>XDD:400</td>\n",
       "      <td>10X298_8</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>5000</td>\n",
       "      <td>10037</td>\n",
       "      <td>3453.0</td>\n",
       "      <td>9027.0</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>XDD:385</td>\n",
       "      <td>10X252_6</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14</td>\n",
       "      <td>Ventral midbrain</td>\n",
       "      <td>5000</td>\n",
       "      <td>2305</td>\n",
       "      <td>2429.0</td>\n",
       "      <td>5049.0</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>XDD:385</td>\n",
       "      <td>10X252_7</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14</td>\n",
       "      <td>Ventral midbrain</td>\n",
       "      <td>5000</td>\n",
       "      <td>2846</td>\n",
       "      <td>2508.5</td>\n",
       "      <td>5151.5</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>XDD:385</td>\n",
       "      <td>10X254_3</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14</td>\n",
       "      <td>Cerebellum</td>\n",
       "      <td>5000</td>\n",
       "      <td>3782</td>\n",
       "      <td>3743.0</td>\n",
       "      <td>9737.5</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>XDD:385</td>\n",
       "      <td>10X254_4</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14</td>\n",
       "      <td>Cerebellum</td>\n",
       "      <td>5000</td>\n",
       "      <td>6932</td>\n",
       "      <td>3198.0</td>\n",
       "      <td>7468.0</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>XDD:385</td>\n",
       "      <td>10X255_4</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14</td>\n",
       "      <td>Cerebellum</td>\n",
       "      <td>5000</td>\n",
       "      <td>4720</td>\n",
       "      <td>4468.5</td>\n",
       "      <td>12957.0</td>\n",
       "      <td>Passed</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Specimen Sample_ID   Age  Age(PCW)  Dissected_region  Target cell number  \\\n",
       "10   XDD:400  10X298_1   5.5         5         Forebrain                5000   \n",
       "11   XDD:400  10X298_2   5.5         5         Forebrain                5000   \n",
       "12   XDD:400  10X298_6   5.5         5          Midbrain                5000   \n",
       "13   XDD:400  10X298_7   5.5         5          Midbrain                5000   \n",
       "14   XDD:400  10X298_8   5.5         5          Midbrain                5000   \n",
       "..       ...       ...   ...       ...               ...                 ...   \n",
       "397  XDD:385  10X252_6  14.0        14  Ventral midbrain                5000   \n",
       "398  XDD:385  10X252_7  14.0        14  Ventral midbrain                5000   \n",
       "399  XDD:385  10X254_3  14.0        14        Cerebellum                5000   \n",
       "400  XDD:385  10X254_4  14.0        14        Cerebellum                5000   \n",
       "401  XDD:385  10X255_4  14.0        14        Cerebellum                5000   \n",
       "\n",
       "     Retrieved cell number  Median genes/cell  Median UMIs/cell      QC  ...  \\\n",
       "10                   11183             5113.0           15327.0  Passed  ...   \n",
       "11                   10351             4602.0           13209.0  Passed  ...   \n",
       "12                   10098             3065.0            7618.0  Passed  ...   \n",
       "13                   10072             3195.0            7974.5  Passed  ...   \n",
       "14                   10037             3453.0            9027.0  Passed  ...   \n",
       "..                     ...                ...               ...     ...  ...   \n",
       "397                   2305             2429.0            5049.0  Passed  ...   \n",
       "398                   2846             2508.5            5151.5  Passed  ...   \n",
       "399                   3782             3743.0            9737.5  Passed  ...   \n",
       "400                   6932             3198.0            7468.0  Passed  ...   \n",
       "401                   4720             4468.5           12957.0  Passed  ...   \n",
       "\n",
       "    Unnamed: 18 Unnamed: 19 Unnamed: 20  Unnamed: 21  Unnamed: 22  \\\n",
       "10          NaN         NaN         NaN          NaN          NaN   \n",
       "11          NaN         NaN         NaN          NaN          NaN   \n",
       "12          NaN         NaN         NaN          NaN          NaN   \n",
       "13          NaN         NaN         NaN          NaN          NaN   \n",
       "14          NaN         NaN         NaN          NaN          NaN   \n",
       "..          ...         ...         ...          ...          ...   \n",
       "397         NaN         NaN         NaN          NaN          NaN   \n",
       "398         NaN         NaN         NaN          NaN          NaN   \n",
       "399         NaN         NaN         NaN          NaN          NaN   \n",
       "400         NaN         NaN         NaN          NaN          NaN   \n",
       "401         NaN         NaN         NaN          NaN          NaN   \n",
       "\n",
       "     Unnamed: 23  Unnamed: 24  Unnamed: 25  Unnamed: 26  Unnamed: 27  \n",
       "10           NaN          NaN          NaN          NaN          NaN  \n",
       "11           NaN          NaN          NaN          NaN          NaN  \n",
       "12           NaN          NaN          NaN          NaN          NaN  \n",
       "13           NaN          NaN          NaN          NaN          NaN  \n",
       "14           NaN          NaN          NaN          NaN          NaN  \n",
       "..           ...          ...          ...          ...          ...  \n",
       "397          NaN          NaN          NaN          NaN          NaN  \n",
       "398          NaN          NaN          NaN          NaN          NaN  \n",
       "399          NaN          NaN          NaN          NaN          NaN  \n",
       "400          NaN          NaN          NaN          NaN          NaN  \n",
       "401          NaN          NaN          NaN          NaN          NaN  \n",
       "\n",
       "[69 rows x 28 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lmf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = EGAD50000001295.merge(\n",
    "    lmf,\n",
    "    on='Sample_ID',\n",
    "    how='left'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all(merged[\"Specimen\"]==merged[\"subject_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged[\"Technology\"]='10X_scRNA-seq'\n",
    "merged[\"Study\"]='Braun_et_al_2023'\n",
    "merged[\"Accession_ID\"]='EGAD50000001295'\n",
    "merged[\"Organ\"]= \"Brain\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>description</th>\n",
       "      <th>biological_sex</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>extra_attributes</th>\n",
       "      <th>submission_accession_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "      <th>Unnamed: 23</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "      <th>Unnamed: 25</th>\n",
       "      <th>Unnamed: 26</th>\n",
       "      <th>Unnamed: 27</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Study</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Organ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGAN50000223063</td>\n",
       "      <td>10X188_4_ABC_1</td>\n",
       "      <td>10X188_4</td>\n",
       "      <td>XDD:351_Subcortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGAN50000223064</td>\n",
       "      <td>10X302_1_AB_1</td>\n",
       "      <td>10X302_1</td>\n",
       "      <td>XDD:398_Hindbrain_7w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:398</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGAN50000223065</td>\n",
       "      <td>10X252_4_AB_1</td>\n",
       "      <td>10X252_4</td>\n",
       "      <td>XDD:385_Hippocampus_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGAN50000223066</td>\n",
       "      <td>10X186_1_ABC_1</td>\n",
       "      <td>10X186_1</td>\n",
       "      <td>XDD:351_Cerebellum_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGAN50000223067</td>\n",
       "      <td>10X287_7_AB_1</td>\n",
       "      <td>10X287_7</td>\n",
       "      <td>XDD:395_Brain_6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:395</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>EGAN50000223127</td>\n",
       "      <td>10X254_8_AB_1</td>\n",
       "      <td>10X254_8</td>\n",
       "      <td>XDD:385_Thalamus_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>EGAN50000223128</td>\n",
       "      <td>10X254_4_AB_1</td>\n",
       "      <td>10X254_4</td>\n",
       "      <td>XDD:385_Cerebellum_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>EGAN50000223129</td>\n",
       "      <td>10X302_2_AB_1</td>\n",
       "      <td>10X302_2</td>\n",
       "      <td>XDD:398_Hindbrain_7w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:398</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>EGAN50000223130</td>\n",
       "      <td>10X298_4_AB_1</td>\n",
       "      <td>10X298_4</td>\n",
       "      <td>XDD:400_Hindbrain_5.5w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:400</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>EGAN50000223131</td>\n",
       "      <td>10X187_1_ABC_1</td>\n",
       "      <td>10X187_1</td>\n",
       "      <td>XDD:351_Cortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession_id           alias Sample_ID              description  \\\n",
       "0   EGAN50000223063  10X188_4_ABC_1  10X188_4    XDD:351_Subcortex_12w   \n",
       "1   EGAN50000223064   10X302_1_AB_1  10X302_1     XDD:398_Hindbrain_7w   \n",
       "2   EGAN50000223065   10X252_4_AB_1  10X252_4  XDD:385_Hippocampus_14w   \n",
       "3   EGAN50000223066  10X186_1_ABC_1  10X186_1   XDD:351_Cerebellum_12w   \n",
       "4   EGAN50000223067   10X287_7_AB_1  10X287_7         XDD:395_Brain_6w   \n",
       "..              ...             ...       ...                      ...   \n",
       "64  EGAN50000223127   10X254_8_AB_1  10X254_8     XDD:385_Thalamus_14w   \n",
       "65  EGAN50000223128   10X254_4_AB_1  10X254_4   XDD:385_Cerebellum_14w   \n",
       "66  EGAN50000223129   10X302_2_AB_1  10X302_2     XDD:398_Hindbrain_7w   \n",
       "67  EGAN50000223130   10X298_4_AB_1  10X298_4   XDD:400_Hindbrain_5.5w   \n",
       "68  EGAN50000223131  10X187_1_ABC_1  10X187_1       XDD:351_Cortex_12w   \n",
       "\n",
       "   biological_sex subject_id phenotype extra_attributes  \\\n",
       "0         unknown    XDD:351    normal             [{}]   \n",
       "1         unknown    XDD:398    normal             [{}]   \n",
       "2         unknown    XDD:385    normal             [{}]   \n",
       "3         unknown    XDD:351    normal             [{}]   \n",
       "4         unknown    XDD:395    normal             [{}]   \n",
       "..            ...        ...       ...              ...   \n",
       "64        unknown    XDD:385    normal             [{}]   \n",
       "65        unknown    XDD:385    normal             [{}]   \n",
       "66        unknown    XDD:398    normal             [{}]   \n",
       "67        unknown    XDD:400    normal             [{}]   \n",
       "68        unknown    XDD:351    normal             [{}]   \n",
       "\n",
       "   submission_accession_id                     created_at  ... Unnamed: 22  \\\n",
       "0           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "1           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "2           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "3           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "4           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "..                     ...                            ...  ...         ...   \n",
       "64          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "65          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "66          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "67          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "68          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "\n",
       "   Unnamed: 23 Unnamed: 24  Unnamed: 25  Unnamed: 26 Unnamed: 27  \\\n",
       "0          NaN         NaN          NaN          NaN         NaN   \n",
       "1          NaN         NaN          NaN          NaN         NaN   \n",
       "2          NaN         NaN          NaN          NaN         NaN   \n",
       "3          NaN         NaN          NaN          NaN         NaN   \n",
       "4          NaN         NaN          NaN          NaN         NaN   \n",
       "..         ...         ...          ...          ...         ...   \n",
       "64         NaN         NaN          NaN          NaN         NaN   \n",
       "65         NaN         NaN          NaN          NaN         NaN   \n",
       "66         NaN         NaN          NaN          NaN         NaN   \n",
       "67         NaN         NaN          NaN          NaN         NaN   \n",
       "68         NaN         NaN          NaN          NaN         NaN   \n",
       "\n",
       "       Technology             Study     Accession_ID  Organ  \n",
       "0   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "1   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "2   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "3   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "4   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "..            ...               ...              ...    ...  \n",
       "64  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "65  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "66  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "67  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "68  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "\n",
       "[69 rows x 43 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.rename(\n",
    "    columns={\n",
    "        'description': 'Donor_Anatomical_Site_Age',\n",
    "        'Dissected_region': 'Anatomical_Site',\n",
    "    },\n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Donor_Anatomical_Site_Age</th>\n",
       "      <th>biological_sex</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>extra_attributes</th>\n",
       "      <th>submission_accession_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "      <th>Unnamed: 23</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "      <th>Unnamed: 25</th>\n",
       "      <th>Unnamed: 26</th>\n",
       "      <th>Unnamed: 27</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Study</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Organ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGAN50000223063</td>\n",
       "      <td>10X188_4_ABC_1</td>\n",
       "      <td>10X188_4</td>\n",
       "      <td>XDD:351_Subcortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGAN50000223064</td>\n",
       "      <td>10X302_1_AB_1</td>\n",
       "      <td>10X302_1</td>\n",
       "      <td>XDD:398_Hindbrain_7w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:398</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGAN50000223065</td>\n",
       "      <td>10X252_4_AB_1</td>\n",
       "      <td>10X252_4</td>\n",
       "      <td>XDD:385_Hippocampus_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGAN50000223066</td>\n",
       "      <td>10X186_1_ABC_1</td>\n",
       "      <td>10X186_1</td>\n",
       "      <td>XDD:351_Cerebellum_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGAN50000223067</td>\n",
       "      <td>10X287_7_AB_1</td>\n",
       "      <td>10X287_7</td>\n",
       "      <td>XDD:395_Brain_6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:395</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>EGAN50000223127</td>\n",
       "      <td>10X254_8_AB_1</td>\n",
       "      <td>10X254_8</td>\n",
       "      <td>XDD:385_Thalamus_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>EGAN50000223128</td>\n",
       "      <td>10X254_4_AB_1</td>\n",
       "      <td>10X254_4</td>\n",
       "      <td>XDD:385_Cerebellum_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>EGAN50000223129</td>\n",
       "      <td>10X302_2_AB_1</td>\n",
       "      <td>10X302_2</td>\n",
       "      <td>XDD:398_Hindbrain_7w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:398</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>EGAN50000223130</td>\n",
       "      <td>10X298_4_AB_1</td>\n",
       "      <td>10X298_4</td>\n",
       "      <td>XDD:400_Hindbrain_5.5w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:400</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>EGAN50000223131</td>\n",
       "      <td>10X187_1_ABC_1</td>\n",
       "      <td>10X187_1</td>\n",
       "      <td>XDD:351_Cortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>EGA50000000829</td>\n",
       "      <td>2025-02-10 15:54:26.391641+01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession_id           alias Sample_ID Donor_Anatomical_Site_Age  \\\n",
       "0   EGAN50000223063  10X188_4_ABC_1  10X188_4     XDD:351_Subcortex_12w   \n",
       "1   EGAN50000223064   10X302_1_AB_1  10X302_1      XDD:398_Hindbrain_7w   \n",
       "2   EGAN50000223065   10X252_4_AB_1  10X252_4   XDD:385_Hippocampus_14w   \n",
       "3   EGAN50000223066  10X186_1_ABC_1  10X186_1    XDD:351_Cerebellum_12w   \n",
       "4   EGAN50000223067   10X287_7_AB_1  10X287_7          XDD:395_Brain_6w   \n",
       "..              ...             ...       ...                       ...   \n",
       "64  EGAN50000223127   10X254_8_AB_1  10X254_8      XDD:385_Thalamus_14w   \n",
       "65  EGAN50000223128   10X254_4_AB_1  10X254_4    XDD:385_Cerebellum_14w   \n",
       "66  EGAN50000223129   10X302_2_AB_1  10X302_2      XDD:398_Hindbrain_7w   \n",
       "67  EGAN50000223130   10X298_4_AB_1  10X298_4    XDD:400_Hindbrain_5.5w   \n",
       "68  EGAN50000223131  10X187_1_ABC_1  10X187_1        XDD:351_Cortex_12w   \n",
       "\n",
       "   biological_sex subject_id phenotype extra_attributes  \\\n",
       "0         unknown    XDD:351    normal             [{}]   \n",
       "1         unknown    XDD:398    normal             [{}]   \n",
       "2         unknown    XDD:385    normal             [{}]   \n",
       "3         unknown    XDD:351    normal             [{}]   \n",
       "4         unknown    XDD:395    normal             [{}]   \n",
       "..            ...        ...       ...              ...   \n",
       "64        unknown    XDD:385    normal             [{}]   \n",
       "65        unknown    XDD:385    normal             [{}]   \n",
       "66        unknown    XDD:398    normal             [{}]   \n",
       "67        unknown    XDD:400    normal             [{}]   \n",
       "68        unknown    XDD:351    normal             [{}]   \n",
       "\n",
       "   submission_accession_id                     created_at  ... Unnamed: 22  \\\n",
       "0           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "1           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "2           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "3           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "4           EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "..                     ...                            ...  ...         ...   \n",
       "64          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "65          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "66          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "67          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "68          EGA50000000829  2025-02-10 15:54:26.391641+01  ...         NaN   \n",
       "\n",
       "   Unnamed: 23 Unnamed: 24  Unnamed: 25  Unnamed: 26 Unnamed: 27  \\\n",
       "0          NaN         NaN          NaN          NaN         NaN   \n",
       "1          NaN         NaN          NaN          NaN         NaN   \n",
       "2          NaN         NaN          NaN          NaN         NaN   \n",
       "3          NaN         NaN          NaN          NaN         NaN   \n",
       "4          NaN         NaN          NaN          NaN         NaN   \n",
       "..         ...         ...          ...          ...         ...   \n",
       "64         NaN         NaN          NaN          NaN         NaN   \n",
       "65         NaN         NaN          NaN          NaN         NaN   \n",
       "66         NaN         NaN          NaN          NaN         NaN   \n",
       "67         NaN         NaN          NaN          NaN         NaN   \n",
       "68         NaN         NaN          NaN          NaN         NaN   \n",
       "\n",
       "       Technology             Study     Accession_ID  Organ  \n",
       "0   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "1   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "2   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "3   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "4   10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "..            ...               ...              ...    ...  \n",
       "64  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "65  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "66  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "67  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "68  10X_scRNA-seq  Braun_et_al_2023  EGAD50000001295  Brain  \n",
       "\n",
       "[69 rows x 43 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "chem_map = {\n",
    "    'v2': \"10x3'v2\",\n",
    "    'v3': \"10x3'v3\"\n",
    "}\n",
    "mapped = merged['Chemistry'].map(chem_map)\n",
    "merged['10X_Chemistry'] = mapped.fillna(\"NA\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10X_Chemistry\n",
       "10x3'v3    68\n",
       "10x3'v2     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged['10X_Chemistry'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chemistry\n",
       "v3    68\n",
       "v2     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged['Chemistry'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of columns to remove\n",
    "cols_to_drop = [f'Unnamed: {i}' for i in range(12, 28)]\n",
    "\n",
    "# drop them (returns a new DataFrame)\n",
    "merged = merged.drop(columns=cols_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "EGAD00001006049[\"Age(PCW)\"]= EGAD00001006049[\"Age\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession_id', 'alias', 'Sample_ID', 'Donor_Anatomical_Site_Age',\n",
       "       'biological_sex', 'subject_id', 'phenotype', 'extra_attributes',\n",
       "       'submission_accession_id', 'created_at', 'edited_at', 'Method',\n",
       "       'Specimen', 'Age', 'Age(PCW)', 'Anatomical_Site', 'Target cell number',\n",
       "       'Retrieved cell number', 'Median genes/cell', 'Median UMIs/cell', 'QC',\n",
       "       'Chemistry', 'Source', 'Technology', 'Study', 'Accession_ID', 'Organ',\n",
       "       '10X_Chemistry'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'accession_id', 'alias', 'Sample_ID', 'Age',\n",
       "       'Donor_Anatomical_Site_Age', 'Sex', 'Donor_ID', 'Disease_Status',\n",
       "       'Further_Attributes', 'extra_attributes', 'Organ', 'Technology',\n",
       "       'Study', 'Accession_ID', 'Method', 'Anatomical_Site', '10X_Chemistry',\n",
       "       'Age(PCW)'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGAD00001006049.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       Total\n",
       "1       Total\n",
       "2       Total\n",
       "3       Total\n",
       "4       Total\n",
       "        ...  \n",
       "483    Total \n",
       "484    Total \n",
       "485    Total \n",
       "486    Total \n",
       "487    Total \n",
       "Name: FACS_Sorting, Length: 482, dtype: object"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['FACS_Sorting']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.rename(\n",
    "    columns={\n",
    "        'biological_sex': 'Sex',\n",
    "        'subject_id': 'Donor_ID',\n",
    "    },\n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged[\"Organism\"]= \"Human\"\n",
    "merged[\"Multiplexed\"]= \"No\"\n",
    "merged[\"Expected_Donors\"]=1\n",
    "merged[\"FACS_Sorting\"]=\"Total\"\n",
    "merged[\"Dataset\"]=\"Linnarson\"\n",
    "merged['Disease_Status']='normal'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_EGAD50000001295.csv') # Linnarson data metadata from EGA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>accession_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Donor_Anatomical_Site_Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Disease_Status</th>\n",
       "      <th>Further_Attributes</th>\n",
       "      <th>extra_attributes</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Study</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Method</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Age(PCW)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>EGAN00002408138</td>\n",
       "      <td>10X147_5_AB_1</td>\n",
       "      <td>10X147_5</td>\n",
       "      <td>6</td>\n",
       "      <td>XDD:326_Spinal_cord_lumbar_6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:326</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631142</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Spinal_cord_lumbar</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>EGAN00002408139</td>\n",
       "      <td>10X101_8_GRCh38-3_0_0</td>\n",
       "      <td>10X101_8</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631144</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>EGAN00002408140</td>\n",
       "      <td>10X101_7_GRCh38-3_0_0</td>\n",
       "      <td>10X101_7</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631145</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>EGAN00002408141</td>\n",
       "      <td>10X101_6_GRCh38-3_0_0</td>\n",
       "      <td>10X101_6</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631146</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>EGAN00002408142</td>\n",
       "      <td>10X101_5_GRCh38-3_0_0</td>\n",
       "      <td>10X101_5</td>\n",
       "      <td>10</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631147</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>374</td>\n",
       "      <td>EGAN00002408512</td>\n",
       "      <td>10X118_3_GRCh38-3_0_0</td>\n",
       "      <td>10X118_3</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Tel/diencephalon_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Tel/diencephalon</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>6.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>375</td>\n",
       "      <td>EGAN00002408513</td>\n",
       "      <td>10X118_4_GRCh38-3_0_0</td>\n",
       "      <td>10X118_4</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Tel/diencephalon_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Tel/diencephalon</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>6.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>376</td>\n",
       "      <td>EGAN00002408514</td>\n",
       "      <td>10X118_5_GRCh38-3_0_0</td>\n",
       "      <td>10X118_5</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Medulla_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Medulla</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>6.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>377</td>\n",
       "      <td>EGAN00002408515</td>\n",
       "      <td>10X118_6_GRCh38-3_0_0</td>\n",
       "      <td>10X118_6</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Medulla_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Medulla</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>6.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>378</td>\n",
       "      <td>EGAN00002408516</td>\n",
       "      <td>10X118_7_GRCh38-3_0_0</td>\n",
       "      <td>10X118_7</td>\n",
       "      <td>6.6</td>\n",
       "      <td>BRC2106A_Medulla_6.6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BRC2106A</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{\"tag\": \"case_or_control\", \"unit\": null, \"val...</td>\n",
       "      <td>EGA00001631556</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Medulla</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>6.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>379 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0     accession_id                  alias Sample_ID  Age  \\\n",
       "0             0  EGAN00002408138          10X147_5_AB_1  10X147_5    6   \n",
       "1             1  EGAN00002408139  10X101_8_GRCh38-3_0_0  10X101_8   10   \n",
       "2             2  EGAN00002408140  10X101_7_GRCh38-3_0_0  10X101_7   10   \n",
       "3             3  EGAN00002408141  10X101_6_GRCh38-3_0_0  10X101_6   10   \n",
       "4             4  EGAN00002408142  10X101_5_GRCh38-3_0_0  10X101_5   10   \n",
       "..          ...              ...                    ...       ...  ...   \n",
       "374         374  EGAN00002408512  10X118_3_GRCh38-3_0_0  10X118_3  6.6   \n",
       "375         375  EGAN00002408513  10X118_4_GRCh38-3_0_0  10X118_4  6.6   \n",
       "376         376  EGAN00002408514  10X118_5_GRCh38-3_0_0  10X118_5  6.6   \n",
       "377         377  EGAN00002408515  10X118_6_GRCh38-3_0_0  10X118_6  6.6   \n",
       "378         378  EGAN00002408516  10X118_7_GRCh38-3_0_0  10X118_7  6.6   \n",
       "\n",
       "          Donor_Anatomical_Site_Age      Sex  Donor_ID Disease_Status  \\\n",
       "0     XDD:326_Spinal_cord_lumbar_6w  unknown   XDD:326         normal   \n",
       "1             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "2             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "3             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "4             XHU:297_Forebrain_10w  unknown   XHU:297         normal   \n",
       "..                              ...      ...       ...            ...   \n",
       "374  BRC2106A_Tel/diencephalon_6.6w  unknown  BRC2106A         normal   \n",
       "375  BRC2106A_Tel/diencephalon_6.6w  unknown  BRC2106A         normal   \n",
       "376           BRC2106A_Medulla_6.6w  unknown  BRC2106A         normal   \n",
       "377           BRC2106A_Medulla_6.6w  unknown  BRC2106A         normal   \n",
       "378           BRC2106A_Medulla_6.6w  unknown  BRC2106A         normal   \n",
       "\n",
       "                                    Further_Attributes extra_attributes  \\\n",
       "0    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631142   \n",
       "1    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631144   \n",
       "2    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631145   \n",
       "3    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631146   \n",
       "4    [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631147   \n",
       "..                                                 ...              ...   \n",
       "374  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "375  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "376  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "377  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "378  [{\"tag\": \"case_or_control\", \"unit\": null, \"val...   EGA00001631556   \n",
       "\n",
       "     Organ     Technology             Study     Accession_ID Method  \\\n",
       "0    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "1    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "2    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "3    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "4    Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "..     ...            ...               ...              ...    ...   \n",
       "374  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "375  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "376  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "377  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "378  Brain  10X_scRNA-seq  Braun_et_al_2023  EGAD00001006049   3GEX   \n",
       "\n",
       "        Anatomical_Site 10X_Chemistry Age(PCW)  \n",
       "0    Spinal_cord_lumbar       10x3'v2        6  \n",
       "1             Forebrain       10x3'v2       10  \n",
       "2             Forebrain       10x3'v2       10  \n",
       "3             Forebrain       10x3'v2       10  \n",
       "4             Forebrain       10x3'v2       10  \n",
       "..                  ...           ...      ...  \n",
       "374    Tel/diencephalon       10x3'v2      6.6  \n",
       "375    Tel/diencephalon       10x3'v2      6.6  \n",
       "376             Medulla       10x3'v2      6.6  \n",
       "377             Medulla       10x3'v2      6.6  \n",
       "378             Medulla       10x3'v2      6.6  \n",
       "\n",
       "[379 rows x 19 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGAD00001006049"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "EGAD00001006049[\"Organism\"]= \"Human\"\n",
    "EGAD00001006049[\"Multiplexed\"]= \"No\"\n",
    "EGAD00001006049[\"Expected_Donors\"]=1\n",
    "EGAD00001006049[\"FACS_Sorting\"]=\"Total\"\n",
    "EGAD00001006049[\"Dataset\"]=\"Linnarson\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "EGAD00001006049.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_EGAD00001006049.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession_id', 'alias', 'Sample_ID', 'Donor_Anatomical_Site_Age',\n",
       "       'Sex', 'Donor_ID', 'phenotype', 'extra_attributes',\n",
       "       'submission_accession_id', 'created_at', 'edited_at', 'Method',\n",
       "       'Specimen', 'Age', 'Age(PCW)', 'Anatomical_Site', 'Target cell number',\n",
       "       'Retrieved cell number', 'Median genes/cell', 'Median UMIs/cell', 'QC',\n",
       "       'Chemistry', 'Source', 'Technology', 'Study', 'Accession_ID', 'Organ',\n",
       "       '10X_Chemistry', 'Organism', 'Multiplexed', 'Expected_Donors',\n",
       "       'FACS_Sorting'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'10X_Chemistry',\n",
       " 'Accession_ID',\n",
       " 'Age',\n",
       " 'Age(PCW)',\n",
       " 'Anatomical_Site',\n",
       " 'Dataset',\n",
       " 'Disease_Status',\n",
       " 'Donor_Anatomical_Site_Age',\n",
       " 'Donor_ID',\n",
       " 'Expected_Donors',\n",
       " 'FACS_Sorting',\n",
       " 'Method',\n",
       " 'Multiplexed',\n",
       " 'Organ',\n",
       " 'Organism',\n",
       " 'Sample_ID',\n",
       " 'Sex',\n",
       " 'Study',\n",
       " 'Technology',\n",
       " 'accession_id',\n",
       " 'alias',\n",
       " 'extra_attributes'}"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(EGAD00001006049.columns) & set(merged.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Chemistry',\n",
       " 'Median UMIs/cell',\n",
       " 'Median genes/cell',\n",
       " 'QC',\n",
       " 'Retrieved cell number',\n",
       " 'Source',\n",
       " 'Specimen',\n",
       " 'Target cell number',\n",
       " 'created_at',\n",
       " 'edited_at',\n",
       " 'phenotype',\n",
       " 'submission_accession_id'}"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(merged.columns) - set(EGAD00001006049.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      normal\n",
       "1      normal\n",
       "2      normal\n",
       "3      normal\n",
       "4      normal\n",
       "        ...  \n",
       "374    normal\n",
       "375    normal\n",
       "376    normal\n",
       "377    normal\n",
       "378    normal\n",
       "Name: Disease_Status, Length: 379, dtype: object"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGAD00001006049['Disease_Status']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1) figure out which columns both DataFrames share\n",
    "common_cols = EGAD00001006049.columns.intersection(merged.columns)\n",
    "\n",
    "# 2) vertically stack only those columns\n",
    "combined = pd.concat(\n",
    "    [EGAD00001006049[common_cols], merged[common_cols]],\n",
    "    ignore_index=True,\n",
    "    axis=0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Donor_Anatomical_Site_Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Disease_Status</th>\n",
       "      <th>extra_attributes</th>\n",
       "      <th>Organ</th>\n",
       "      <th>...</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Method</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Multiplexed</th>\n",
       "      <th>Expected_Donors</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>EGAN50000223118</td>\n",
       "      <td>10X187_8_ABC_1</td>\n",
       "      <td>10X187_8</td>\n",
       "      <td>12.0</td>\n",
       "      <td>XDD:351_Subcortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Subctx (forebrain)</td>\n",
       "      <td>10x3'v3</td>\n",
       "      <td>12</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession_id           alias Sample_ID   Age  \\\n",
       "434  EGAN50000223118  10X187_8_ABC_1  10X187_8  12.0   \n",
       "\n",
       "    Donor_Anatomical_Site_Age      Sex Donor_ID Disease_Status  \\\n",
       "434     XDD:351_Subcortex_12w  unknown  XDD:351         normal   \n",
       "\n",
       "    extra_attributes  Organ  ...     Accession_ID Method     Anatomical_Site  \\\n",
       "434             [{}]  Brain  ...  EGAD50000001295   3GEX  Subctx (forebrain)   \n",
       "\n",
       "    10X_Chemistry Age(PCW) Organism Multiplexed Expected_Donors FACS_Sorting  \\\n",
       "434       10x3'v3       12    Human          No               1        Total   \n",
       "\n",
       "       Dataset  \n",
       "434  Linnarson  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined[combined[\"Sample_ID\"]=='10X187_8']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_Metadata_Final.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "linnarson= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/Linnarson_Metadata_Final.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>alias</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Donor_Anatomical_Site_Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Disease_Status</th>\n",
       "      <th>extra_attributes</th>\n",
       "      <th>Organ</th>\n",
       "      <th>...</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Method</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Age(PCW)</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Multiplexed</th>\n",
       "      <th>Expected_Donors</th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGAN00002408138</td>\n",
       "      <td>10X147_5_AB_1</td>\n",
       "      <td>10X147_5</td>\n",
       "      <td>6.0</td>\n",
       "      <td>XDD:326_Spinal_cord_lumbar_6w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:326</td>\n",
       "      <td>normal</td>\n",
       "      <td>EGA00001631142</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Spinal_cord_lumbar</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGAN00002408139</td>\n",
       "      <td>10X101_8_GRCh38-3_0_0</td>\n",
       "      <td>10X101_8</td>\n",
       "      <td>10.0</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>EGA00001631144</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGAN00002408140</td>\n",
       "      <td>10X101_7_GRCh38-3_0_0</td>\n",
       "      <td>10X101_7</td>\n",
       "      <td>10.0</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>EGA00001631145</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGAN00002408141</td>\n",
       "      <td>10X101_6_GRCh38-3_0_0</td>\n",
       "      <td>10X101_6</td>\n",
       "      <td>10.0</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>EGA00001631146</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGAN00002408142</td>\n",
       "      <td>10X101_5_GRCh38-3_0_0</td>\n",
       "      <td>10X101_5</td>\n",
       "      <td>10.0</td>\n",
       "      <td>XHU:297_Forebrain_10w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>normal</td>\n",
       "      <td>EGA00001631147</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>EGAN50000223127</td>\n",
       "      <td>10X254_8_AB_1</td>\n",
       "      <td>10X254_8</td>\n",
       "      <td>14.0</td>\n",
       "      <td>XDD:385_Thalamus_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Thalamus</td>\n",
       "      <td>10x3'v3</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>EGAN50000223128</td>\n",
       "      <td>10X254_4_AB_1</td>\n",
       "      <td>10X254_4</td>\n",
       "      <td>14.0</td>\n",
       "      <td>XDD:385_Cerebellum_14w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:385</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Cerebellum</td>\n",
       "      <td>10x3'v3</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>EGAN50000223129</td>\n",
       "      <td>10X302_2_AB_1</td>\n",
       "      <td>10X302_2</td>\n",
       "      <td>7.0</td>\n",
       "      <td>XDD:398_Hindbrain_7w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:398</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Hindbrain</td>\n",
       "      <td>10x3'v3</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>EGAN50000223130</td>\n",
       "      <td>10X298_4_AB_1</td>\n",
       "      <td>10X298_4</td>\n",
       "      <td>5.5</td>\n",
       "      <td>XDD:400_Hindbrain_5.5w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:400</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Hindbrain</td>\n",
       "      <td>10x3'v3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>EGAN50000223131</td>\n",
       "      <td>10X187_1_ABC_1</td>\n",
       "      <td>10X187_1</td>\n",
       "      <td>12.0</td>\n",
       "      <td>XDD:351_Cortex_12w</td>\n",
       "      <td>unknown</td>\n",
       "      <td>XDD:351</td>\n",
       "      <td>normal</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>Brain</td>\n",
       "      <td>...</td>\n",
       "      <td>EGAD50000001295</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>10x3'v3</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>Total</td>\n",
       "      <td>Linnarson</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>448 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession_id                  alias Sample_ID   Age  \\\n",
       "0    EGAN00002408138          10X147_5_AB_1  10X147_5   6.0   \n",
       "1    EGAN00002408139  10X101_8_GRCh38-3_0_0  10X101_8  10.0   \n",
       "2    EGAN00002408140  10X101_7_GRCh38-3_0_0  10X101_7  10.0   \n",
       "3    EGAN00002408141  10X101_6_GRCh38-3_0_0  10X101_6  10.0   \n",
       "4    EGAN00002408142  10X101_5_GRCh38-3_0_0  10X101_5  10.0   \n",
       "..               ...                    ...       ...   ...   \n",
       "443  EGAN50000223127          10X254_8_AB_1  10X254_8  14.0   \n",
       "444  EGAN50000223128          10X254_4_AB_1  10X254_4  14.0   \n",
       "445  EGAN50000223129          10X302_2_AB_1  10X302_2   7.0   \n",
       "446  EGAN50000223130          10X298_4_AB_1  10X298_4   5.5   \n",
       "447  EGAN50000223131         10X187_1_ABC_1  10X187_1  12.0   \n",
       "\n",
       "         Donor_Anatomical_Site_Age      Sex Donor_ID Disease_Status  \\\n",
       "0    XDD:326_Spinal_cord_lumbar_6w  unknown  XDD:326         normal   \n",
       "1            XHU:297_Forebrain_10w  unknown  XHU:297         normal   \n",
       "2            XHU:297_Forebrain_10w  unknown  XHU:297         normal   \n",
       "3            XHU:297_Forebrain_10w  unknown  XHU:297         normal   \n",
       "4            XHU:297_Forebrain_10w  unknown  XHU:297         normal   \n",
       "..                             ...      ...      ...            ...   \n",
       "443           XDD:385_Thalamus_14w  unknown  XDD:385         normal   \n",
       "444         XDD:385_Cerebellum_14w  unknown  XDD:385         normal   \n",
       "445           XDD:398_Hindbrain_7w  unknown  XDD:398         normal   \n",
       "446         XDD:400_Hindbrain_5.5w  unknown  XDD:400         normal   \n",
       "447             XDD:351_Cortex_12w  unknown  XDD:351         normal   \n",
       "\n",
       "    extra_attributes  Organ  ...     Accession_ID Method     Anatomical_Site  \\\n",
       "0     EGA00001631142  Brain  ...  EGAD00001006049   3GEX  Spinal_cord_lumbar   \n",
       "1     EGA00001631144  Brain  ...  EGAD00001006049   3GEX           Forebrain   \n",
       "2     EGA00001631145  Brain  ...  EGAD00001006049   3GEX           Forebrain   \n",
       "3     EGA00001631146  Brain  ...  EGAD00001006049   3GEX           Forebrain   \n",
       "4     EGA00001631147  Brain  ...  EGAD00001006049   3GEX           Forebrain   \n",
       "..               ...    ...  ...              ...    ...                 ...   \n",
       "443             [{}]  Brain  ...  EGAD50000001295   3GEX            Thalamus   \n",
       "444             [{}]  Brain  ...  EGAD50000001295   3GEX          Cerebellum   \n",
       "445             [{}]  Brain  ...  EGAD50000001295   3GEX           Hindbrain   \n",
       "446             [{}]  Brain  ...  EGAD50000001295   3GEX           Hindbrain   \n",
       "447             [{}]  Brain  ...  EGAD50000001295   3GEX              Cortex   \n",
       "\n",
       "    10X_Chemistry Age(PCW) Organism  Multiplexed Expected_Donors FACS_Sorting  \\\n",
       "0         10x3'v2      6.0    Human           No               1        Total   \n",
       "1         10x3'v2     10.0    Human           No               1        Total   \n",
       "2         10x3'v2     10.0    Human           No               1        Total   \n",
       "3         10x3'v2     10.0    Human           No               1        Total   \n",
       "4         10x3'v2     10.0    Human           No               1        Total   \n",
       "..            ...      ...      ...          ...             ...          ...   \n",
       "443       10x3'v3     14.0    Human           No               1        Total   \n",
       "444       10x3'v3     14.0    Human           No               1        Total   \n",
       "445       10x3'v3      7.0    Human           No               1        Total   \n",
       "446       10x3'v3      5.0    Human           No               1        Total   \n",
       "447       10x3'v3     12.0    Human           No               1        Total   \n",
       "\n",
       "       Dataset  \n",
       "0    Linnarson  \n",
       "1    Linnarson  \n",
       "2    Linnarson  \n",
       "3    Linnarson  \n",
       "4    Linnarson  \n",
       "..         ...  \n",
       "443  Linnarson  \n",
       "444  Linnarson  \n",
       "445  Linnarson  \n",
       "446  Linnarson  \n",
       "447  Linnarson  \n",
       "\n",
       "[448 rows x 22 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linnarson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_METADATA.csv', index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'10X_Chemistry',\n",
       " 'Accession_ID',\n",
       " 'Age',\n",
       " 'Age(PCW)',\n",
       " 'Anatomical_Site',\n",
       " 'Dataset',\n",
       " 'Donor_ID',\n",
       " 'Expected_Donors',\n",
       " 'FACS_Sorting',\n",
       " 'Method',\n",
       " 'Multiplexed',\n",
       " 'Organ',\n",
       " 'Organism',\n",
       " 'Sample_ID',\n",
       " 'Sex',\n",
       " 'Study',\n",
       " 'Technology'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(linnarson.columns) & set(metadata.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "accession_id                  object\n",
       "alias                         object\n",
       "Sample_ID                     object\n",
       "Age                          float64\n",
       "Donor_Anatomical_Site_Age     object\n",
       "Sex                           object\n",
       "Donor_ID                      object\n",
       "Disease_Status                object\n",
       "extra_attributes              object\n",
       "Organ                         object\n",
       "Technology                    object\n",
       "Study                         object\n",
       "Accession_ID                  object\n",
       "Method                        object\n",
       "Anatomical_Site               object\n",
       "10X_Chemistry                 object\n",
       "Age(PCW)                     float64\n",
       "Organism                      object\n",
       "Multiplexed                   object\n",
       "Expected_Donors                int64\n",
       "FACS_Sorting                  object\n",
       "Dataset                       object\n",
       "dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linnarson.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "linnarson['Age(PCW)']= linnarson['Age(PCW)'].astype(str)\n",
    "metadata['Age(PCW)']= metadata['Age(PCW)'].astype(str)\n",
    "linnarson['Age']= linnarson['Age'].astype(str)\n",
    "metadata['Age']= metadata['Age'].astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples= ['10X147_1_ABC_1',\n",
    " '10X186_1_ABC_1',\n",
    " '10X186_4_ABC_1',\n",
    " '10X187_1_ABC_1',\n",
    " '10X187_2_ABC_1',\n",
    " '10X187_3_ABC_1',\n",
    " '10X187_4_ABC_1',\n",
    " '10X187_6_ABC_1',\n",
    " '10X187_8_ABC_1',\n",
    " '10X188_1_ABC_1',\n",
    " '10X188_2_ABC_1',\n",
    " '10X188_3_ABC_1',\n",
    " '10X188_4_ABC_1',\n",
    " '10X251_7_AB_1',\n",
    " '10X251_8_AB_1',\n",
    " '10X252_1_AB_1',\n",
    " '10X252_2_AB_1',\n",
    " '10X252_3_AB_1',\n",
    " '10X252_4_AB_1',\n",
    " '10X252_5_AB_1',\n",
    " '10X252_6_AB_1',\n",
    " '10X252_7_AB_1',\n",
    " '10X254_2_AB_1',\n",
    " '10X254_3_AB_1',\n",
    " '10X254_4_AB_1',\n",
    " '10X254_5_AB_1',\n",
    " '10X254_6_AB_1',\n",
    " '10X254_7_AB_1',\n",
    " '10X254_8_AB_1',\n",
    " '10X255_1_AB_1',\n",
    " '10X255_2_AB_1',\n",
    " '10X255_3_AB_1',\n",
    " '10X255_4_AB_1',\n",
    " '10X255_6_AB_1',\n",
    " '10X255_7_AB_1',\n",
    " '10X257_1_AB_1',\n",
    " '10X257_2_AB_1',\n",
    " '10X257_4_AB_1',\n",
    " '10X257_5_AB_1',\n",
    " '10X257_6_AB_1',\n",
    " '10X287_4_AB_1',\n",
    " '10X287_5_AB_1',\n",
    " '10X287_6_AB_1',\n",
    " '10X287_7_AB_1',\n",
    " '10X287_8_AB_1',\n",
    " '10X288_1_AB_1',\n",
    " '10X288_2_AB_1',\n",
    " '10X288_3_AB_1',\n",
    " '10X288_4_AB_1',\n",
    " '10X288_5_AB_1',\n",
    " '10X288_6_AB_1',\n",
    " '10X288_7_AB_1',\n",
    " '10X288_8_AB_1',\n",
    " '10X298_1_AB_1',\n",
    " '10X298_2_AB_1',\n",
    " '10X298_3_AB_1',\n",
    " '10X298_4_AB_1',\n",
    " '10X298_5_AB_1',\n",
    " '10X298_6_AB_1',\n",
    " '10X298_7_AB_1',\n",
    " '10X298_8_AB_1',\n",
    " '10X302_1_AB_1',\n",
    " '10X302_2_AB_1',\n",
    " '10X302_3_AB_1',\n",
    " '10X302_4_AB_1',\n",
    " '10X302_5_AB_1',\n",
    " '10X302_6_AB_1',\n",
    " '10X302_7_AB_1',\n",
    " '10X302_8_AB_1']\n",
    "# Ensure 'samples' is a set of aliases you want to match\n",
    "linnarson.loc[linnarson['alias'].isin(samples), 'Sample_ID'] = \\\n",
    "    linnarson.loc[linnarson['alias'].isin(samples), 'alias']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "shared_columns = list(set(linnarson.columns) & set(metadata.columns))\n",
    "combined_shared = pd.concat([linnarson[shared_columns], metadata[shared_columns]], axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_shared.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_BRAIN_METADATA_FINAL.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_shared2= pd.read_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_BRAIN_METADATA_FINAL.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FACS_Sorting</th>\n",
       "      <th>Age</th>\n",
       "      <th>Method</th>\n",
       "      <th>Multiplexed</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Organ</th>\n",
       "      <th>10X_Chemistry</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Accession_ID</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Expected_Donors</th>\n",
       "      <th>Study</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Age(PCW)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Total</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>No</td>\n",
       "      <td>Linnarson</td>\n",
       "      <td>Spinal_cord_lumbar</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>XDD:326</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>Human</td>\n",
       "      <td>10X147_5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Total</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>No</td>\n",
       "      <td>Linnarson</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>Human</td>\n",
       "      <td>10X101_8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Total</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>No</td>\n",
       "      <td>Linnarson</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>Human</td>\n",
       "      <td>10X101_7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Total</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>No</td>\n",
       "      <td>Linnarson</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>Human</td>\n",
       "      <td>10X101_6</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Total</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3GEX</td>\n",
       "      <td>No</td>\n",
       "      <td>Linnarson</td>\n",
       "      <td>Forebrain</td>\n",
       "      <td>Brain</td>\n",
       "      <td>10x3'v2</td>\n",
       "      <td>XHU:297</td>\n",
       "      <td>EGAD00001006049</td>\n",
       "      <td>Human</td>\n",
       "      <td>10X101_5</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1</td>\n",
       "      <td>Braun_et_al_2023</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>Total</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HCA_BRA_F13774552</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926</th>\n",
       "      <td>Total</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HCA_BRA_F13774553</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927</th>\n",
       "      <td>Total</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>sb-ant</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HCA_BRA_F13774554</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>928</th>\n",
       "      <td>Total</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>sb-post</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HCA_BRA_F13774555</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>929</th>\n",
       "      <td>Total</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cross_Organ</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>Meninges</td>\n",
       "      <td>10x_multiome</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>HCA_BRA_F13774556</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>3</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>9.2,9.9,12.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>930 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    FACS_Sorting           Age        Method Multiplexed      Dataset  \\\n",
       "0          Total           6.0          3GEX          No    Linnarson   \n",
       "1          Total          10.0          3GEX          No    Linnarson   \n",
       "2          Total          10.0          3GEX          No    Linnarson   \n",
       "3          Total          10.0          3GEX          No    Linnarson   \n",
       "4          Total          10.0          3GEX          No    Linnarson   \n",
       "..           ...           ...           ...         ...          ...   \n",
       "925       Total   9.2,9.9,12.5  10X_multiome         Yes  Cross_Organ   \n",
       "926       Total   9.2,9.9,12.5  10X_multiome         Yes  Cross_Organ   \n",
       "927       Total   9.2,9.9,12.5  10X_multiome         Yes  Cross_Organ   \n",
       "928       Total   9.2,9.9,12.5  10X_multiome         Yes  Cross_Organ   \n",
       "929       Total   9.2,9.9,12.5  10X_multiome         Yes  Cross_Organ   \n",
       "\n",
       "        Anatomical_Site     Organ 10X_Chemistry     Donor_ID     Accession_ID  \\\n",
       "0    Spinal_cord_lumbar     Brain       10x3'v2      XDD:326  EGAD00001006049   \n",
       "1             Forebrain     Brain       10x3'v2      XHU:297  EGAD00001006049   \n",
       "2             Forebrain     Brain       10x3'v2      XHU:297  EGAD00001006049   \n",
       "3             Forebrain     Brain       10x3'v2      XHU:297  EGAD00001006049   \n",
       "4             Forebrain     Brain       10x3'v2      XHU:297  EGAD00001006049   \n",
       "..                  ...       ...           ...          ...              ...   \n",
       "925         multiplexed  Meninges  10x_multiome  multiplexed           Sanger   \n",
       "926         multiplexed  Meninges  10x_multiome  multiplexed           Sanger   \n",
       "927              sb-ant  Meninges  10x_multiome  multiplexed           Sanger   \n",
       "928             sb-post  Meninges  10x_multiome  multiplexed           Sanger   \n",
       "929            sagittal  Meninges  10x_multiome  multiplexed           Sanger   \n",
       "\n",
       "         Organism          Sample_ID      Sex  Expected_Donors  \\\n",
       "0           Human           10X147_5  unknown                1   \n",
       "1           Human           10X101_8  unknown                1   \n",
       "2           Human           10X101_7  unknown                1   \n",
       "3           Human           10X101_6  unknown                1   \n",
       "4           Human           10X101_5  unknown                1   \n",
       "..            ...                ...      ...              ...   \n",
       "925  Homo sapiens  HCA_BRA_F13774552  Unknown                3   \n",
       "926  Homo sapiens  HCA_BRA_F13774553  Unknown                3   \n",
       "927  Homo sapiens  HCA_BRA_F13774554  Unknown                3   \n",
       "928  Homo sapiens  HCA_BRA_F13774555  Unknown                3   \n",
       "929  Homo sapiens  HCA_BRA_F13774556  Unknown                3   \n",
       "\n",
       "                Study     Technology      Age(PCW)  \n",
       "0    Braun_et_al_2023  10X_scRNA-seq           6.0  \n",
       "1    Braun_et_al_2023  10X_scRNA-seq          10.0  \n",
       "2    Braun_et_al_2023  10X_scRNA-seq          10.0  \n",
       "3    Braun_et_al_2023  10X_scRNA-seq          10.0  \n",
       "4    Braun_et_al_2023  10X_scRNA-seq          10.0  \n",
       "..                ...            ...           ...  \n",
       "925            Sanger   10X_multiome  9.2,9.9,12.5  \n",
       "926            Sanger   10X_multiome  9.2,9.9,12.5  \n",
       "927            Sanger   10X_multiome  9.2,9.9,12.5  \n",
       "928            Sanger   10X_multiome  9.2,9.9,12.5  \n",
       "929            Sanger   10X_multiome  9.2,9.9,12.5  \n",
       "\n",
       "[930 rows x 17 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_shared2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_all = pd.concat([linnarson, metadata], axis=0, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_all.to_csv('/nfs/team298/sm54/BoneAtlasProject/metadata/sample_metadata/COMBINED_BONE_CROSS_ORGAN_BRAIN_METADATA_ALL_COLUMNS.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "boneatlas",
   "language": "python",
   "name": "boneatlas"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
