{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aware-lunch",
   "metadata": {},
   "source": [
    "# Read in required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "posted-alexandria",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import matplotlib as mpl\n",
    "import scipy\n",
    "import scrublet as scr\n",
    "#import scvi\n",
    "import hashlib\n",
    "\n",
    "from matplotlib import rcParams\n",
    "rcParams['pdf.fonttype'] = 42 # enables correct plotting of text\n",
    "import seaborn as sns\n",
    "import anndata\n",
    "import os\n",
    "import gc\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "balanced-bangkok",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/nfs/users/nfs_s/sm54/repos/cns-haematopoiesis/analysis/01_preprocess_scRNAseq'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "vital-globe",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "electoral-element",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code provided by Vitalii\n",
    "\n",
    "def read_10x_output(smp_list, metadata=None, type = 'raw'):\n",
    "    import os\n",
    "    \n",
    "    #Writing output from separate samples, processed using CellRanger, into a dictionary of Scanpy objects:\n",
    "    ad = {}\n",
    "\n",
    "    #Generate AnnData for each sample\n",
    "    for sample_name in smp_list:\n",
    "        path = sample_name\n",
    "        for i in os.listdir(path):\n",
    "            if type in i and 'h5' in i:\n",
    "                file = f\"{sample_name}_knee_FPR_0.01_filtered.h5\"\n",
    "                ad[sample_name] = sc.read_10x_h5(sample_name +'/'+file)\n",
    "                scrub = scr.Scrublet(ad[sample_name].X.copy())\n",
    "                doublet_scores, predicted_doublets = scrub.scrub_doublets()\n",
    "                ad[sample_name].obs['doublet_scores'] = doublet_scores\n",
    "                ad[sample_name].obs['predicted_doublets'] = predicted_doublets\n",
    "                ad[sample_name].var['ENSEMBL'] = ad[sample_name].var.index\n",
    "                ad[sample_name].var.rename(columns = {'gene_ids':'SYMBOL'}, inplace = True)\n",
    "                ad[sample_name].var.index = ad[sample_name].var['SYMBOL']\n",
    "                ad[sample_name].var_names_make_unique() \n",
    "                ad[sample_name].obs['sample_id'] = sample_name\n",
    "                ad[sample_name].obs['barcode'] = ad[sample_name].obs_names\n",
    "                ad[sample_name].obs_names = ad[sample_name].obs['barcode']+\"_\"+ad[sample_name].obs['sample_id']\n",
    "\n",
    "    #Merge AnnData objects from all the samples together    \n",
    "    from scipy.sparse import vstack\n",
    "    stack = vstack([ad[x].X for x in smp_list]) # stack data\n",
    "    adata = sc.AnnData(stack, var = ad[smp_list[0]].var)\n",
    "    adata.obs = pd.concat([ad[x].obs for x in smp_list], axis = 0)\n",
    "    \n",
    "    for c in adata.obs.columns:\n",
    "        adata.obs[c] = adata.obs[c].astype('str')\n",
    "    adata.obs =  adata.obs.copy()\n",
    "\n",
    "    return adata, ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "spread-navigator",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata= pd.read_csv(\"/lustre/scratch126/cellgen/team298/sm54/Data_Integration/Integrated_Object/metadata/metadata_harmonised.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "smart-storm",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/lustre/scratch126/cellgen/team205/sharedData/kt21/all_multiome_cellbender\n"
     ]
    }
   ],
   "source": [
    "cd /lustre/scratch126/cellgen/team205/sharedData/kt21/all_multiome_cellbender"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "equivalent-sunday",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organ</th>\n",
       "      <th>Run_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Anatomical_Site</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>Technology</th>\n",
       "      <th>iRODs path</th>\n",
       "      <th>Farm Path</th>\n",
       "      <th>CellBender_Output_Path</th>\n",
       "      <th>Demultiplexed_Barcodes_Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Skull</td>\n",
       "      <td>HCA_BN_F12627470_and_HCA_BN_F12605339</td>\n",
       "      <td>5.8, 7.6, 8.3</td>\n",
       "      <td>knee</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc201...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sharedData/...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Skull</td>\n",
       "      <td>HCA_BN_F12627471_and_HCA_BN_F12605340</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>hip</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc201...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sharedData/...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Skull</td>\n",
       "      <td>HCA_BN_F12627472_and_HCA_BN_F12605341</td>\n",
       "      <td>5.8, 8.3</td>\n",
       "      <td>hip/shoulder</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc201...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sharedData/...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Skull</td>\n",
       "      <td>HCA_BN_F12627473_and_HCA_BN_F12605342</td>\n",
       "      <td>7.6, 8.3</td>\n",
       "      <td>shoulder</td>\n",
       "      <td>multiplexed</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc201...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sharedData/...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Skull</td>\n",
       "      <td>HCA_BN_F12874040_and_HCA_BN_F12865672</td>\n",
       "      <td>9.4</td>\n",
       "      <td>hip</td>\n",
       "      <td>2384</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc201...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sharedData/...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Spine</td>\n",
       "      <td>41865_SB_200532_11018320</td>\n",
       "      <td>17</td>\n",
       "      <td>cervical</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>/seq/illumina/runs/41/41865/cellranger/cellran...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Spine</td>\n",
       "      <td>41865_SB_200532_11018321</td>\n",
       "      <td>17</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>/seq/illumina/runs/41/41865/cellranger/cellran...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>Spine</td>\n",
       "      <td>41865_SB_200532_11018322</td>\n",
       "      <td>17</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>/seq/illumina/runs/41/41865/cellranger/cellran...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>Spine</td>\n",
       "      <td>41865_SB_200532_11018323</td>\n",
       "      <td>17</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>/seq/illumina/runs/41/41865/cellranger/cellran...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Spine</td>\n",
       "      <td>41865_SB_200532_11018324</td>\n",
       "      <td>17</td>\n",
       "      <td>sacral</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>10X_scRNA-seq</td>\n",
       "      <td>/seq/illumina/runs/41/41865/cellranger/cellran...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>/lustre/scratch126/cellgen/team298/sm54/Data_I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Organ                                 Run_ID            Age  \\\n",
       "0     Skull  HCA_BN_F12627470_and_HCA_BN_F12605339  5.8, 7.6, 8.3   \n",
       "1     Skull  HCA_BN_F12627471_and_HCA_BN_F12605340       7.6, 8.3   \n",
       "2     Skull  HCA_BN_F12627472_and_HCA_BN_F12605341       5.8, 8.3   \n",
       "3     Skull  HCA_BN_F12627473_and_HCA_BN_F12605342       7.6, 8.3   \n",
       "4     Skull  HCA_BN_F12874040_and_HCA_BN_F12865672            9.4   \n",
       "..      ...                                    ...            ...   \n",
       "127  Spine                41865_SB_200532_11018320             17   \n",
       "128  Spine                41865_SB_200532_11018321             17   \n",
       "129  Spine                41865_SB_200532_11018322             17   \n",
       "130  Spine                41865_SB_200532_11018323             17   \n",
       "131  Spine                41865_SB_200532_11018324             17   \n",
       "\n",
       "    Anatomical_Site     Donor_ID    Technology   \\\n",
       "0              knee  multiplexed   10X_multiome   \n",
       "1               hip  multiplexed   10X_multiome   \n",
       "2      hip/shoulder  multiplexed   10X_multiome   \n",
       "3          shoulder  multiplexed   10X_multiome   \n",
       "4               hip         2384   10X_multiome   \n",
       "..              ...          ...            ...   \n",
       "127       cervical     HDBR15948  10X_scRNA-seq   \n",
       "128       thoracic     HDBR15948  10X_scRNA-seq   \n",
       "129       thoracic     HDBR15948  10X_scRNA-seq   \n",
       "130         lumbar     HDBR15948  10X_scRNA-seq   \n",
       "131          sacral    HDBR15948  10X_scRNA-seq   \n",
       "\n",
       "                                            iRODs path  \\\n",
       "0    /seq/illumina/cellranger-arc/cellranger-arc201...   \n",
       "1    /seq/illumina/cellranger-arc/cellranger-arc201...   \n",
       "2    /seq/illumina/cellranger-arc/cellranger-arc201...   \n",
       "3    /seq/illumina/cellranger-arc/cellranger-arc201...   \n",
       "4    /seq/illumina/cellranger-arc/cellranger-arc201...   \n",
       "..                                                 ...   \n",
       "127  /seq/illumina/runs/41/41865/cellranger/cellran...   \n",
       "128  /seq/illumina/runs/41/41865/cellranger/cellran...   \n",
       "129  /seq/illumina/runs/41/41865/cellranger/cellran...   \n",
       "130  /seq/illumina/runs/41/41865/cellranger/cellran...   \n",
       "131  /seq/illumina/runs/41/41865/cellranger/cellran...   \n",
       "\n",
       "                                            Farm Path   \\\n",
       "0    /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "1    /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "2    /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "3    /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "4    /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "..                                                 ...   \n",
       "127  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "128  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "129  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "130  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "131  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "\n",
       "                                CellBender_Output_Path  \\\n",
       "0    /lustre/scratch126/cellgen/team205/sharedData/...   \n",
       "1    /lustre/scratch126/cellgen/team205/sharedData/...   \n",
       "2    /lustre/scratch126/cellgen/team205/sharedData/...   \n",
       "3    /lustre/scratch126/cellgen/team205/sharedData/...   \n",
       "4    /lustre/scratch126/cellgen/team205/sharedData/...   \n",
       "..                                                 ...   \n",
       "127  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "128  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "129  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "130  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "131  /lustre/scratch126/cellgen/team298/sm54/Data_I...   \n",
       "\n",
       "    Demultiplexed_Barcodes_Location  \n",
       "0                               NaN  \n",
       "1                               NaN  \n",
       "2                               NaN  \n",
       "3                               NaN  \n",
       "4                               NaN  \n",
       "..                              ...  \n",
       "127                             NaN  \n",
       "128                             NaN  \n",
       "129                             NaN  \n",
       "130                             NaN  \n",
       "131                             NaN  \n",
       "\n",
       "[132 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "exciting-pencil",
   "metadata": {},
   "outputs": [],
   "source": [
    "arr = os.listdir()\n",
    "\n",
    "sample_list= metadata[\"Run_ID\"].tolist()\n",
    "\n",
    "sample_list = [x for x in sample_list if str(x) != 'nan']\n",
    "\n",
    "\n",
    "filter_data = [x for x in arr if\n",
    "              any(y in x for y in sample_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "impaired-spare",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HCA_BN_F13109713_and_HCA_BN_F13101383',\n",
       " 'HCA_BN_F12863835_and_HCA_BN_F12865000',\n",
       " 'HCA_BN_F12808062_and_HCA_BN_F12803238',\n",
       " 'HCA_BN_F12482359_and_HCA_BN_F12477271',\n",
       " 'HCA_BRA_F13774555_and_HCA_BRA_F13852282',\n",
       " 'HCA_BN_F13177436_and_HCA_BN_F13177209',\n",
       " 'HCA_BN_F12482364_and_HCA_BN_F12477276',\n",
       " 'HCA_BN_F13177435_and_HCA_BN_F13177208',\n",
       " 'HCA_BN_F12874046_and_HCA_BN_F12865678',\n",
       " 'HCA_BN_F12482365_and_HCA_BN_F12477277',\n",
       " 'HCA_BN_F13109711_and_HCA_BN_F13101381',\n",
       " 'HCA_BN_F12874040_and_HCA_BN_F12865672',\n",
       " 'HCA_BN_F13109712_and_HCA_BN_F13101382',\n",
       " 'HCA_BRA_F13168992_and_HCA_BRA_F13168608',\n",
       " 'HCA_BN_F12874042_and_HCA_BN_F12865674',\n",
       " 'HCA_BN_F12874043_and_HCA_BN_F12865675',\n",
       " 'HCA_BRA_F13168988_and_HCA_BRA_F13168604',\n",
       " 'HCA_BN_F12482366_and_HCA_BN_F12477278',\n",
       " 'HCA_BRA_F13774556_and_HCA_BRA_F13852283',\n",
       " 'HCA_BRA_F13168989_and_HCA_BRA_F13168605',\n",
       " 'HCA_BN_F12966456_and_HCA_BN_F12951832',\n",
       " 'HCA_BN_F12808059_and_HCA_BN_F12803235',\n",
       " 'HCA_BN_F12922483_and_HCA_BN_F12918721',\n",
       " 'HCA_BN_F12922481_and_HCA_BN_F12918719',\n",
       " 'HCA_BRA_F13774552_and_HCA_BRA_F13768422',\n",
       " 'HCA_BN_F13177437_and_HCA_BN_F13177210',\n",
       " 'HCA_BN_F12863837_and_HCA_BN_F12865002',\n",
       " 'HCA_BRA_F13168991_and_HCA_BRA_F13168607',\n",
       " 'HCA_BN_F13177434_and_HCA_BN_F13177207',\n",
       " 'HCA_BN_F12808058_and_HCA_BN_F12803234',\n",
       " 'HCA_BN_F13109719_and_HCA_BN_F13128947',\n",
       " 'HCA_BN_F13109714_and_HCA_BN_F13128942',\n",
       " 'HCA_BN_F12966458_and_HCA_BN_F12951834',\n",
       " 'HCA_BN_F12808060_and_HCA_BN_F12803236',\n",
       " 'HCA_BN_F12874047_and_HCA_BN_F12865679',\n",
       " 'HCA_BN_F13109718_and_HCA_BN_F13128946',\n",
       " 'HCA_BN_F12482360_and_HCA_BN_F12477272',\n",
       " 'HCA_BRA_F13774554_and_HCA_BRA_F13852281',\n",
       " 'HCA_BN_F12966460_and_HCA_BN_F12951836',\n",
       " 'HCA_BRA_F13168990_and_HCA_BRA_F13168606',\n",
       " 'HCA_BN_F12922486_and_HCA_BN_F12918724',\n",
       " 'HCA_BN_F12808061_and_HCA_BN_F12803237',\n",
       " 'HCA_BN_F12966461_and_HCA_BN_F12951837',\n",
       " 'HCA_BN_F12966459_and_HCA_BN_F12951835',\n",
       " 'HCA_BN_F12627470_and_HCA_BN_F12605339',\n",
       " 'HCA_BN_F12922484_and_HCA_BN_F12918722',\n",
       " 'HCA_BRA_F13774551_and_HCA_BRA_F13768421',\n",
       " 'HCA_BN_F12922482_and_HCA_BN_F12918720',\n",
       " 'HCA_BN_F12874045_and_HCA_BN_F12865677',\n",
       " 'HCA_BN_F12482361_and_HCA_BN_F12477273',\n",
       " 'HCA_BN_F12627471_and_HCA_BN_F12605340',\n",
       " 'HCA_BRA_F13774553_and_HCA_BRA_F13768423',\n",
       " 'HCA_BN_F13109710_and_HCA_BN_F13101380',\n",
       " 'HCA_BN_F12627473_and_HCA_BN_F12605342',\n",
       " 'HCA_BN_F13109715_and_HCA_BN_F13128943',\n",
       " 'HCA_BN_F12874041_and_HCA_BN_F12865673',\n",
       " 'HCA_BN_F12863836_and_HCA_BN_F12865001',\n",
       " 'HCA_BN_F12627472_and_HCA_BN_F12605341',\n",
       " 'HCA_BN_F12922487_and_HCA_BN_F12918725',\n",
       " 'HCA_BRA_F13168993_and_HCA_BRA_F13168609',\n",
       " 'HCA_BN_F12922485_and_HCA_BN_F12918723',\n",
       " 'HCA_BN_F12966455_and_HCA_BN_F12951831',\n",
       " 'HCA_BN_F12966457_and_HCA_BN_F12951833',\n",
       " 'HCA_BN_F12874044_and_HCA_BN_F12865676',\n",
       " 'HCA_BN_F12482362_and_HCA_BN_F12477274',\n",
       " 'HCA_BN_F12482363_and_HCA_BN_F12477275']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "italic-railway",
   "metadata": {},
   "source": [
    "# Skull and meninges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "renewable-hundred",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(filter_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "refined-closure",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_data.remove('HCA_BRA_F13774552_and_HCA_BRA_F13768422')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "personalized-typing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(filter_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pressed-milan",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "received-transcript",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.77\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.8%\n",
      "Elapsed time: 73.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.77\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.8%\n",
      "Elapsed time: 76.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.77\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.8%\n",
      "Elapsed time: 75.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.77\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.8%\n",
      "Elapsed time: 74.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 79.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 80.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 85.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 84.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 82.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 79.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 80.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 34.2%\n",
      "Estimated detectable doublet fraction = 79.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 43.2%\n",
      "Elapsed time: 39.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 34.2%\n",
      "Estimated detectable doublet fraction = 79.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 43.2%\n",
      "Elapsed time: 37.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 34.2%\n",
      "Estimated detectable doublet fraction = 79.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 43.2%\n",
      "Elapsed time: 37.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 34.2%\n",
      "Estimated detectable doublet fraction = 79.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 43.2%\n",
      "Elapsed time: 38.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 11.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 11.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 12.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 12.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 12.5%\n",
      "Estimated detectable doublet fraction = 54.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 23.0%\n",
      "Elapsed time: 32.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 12.5%\n",
      "Estimated detectable doublet fraction = 54.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 23.0%\n",
      "Elapsed time: 33.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 12.5%\n",
      "Estimated detectable doublet fraction = 54.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 23.0%\n",
      "Elapsed time: 32.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 12.5%\n",
      "Estimated detectable doublet fraction = 54.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 23.0%\n",
      "Elapsed time: 31.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.4%\n",
      "Elapsed time: 32.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.4%\n",
      "Elapsed time: 30.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.4%\n",
      "Elapsed time: 30.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.4%\n",
      "Elapsed time: 30.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.61\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.6%\n",
      "Elapsed time: 8.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.61\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.6%\n",
      "Elapsed time: 8.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.61\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.6%\n",
      "Elapsed time: 8.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.61\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.6%\n",
      "Elapsed time: 7.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 6.8%\n",
      "Estimated detectable doublet fraction = 30.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 32.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 6.8%\n",
      "Estimated detectable doublet fraction = 30.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 31.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 6.8%\n",
      "Estimated detectable doublet fraction = 30.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 30.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 6.8%\n",
      "Estimated detectable doublet fraction = 30.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 31.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.48\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 52.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.48\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 51.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.48\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 49.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.48\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 51.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.5%\n",
      "Elapsed time: 14.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.5%\n",
      "Elapsed time: 15.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.5%\n",
      "Elapsed time: 14.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.5%\n",
      "Elapsed time: 17.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.06\n",
      "Detected doublet rate = 87.4%\n",
      "Estimated detectable doublet fraction = 97.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.4%\n",
      "Elapsed time: 53.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.06\n",
      "Detected doublet rate = 87.4%\n",
      "Estimated detectable doublet fraction = 97.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.4%\n",
      "Elapsed time: 55.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.06\n",
      "Detected doublet rate = 87.4%\n",
      "Estimated detectable doublet fraction = 97.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.4%\n",
      "Elapsed time: 56.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.06\n",
      "Detected doublet rate = 87.4%\n",
      "Estimated detectable doublet fraction = 97.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.4%\n",
      "Elapsed time: 54.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 29.4%\n",
      "Estimated detectable doublet fraction = 66.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 44.0%\n",
      "Elapsed time: 45.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 29.4%\n",
      "Estimated detectable doublet fraction = 66.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 44.0%\n",
      "Elapsed time: 42.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 29.4%\n",
      "Estimated detectable doublet fraction = 66.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 44.0%\n",
      "Elapsed time: 44.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 29.4%\n",
      "Estimated detectable doublet fraction = 66.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 44.0%\n",
      "Elapsed time: 45.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.04\n",
      "Detected doublet rate = 88.2%\n",
      "Estimated detectable doublet fraction = 98.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.3%\n",
      "Elapsed time: 54.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.04\n",
      "Detected doublet rate = 88.2%\n",
      "Estimated detectable doublet fraction = 98.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.3%\n",
      "Elapsed time: 55.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.04\n",
      "Detected doublet rate = 88.2%\n",
      "Estimated detectable doublet fraction = 98.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.3%\n",
      "Elapsed time: 54.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.04\n",
      "Detected doublet rate = 88.2%\n",
      "Estimated detectable doublet fraction = 98.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 89.3%\n",
      "Elapsed time: 55.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 2.0%\n",
      "Estimated detectable doublet fraction = 18.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.1%\n",
      "Elapsed time: 8.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 2.0%\n",
      "Estimated detectable doublet fraction = 18.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.1%\n",
      "Elapsed time: 9.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 2.0%\n",
      "Estimated detectable doublet fraction = 18.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.1%\n",
      "Elapsed time: 9.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 2.0%\n",
      "Estimated detectable doublet fraction = 18.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.1%\n",
      "Elapsed time: 9.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 45.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 44.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 45.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 44.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 16.2%\n",
      "Estimated detectable doublet fraction = 54.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.9%\n",
      "Elapsed time: 30.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 16.2%\n",
      "Estimated detectable doublet fraction = 54.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.9%\n",
      "Elapsed time: 30.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 16.2%\n",
      "Estimated detectable doublet fraction = 54.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.9%\n",
      "Elapsed time: 30.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 16.2%\n",
      "Estimated detectable doublet fraction = 54.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.9%\n",
      "Elapsed time: 31.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.4%\n",
      "Estimated detectable doublet fraction = 73.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.7%\n",
      "Elapsed time: 13.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.4%\n",
      "Estimated detectable doublet fraction = 73.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.7%\n",
      "Elapsed time: 13.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.4%\n",
      "Estimated detectable doublet fraction = 73.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.7%\n",
      "Elapsed time: 13.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.4%\n",
      "Estimated detectable doublet fraction = 73.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.7%\n",
      "Elapsed time: 13.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 31.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 31.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 31.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 30.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 31.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 31.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 32.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Warning: failed to automatically identify doublet score threshold. Run `call_doublets` with user-specified threshold.\n",
      "Elapsed time: 33.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Warning: failed to automatically identify doublet score threshold. Run `call_doublets` with user-specified threshold.\n",
      "Elapsed time: 35.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Warning: failed to automatically identify doublet score threshold. Run `call_doublets` with user-specified threshold.\n",
      "Elapsed time: 34.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Warning: failed to automatically identify doublet score threshold. Run `call_doublets` with user-specified threshold.\n",
      "Elapsed time: 33.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 11.8%\n",
      "Estimated detectable doublet fraction = 53.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 36.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 11.8%\n",
      "Estimated detectable doublet fraction = 53.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 38.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 11.8%\n",
      "Estimated detectable doublet fraction = 53.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 37.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 11.8%\n",
      "Estimated detectable doublet fraction = 53.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.3%\n",
      "Elapsed time: 39.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.12\n",
      "Detected doublet rate = 34.4%\n",
      "Estimated detectable doublet fraction = 80.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.9%\n",
      "Elapsed time: 37.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.12\n",
      "Detected doublet rate = 34.4%\n",
      "Estimated detectable doublet fraction = 80.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.9%\n",
      "Elapsed time: 38.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.12\n",
      "Detected doublet rate = 34.4%\n",
      "Estimated detectable doublet fraction = 80.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.9%\n",
      "Elapsed time: 38.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.12\n",
      "Detected doublet rate = 34.4%\n",
      "Estimated detectable doublet fraction = 80.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.9%\n",
      "Elapsed time: 38.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 100.0%\n",
      "Elapsed time: 11.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 100.0%\n",
      "Elapsed time: 11.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 100.0%\n",
      "Elapsed time: 11.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 100.0%\n",
      "Elapsed time: 13.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 35.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 34.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 35.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 35.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 35.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 35.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.54\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 35.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 30.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 30.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 29.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 31.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.0%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.1%\n",
      "Elapsed time: 48.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.0%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.1%\n",
      "Elapsed time: 46.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.0%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.1%\n",
      "Elapsed time: 48.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.0%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.1%\n",
      "Elapsed time: 47.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.0%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.1%\n",
      "Elapsed time: 49.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.0%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.1%\n",
      "Elapsed time: 48.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.0%\n",
      "Estimated detectable doublet fraction = 33.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.1%\n",
      "Elapsed time: 46.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.5%\n",
      "Elapsed time: 27.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.5%\n",
      "Elapsed time: 27.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.5%\n",
      "Elapsed time: 26.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.5%\n",
      "Elapsed time: 26.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 21.6%\n",
      "Estimated detectable doublet fraction = 69.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 12.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 21.6%\n",
      "Estimated detectable doublet fraction = 69.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 11.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 21.6%\n",
      "Estimated detectable doublet fraction = 69.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 12.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 21.6%\n",
      "Estimated detectable doublet fraction = 69.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 11.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.5%\n",
      "Elapsed time: 24.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.5%\n",
      "Elapsed time: 25.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.5%\n",
      "Elapsed time: 26.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.5%\n",
      "Elapsed time: 25.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 35.3%\n",
      "Elapsed time: 24.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 35.3%\n",
      "Elapsed time: 24.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 35.3%\n",
      "Elapsed time: 24.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 35.3%\n",
      "Elapsed time: 23.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 55.6%\n",
      "Elapsed time: 13.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 55.6%\n",
      "Elapsed time: 10.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 55.6%\n",
      "Elapsed time: 10.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 55.6%\n",
      "Elapsed time: 10.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 34.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 35.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 35.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 35.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 34.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 34.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 36.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 14.3%\n",
      "Elapsed time: 24.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 14.3%\n",
      "Elapsed time: 24.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 14.3%\n",
      "Elapsed time: 24.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 14.3%\n",
      "Elapsed time: 24.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 31.3%\n",
      "Estimated detectable doublet fraction = 78.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 16.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 31.3%\n",
      "Estimated detectable doublet fraction = 78.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 16.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 31.3%\n",
      "Estimated detectable doublet fraction = 78.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 16.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.11\n",
      "Detected doublet rate = 31.3%\n",
      "Estimated detectable doublet fraction = 78.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 16.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 41.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 41.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 42.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 42.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 12.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 13.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 13.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 13.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 15.4%\n",
      "Elapsed time: 2.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 15.4%\n",
      "Elapsed time: 2.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 15.4%\n",
      "Elapsed time: 2.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 15.4%\n",
      "Elapsed time: 2.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 21.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 22.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 22.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 22.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.8%\n",
      "Estimated detectable doublet fraction = 36.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.3%\n",
      "Elapsed time: 57.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.8%\n",
      "Estimated detectable doublet fraction = 36.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.3%\n",
      "Elapsed time: 60.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.8%\n",
      "Estimated detectable doublet fraction = 36.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.3%\n",
      "Elapsed time: 57.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.8%\n",
      "Estimated detectable doublet fraction = 36.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.3%\n",
      "Elapsed time: 60.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.8%\n",
      "Estimated detectable doublet fraction = 36.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.3%\n",
      "Elapsed time: 59.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.8%\n",
      "Estimated detectable doublet fraction = 36.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.3%\n",
      "Elapsed time: 58.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 10.8%\n",
      "Estimated detectable doublet fraction = 36.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 29.3%\n",
      "Elapsed time: 59.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 19.0%\n",
      "Estimated detectable doublet fraction = 70.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 26.8%\n",
      "Elapsed time: 28.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 19.0%\n",
      "Estimated detectable doublet fraction = 70.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 26.8%\n",
      "Elapsed time: 29.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 19.0%\n",
      "Estimated detectable doublet fraction = 70.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 26.8%\n",
      "Elapsed time: 28.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 19.0%\n",
      "Estimated detectable doublet fraction = 70.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 26.8%\n",
      "Elapsed time: 28.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 43.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 45.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 44.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 43.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 43.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 44.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 33.3%\n",
      "Elapsed time: 44.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 21.4%\n",
      "Elapsed time: 42.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 21.4%\n",
      "Elapsed time: 42.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 21.4%\n",
      "Elapsed time: 42.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 21.4%\n",
      "Elapsed time: 41.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 26.0%\n",
      "Estimated detectable doublet fraction = 64.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 34.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 26.0%\n",
      "Estimated detectable doublet fraction = 64.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 32.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 26.0%\n",
      "Estimated detectable doublet fraction = 64.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 33.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 26.0%\n",
      "Estimated detectable doublet fraction = 64.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 40.2%\n",
      "Elapsed time: 33.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 56.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 54.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 54.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 55.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 55.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 54.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 55.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 18.9%\n",
      "Estimated detectable doublet fraction = 48.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 38.7%\n",
      "Elapsed time: 37.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 18.9%\n",
      "Estimated detectable doublet fraction = 48.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 38.7%\n",
      "Elapsed time: 36.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 18.9%\n",
      "Estimated detectable doublet fraction = 48.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 38.7%\n",
      "Elapsed time: 36.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 18.9%\n",
      "Estimated detectable doublet fraction = 48.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 38.7%\n",
      "Elapsed time: 37.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 27.3%\n",
      "Estimated detectable doublet fraction = 73.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.1%\n",
      "Elapsed time: 40.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 27.3%\n",
      "Estimated detectable doublet fraction = 73.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.1%\n",
      "Elapsed time: 39.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 27.3%\n",
      "Estimated detectable doublet fraction = 73.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.1%\n",
      "Elapsed time: 40.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 27.3%\n",
      "Estimated detectable doublet fraction = 73.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.1%\n",
      "Elapsed time: 40.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 27.3%\n",
      "Estimated detectable doublet fraction = 73.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.1%\n",
      "Elapsed time: 41.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 27.3%\n",
      "Estimated detectable doublet fraction = 73.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.1%\n",
      "Elapsed time: 41.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.14\n",
      "Detected doublet rate = 27.3%\n",
      "Estimated detectable doublet fraction = 73.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.1%\n",
      "Elapsed time: 41.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 30.0%\n",
      "Estimated detectable doublet fraction = 70.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.6%\n",
      "Elapsed time: 32.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 30.0%\n",
      "Estimated detectable doublet fraction = 70.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.6%\n",
      "Elapsed time: 33.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 30.0%\n",
      "Estimated detectable doublet fraction = 70.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.6%\n",
      "Elapsed time: 34.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 30.0%\n",
      "Estimated detectable doublet fraction = 70.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.6%\n",
      "Elapsed time: 32.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 16.6%\n",
      "Estimated detectable doublet fraction = 59.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.0%\n",
      "Elapsed time: 36.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 16.6%\n",
      "Estimated detectable doublet fraction = 59.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.0%\n",
      "Elapsed time: 34.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 16.6%\n",
      "Estimated detectable doublet fraction = 59.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.0%\n",
      "Elapsed time: 34.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 16.6%\n",
      "Estimated detectable doublet fraction = 59.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.0%\n",
      "Elapsed time: 34.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 16.6%\n",
      "Estimated detectable doublet fraction = 59.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.0%\n",
      "Elapsed time: 34.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 16.6%\n",
      "Estimated detectable doublet fraction = 59.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.0%\n",
      "Elapsed time: 34.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 16.6%\n",
      "Estimated detectable doublet fraction = 59.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 28.0%\n",
      "Elapsed time: 34.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.1%\n",
      "Estimated detectable doublet fraction = 68.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.7%\n",
      "Elapsed time: 34.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.1%\n",
      "Estimated detectable doublet fraction = 68.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.7%\n",
      "Elapsed time: 34.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.1%\n",
      "Estimated detectable doublet fraction = 68.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.7%\n",
      "Elapsed time: 34.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.13\n",
      "Detected doublet rate = 25.1%\n",
      "Estimated detectable doublet fraction = 68.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.7%\n",
      "Elapsed time: 34.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 20.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 21.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 21.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 21.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.1%\n",
      "Elapsed time: 12.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.1%\n",
      "Elapsed time: 12.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.1%\n",
      "Elapsed time: 11.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 42.1%\n",
      "Elapsed time: 12.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.0%\n",
      "Elapsed time: 7.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.0%\n",
      "Elapsed time: 11.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.0%\n",
      "Elapsed time: 7.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.0%\n",
      "Elapsed time: 7.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 10.1%\n",
      "Estimated detectable doublet fraction = 27.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.3%\n",
      "Elapsed time: 46.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 10.1%\n",
      "Estimated detectable doublet fraction = 27.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.3%\n",
      "Elapsed time: 46.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 10.1%\n",
      "Estimated detectable doublet fraction = 27.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.3%\n",
      "Elapsed time: 46.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.25\n",
      "Detected doublet rate = 10.1%\n",
      "Estimated detectable doublet fraction = 27.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.3%\n",
      "Elapsed time: 48.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 40.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 40.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 41.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 40.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 53.0%\n",
      "Estimated detectable doublet fraction = 90.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 58.7%\n",
      "Elapsed time: 36.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 53.0%\n",
      "Estimated detectable doublet fraction = 90.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 58.7%\n",
      "Elapsed time: 37.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 53.0%\n",
      "Estimated detectable doublet fraction = 90.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 58.7%\n",
      "Elapsed time: 37.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 53.0%\n",
      "Estimated detectable doublet fraction = 90.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 58.7%\n",
      "Elapsed time: 36.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 53.0%\n",
      "Estimated detectable doublet fraction = 90.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 58.7%\n",
      "Elapsed time: 36.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 53.0%\n",
      "Estimated detectable doublet fraction = 90.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 58.7%\n",
      "Elapsed time: 37.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.07\n",
      "Detected doublet rate = 53.0%\n",
      "Estimated detectable doublet fraction = 90.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 58.7%\n",
      "Elapsed time: 37.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 23.2%\n",
      "Estimated detectable doublet fraction = 50.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 45.5%\n",
      "Elapsed time: 44.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 23.2%\n",
      "Estimated detectable doublet fraction = 50.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 45.5%\n",
      "Elapsed time: 42.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 23.2%\n",
      "Estimated detectable doublet fraction = 50.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 45.5%\n",
      "Elapsed time: 42.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 23.2%\n",
      "Estimated detectable doublet fraction = 50.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 45.5%\n",
      "Elapsed time: 42.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.1%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.0%\n",
      "Elapsed time: 47.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.1%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.0%\n",
      "Elapsed time: 47.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.1%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.0%\n",
      "Elapsed time: 46.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.1%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.0%\n",
      "Elapsed time: 46.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.1%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.0%\n",
      "Elapsed time: 47.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.1%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.0%\n",
      "Elapsed time: 47.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.1%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.0%\n",
      "Elapsed time: 47.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 23.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 23.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 22.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 23.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 18.7%\n",
      "Estimated detectable doublet fraction = 60.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 51.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 18.7%\n",
      "Estimated detectable doublet fraction = 60.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 54.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 18.7%\n",
      "Estimated detectable doublet fraction = 60.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 54.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 18.7%\n",
      "Estimated detectable doublet fraction = 60.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 51.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 18.7%\n",
      "Estimated detectable doublet fraction = 60.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 53.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 18.7%\n",
      "Estimated detectable doublet fraction = 60.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 55.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 18.7%\n",
      "Estimated detectable doublet fraction = 60.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 30.9%\n",
      "Elapsed time: 53.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 22.8%\n",
      "Estimated detectable doublet fraction = 69.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.9%\n",
      "Elapsed time: 31.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 22.8%\n",
      "Estimated detectable doublet fraction = 69.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.9%\n",
      "Elapsed time: 31.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 22.8%\n",
      "Estimated detectable doublet fraction = 69.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.9%\n",
      "Elapsed time: 31.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 22.8%\n",
      "Estimated detectable doublet fraction = 69.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.9%\n",
      "Elapsed time: 31.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 19.9%\n",
      "Estimated detectable doublet fraction = 63.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.3%\n",
      "Elapsed time: 42.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 19.9%\n",
      "Estimated detectable doublet fraction = 63.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.3%\n",
      "Elapsed time: 42.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 19.9%\n",
      "Estimated detectable doublet fraction = 63.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.3%\n",
      "Elapsed time: 42.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 19.9%\n",
      "Estimated detectable doublet fraction = 63.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.3%\n",
      "Elapsed time: 42.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 19.9%\n",
      "Estimated detectable doublet fraction = 63.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.3%\n",
      "Elapsed time: 42.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 19.9%\n",
      "Estimated detectable doublet fraction = 63.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.3%\n",
      "Elapsed time: 41.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 19.9%\n",
      "Estimated detectable doublet fraction = 63.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.3%\n",
      "Elapsed time: 41.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 12.6%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.9%\n",
      "Elapsed time: 51.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 12.6%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.9%\n",
      "Elapsed time: 51.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 12.6%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.9%\n",
      "Elapsed time: 51.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 12.6%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.9%\n",
      "Elapsed time: 51.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 12.6%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.9%\n",
      "Elapsed time: 51.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 12.6%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.9%\n",
      "Elapsed time: 52.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.22\n",
      "Detected doublet rate = 12.6%\n",
      "Estimated detectable doublet fraction = 39.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.9%\n",
      "Elapsed time: 52.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 55.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 55.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 56.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 55.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 57.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 56.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 57.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 21.2%\n",
      "Estimated detectable doublet fraction = 61.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 37.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 21.2%\n",
      "Estimated detectable doublet fraction = 61.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 36.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 21.2%\n",
      "Estimated detectable doublet fraction = 61.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 36.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.15\n",
      "Detected doublet rate = 21.2%\n",
      "Estimated detectable doublet fraction = 61.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 34.4%\n",
      "Elapsed time: 36.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.5%\n",
      "Elapsed time: 2.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.5%\n",
      "Elapsed time: 2.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.5%\n",
      "Elapsed time: 2.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.5%\n",
      "Elapsed time: 2.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.3%\n",
      "Elapsed time: 7.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.3%\n",
      "Elapsed time: 7.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.3%\n",
      "Elapsed time: 7.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.3%\n",
      "Elapsed time: 6.8 seconds\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "adata, ad_list = read_10x_output(\n",
    "    smp_list=filter_data,\n",
    "    \n",
    "    type='filtered'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "complimentary-request",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 924995 × 36601\n",
       "    obs: 'doublet_scores', 'predicted_doublets', 'sample_id', 'barcode'\n",
       "    var: 'SYMBOL', 'ENSEMBL'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "alternative-reader",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.09950522264980756</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>TAAGCCTAGGGACCTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.20180995475113112</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CGCACAATCCAGGTTG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.18670076726342708</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CGCTTCTAGGATTGCT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.30946291560102296</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CTTTAGTTCCCAGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.26082130965593786</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CTGACCAAGGCTGTGC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTTATGTCCTTGCGT-1_HCA_BN_F12482363_and_HCA_BN_F12477275</th>\n",
       "      <td>0.05263157894736842</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F12482363_and_HCA_BN_F12477275</td>\n",
       "      <td>AGTTATGTCCTTGCGT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGCCAATCGATAACC-1_HCA_BN_F12482363_and_HCA_BN_F12477275</th>\n",
       "      <td>0.06005221932114881</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F12482363_and_HCA_BN_F12477275</td>\n",
       "      <td>CAGCCAATCGATAACC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTAAGGTCCACAATA-1_HCA_BN_F12482363_and_HCA_BN_F12477275</th>\n",
       "      <td>0.04608294930875575</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F12482363_and_HCA_BN_F12477275</td>\n",
       "      <td>CCTAAGGTCCACAATA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGTGGTTCAAGTGTTT-1_HCA_BN_F12482363_and_HCA_BN_F12477275</th>\n",
       "      <td>0.03187250996015936</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F12482363_and_HCA_BN_F12477275</td>\n",
       "      <td>CGTGGTTCAAGTGTTT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATCCTCAGGCTTAAC-1_HCA_BN_F12482363_and_HCA_BN_F12477275</th>\n",
       "      <td>0.06557377049180327</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F12482363_and_HCA_BN_F12477275</td>\n",
       "      <td>CATCCTCAGGCTTAAC-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>924995 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         doublet_scores  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...  0.09950522264980756   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...  0.20180995475113112   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...  0.18670076726342708   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...  0.30946291560102296   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...  0.26082130965593786   \n",
       "...                                                                 ...   \n",
       "AGTTATGTCCTTGCGT-1_HCA_BN_F12482363_and_HCA_BN_...  0.05263157894736842   \n",
       "CAGCCAATCGATAACC-1_HCA_BN_F12482363_and_HCA_BN_...  0.06005221932114881   \n",
       "CCTAAGGTCCACAATA-1_HCA_BN_F12482363_and_HCA_BN_...  0.04608294930875575   \n",
       "CGTGGTTCAAGTGTTT-1_HCA_BN_F12482363_and_HCA_BN_...  0.03187250996015936   \n",
       "CATCCTCAGGCTTAAC-1_HCA_BN_F12482363_and_HCA_BN_...  0.06557377049180327   \n",
       "\n",
       "                                                   predicted_doublets  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "...                                                               ...   \n",
       "AGTTATGTCCTTGCGT-1_HCA_BN_F12482363_and_HCA_BN_...              False   \n",
       "CAGCCAATCGATAACC-1_HCA_BN_F12482363_and_HCA_BN_...              False   \n",
       "CCTAAGGTCCACAATA-1_HCA_BN_F12482363_and_HCA_BN_...              False   \n",
       "CGTGGTTCAAGTGTTT-1_HCA_BN_F12482363_and_HCA_BN_...              False   \n",
       "CATCCTCAGGCTTAAC-1_HCA_BN_F12482363_and_HCA_BN_...              False   \n",
       "\n",
       "                                                                                sample_id  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...  HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...  HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...  HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...  HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...  HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "...                                                                                   ...   \n",
       "AGTTATGTCCTTGCGT-1_HCA_BN_F12482363_and_HCA_BN_...  HCA_BN_F12482363_and_HCA_BN_F12477275   \n",
       "CAGCCAATCGATAACC-1_HCA_BN_F12482363_and_HCA_BN_...  HCA_BN_F12482363_and_HCA_BN_F12477275   \n",
       "CCTAAGGTCCACAATA-1_HCA_BN_F12482363_and_HCA_BN_...  HCA_BN_F12482363_and_HCA_BN_F12477275   \n",
       "CGTGGTTCAAGTGTTT-1_HCA_BN_F12482363_and_HCA_BN_...  HCA_BN_F12482363_and_HCA_BN_F12477275   \n",
       "CATCCTCAGGCTTAAC-1_HCA_BN_F12482363_and_HCA_BN_...  HCA_BN_F12482363_and_HCA_BN_F12477275   \n",
       "\n",
       "                                                               barcode  \n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...  TAAGCCTAGGGACCTC-1  \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...  CGCACAATCCAGGTTG-1  \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...  CGCTTCTAGGATTGCT-1  \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...  CTTTAGTTCCCAGTAG-1  \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...  CTGACCAAGGCTGTGC-1  \n",
       "...                                                                ...  \n",
       "AGTTATGTCCTTGCGT-1_HCA_BN_F12482363_and_HCA_BN_...  AGTTATGTCCTTGCGT-1  \n",
       "CAGCCAATCGATAACC-1_HCA_BN_F12482363_and_HCA_BN_...  CAGCCAATCGATAACC-1  \n",
       "CCTAAGGTCCACAATA-1_HCA_BN_F12482363_and_HCA_BN_...  CCTAAGGTCCACAATA-1  \n",
       "CGTGGTTCAAGTGTTT-1_HCA_BN_F12482363_and_HCA_BN_...  CGTGGTTCAAGTGTTT-1  \n",
       "CATCCTCAGGCTTAAC-1_HCA_BN_F12482363_and_HCA_BN_...  CATCCTCAGGCTTAAC-1  \n",
       "\n",
       "[924995 rows x 4 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "angry-restoration",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample_id\n",
       "HCA_BN_F12863835_and_HCA_BN_F12865000    34958\n",
       "HCA_BN_F13109713_and_HCA_BN_F13101383    32474\n",
       "HCA_BN_F13109710_and_HCA_BN_F13101380    22183\n",
       "HCA_BN_F13109711_and_HCA_BN_F13101381    21755\n",
       "HCA_BN_F12966457_and_HCA_BN_F12951833    21730\n",
       "HCA_BN_F13177437_and_HCA_BN_F13177210    21195\n",
       "HCA_BN_F12966460_and_HCA_BN_F12951836    20866\n",
       "HCA_BN_F12966455_and_HCA_BN_F12951831    20620\n",
       "HCA_BN_F12922487_and_HCA_BN_F12918725    20472\n",
       "HCA_BN_F12966459_and_HCA_BN_F12951835    20320\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"sample_id\"].value_counts()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "nearby-niger",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var.drop(columns=['ENSEMBL'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "static-volunteer",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_sample= sc.read_10x_h5('HCA_BRA_F13774552_and_HCA_BRA_F13768422/HCA_BRA_F13774552_and_HCA_BRA_F13768422_filtered.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "painful-awareness",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.17\n",
      "Detected doublet rate = 24.4%\n",
      "Estimated detectable doublet fraction = 65.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 37.2%\n",
      "Elapsed time: 26.6 seconds\n"
     ]
    }
   ],
   "source": [
    "scrub = scr.Scrublet(meninges_sample.X.copy())\n",
    "doublet_scores, predicted_doublets = scrub.scrub_doublets()\n",
    "meninges_sample.obs['doublet_scores'] = doublet_scores\n",
    "meninges_sample.obs['predicted_doublets'] = predicted_doublets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "legal-cheat",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CTGATCACAATATGGA-1</th>\n",
       "      <td>0.226994</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGCCATCATGCTCCC-1</th>\n",
       "      <td>0.242718</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAATATGTCGGTCATG-1</th>\n",
       "      <td>0.234646</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTTAATCGGTCAAT-1</th>\n",
       "      <td>0.212828</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGGTTTGTGACCTGG-1</th>\n",
       "      <td>0.200000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTCTCAAGCTATTAG-1</th>\n",
       "      <td>0.012195</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGCGAGGTGCAACTA-1</th>\n",
       "      <td>0.081081</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGTTCTTGTCGCGCAA-1</th>\n",
       "      <td>0.014573</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCCCAGCCTGCA-1</th>\n",
       "      <td>0.086651</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAGGCTGTGAGCGAA-1</th>\n",
       "      <td>0.119847</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13587 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    doublet_scores  predicted_doublets\n",
       "CTGATCACAATATGGA-1        0.226994                True\n",
       "TGGCCATCATGCTCCC-1        0.242718                True\n",
       "CAATATGTCGGTCATG-1        0.234646                True\n",
       "TGGTTAATCGGTCAAT-1        0.212828                True\n",
       "TCGGTTTGTGACCTGG-1        0.200000                True\n",
       "...                            ...                 ...\n",
       "CTTCTCAAGCTATTAG-1        0.012195               False\n",
       "TCGCGAGGTGCAACTA-1        0.081081               False\n",
       "GGTTCTTGTCGCGCAA-1        0.014573               False\n",
       "TTTGTCCCAGCCTGCA-1        0.086651               False\n",
       "CAAGGCTGTGAGCGAA-1        0.119847               False\n",
       "\n",
       "[13587 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_sample.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "freelance-amateur",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_sample.var['ENSEMBL'] = meninges_sample.var.index\n",
    "meninges_sample.var.rename(columns = {'gene_ids':'SYMBOL'}, inplace = True)\n",
    "meninges_sample.var.index = meninges_sample.var['SYMBOL']\n",
    "meninges_sample.var_names_make_unique()\n",
    "meninges_sample.obs['sample_id'] = 'HCA_BRA_F13774552_and_HCA_BRA_F13768422'\n",
    "meninges_sample.obs['barcode'] = meninges_sample.obs_names\n",
    "meninges_sample.obs_names = meninges_sample.obs['barcode']+\"_\"+meninges_sample.obs['sample_id']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "stable-laundry",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_sample.var.drop(columns=['ENSEMBL'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "thousand-there",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 13587 × 36601\n",
       "    obs: 'doublet_scores', 'predicted_doublets', 'sample_id', 'barcode'\n",
       "    var: 'SYMBOL'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "acoustic-horizon",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CTGATCACAATATGGA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.226994</td>\n",
       "      <td>True</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>CTGATCACAATATGGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGCCATCATGCTCCC-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.242718</td>\n",
       "      <td>True</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>TGGCCATCATGCTCCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAATATGTCGGTCATG-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.234646</td>\n",
       "      <td>True</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>CAATATGTCGGTCATG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTTAATCGGTCAAT-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.212828</td>\n",
       "      <td>True</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>TGGTTAATCGGTCAAT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGGTTTGTGACCTGG-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.200000</td>\n",
       "      <td>True</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>TCGGTTTGTGACCTGG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.012195</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>CTTCTCAAGCTATTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.081081</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>TCGCGAGGTGCAACTA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.014573</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>GGTTCTTGTCGCGCAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.086651</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>TTTGTCCCAGCCTGCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.119847</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>CAAGGCTGTGAGCGAA-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13587 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    doublet_scores  \\\n",
       "CTGATCACAATATGGA-1_HCA_BRA_F13774552_and_HCA_BR...        0.226994   \n",
       "TGGCCATCATGCTCCC-1_HCA_BRA_F13774552_and_HCA_BR...        0.242718   \n",
       "CAATATGTCGGTCATG-1_HCA_BRA_F13774552_and_HCA_BR...        0.234646   \n",
       "TGGTTAATCGGTCAAT-1_HCA_BRA_F13774552_and_HCA_BR...        0.212828   \n",
       "TCGGTTTGTGACCTGG-1_HCA_BRA_F13774552_and_HCA_BR...        0.200000   \n",
       "...                                                            ...   \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...        0.012195   \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...        0.081081   \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...        0.014573   \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...        0.086651   \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...        0.119847   \n",
       "\n",
       "                                                    predicted_doublets  \\\n",
       "CTGATCACAATATGGA-1_HCA_BRA_F13774552_and_HCA_BR...                True   \n",
       "TGGCCATCATGCTCCC-1_HCA_BRA_F13774552_and_HCA_BR...                True   \n",
       "CAATATGTCGGTCATG-1_HCA_BRA_F13774552_and_HCA_BR...                True   \n",
       "TGGTTAATCGGTCAAT-1_HCA_BRA_F13774552_and_HCA_BR...                True   \n",
       "TCGGTTTGTGACCTGG-1_HCA_BRA_F13774552_and_HCA_BR...                True   \n",
       "...                                                                ...   \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...               False   \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...               False   \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...               False   \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...               False   \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...               False   \n",
       "\n",
       "                                                                                  sample_id  \\\n",
       "CTGATCACAATATGGA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "TGGCCATCATGCTCCC-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "CAATATGTCGGTCATG-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "TGGTTAATCGGTCAAT-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "TCGGTTTGTGACCTGG-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "...                                                                                     ...   \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "\n",
       "                                                               barcode  \n",
       "CTGATCACAATATGGA-1_HCA_BRA_F13774552_and_HCA_BR...  CTGATCACAATATGGA-1  \n",
       "TGGCCATCATGCTCCC-1_HCA_BRA_F13774552_and_HCA_BR...  TGGCCATCATGCTCCC-1  \n",
       "CAATATGTCGGTCATG-1_HCA_BRA_F13774552_and_HCA_BR...  CAATATGTCGGTCATG-1  \n",
       "TGGTTAATCGGTCAAT-1_HCA_BRA_F13774552_and_HCA_BR...  TGGTTAATCGGTCAAT-1  \n",
       "TCGGTTTGTGACCTGG-1_HCA_BRA_F13774552_and_HCA_BR...  TCGGTTTGTGACCTGG-1  \n",
       "...                                                                ...  \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...  CTTCTCAAGCTATTAG-1  \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...  TCGCGAGGTGCAACTA-1  \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...  GGTTCTTGTCGCGCAA-1  \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...  TTTGTCCCAGCCTGCA-1  \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...  CAAGGCTGTGAGCGAA-1  \n",
       "\n",
       "[13587 rows x 4 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_sample.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aerial-timing",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_skull=anndata.concat([adata,meninges_sample])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "progressive-condition",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYMBOL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC141272.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [MIR1302-2HG, FAM138A, OR4F5, AL627309.1, AL627309.3, AL627309.2, AL627309.5, AL627309.4, AP006222.2, AL732372.1, OR4F29, AC114498.1, OR4F16, AL669831.2, LINC01409, FAM87B, LINC01128, LINC00115, FAM41C, AL645608.6, AL645608.2, AL645608.4, LINC02593, SAMD11, NOC2L, KLHL17, PLEKHN1, PERM1, AL645608.7, HES4, ISG15, AL645608.1, AGRN, AL645608.5, AL645608.8, RNF223, C1orf159, AL390719.3, LINC01342, AL390719.2, TTLL10-AS1, TTLL10, TNFRSF18, TNFRSF4, SDF4, B3GALT6, C1QTNF12, AL162741.1, UBE2J2, LINC01786, SCNN1D, ACAP3, PUSL1, INTS11, AL139287.1, CPTP, TAS1R3, DVL1, MXRA8, AURKAIP1, CCNL2, MRPL20-AS1, MRPL20, AL391244.2, ANKRD65, AL391244.1, TMEM88B, LINC01770, VWA1, ATAD3C, ATAD3B, ATAD3A, TMEM240, SSU72, AL645728.1, FNDC10, AL691432.4, AL691432.2, MIB2, MMP23B, CDK11B, FO704657.1, SLC35E2B, CDK11A, SLC35E2A, NADK, GNB1, AL109917.1, CALML6, TMEM52, CFAP74, AL391845.2, GABRD, AL391845.1, PRKCZ, AL590822.2, PRKCZ-AS1, FAAP20, AL590822.1, SKI, ...]\n",
       "\n",
       "[36601 rows x 0 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_skull.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "russian-filing",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_skull.obs[\"doublet_scores\"]=meninges_skull.obs[\"doublet_scores\"].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "alleged-working",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "doublet_scores         float64\n",
       "predicted_doublets      object\n",
       "sample_id             category\n",
       "barcode               category\n",
       "dtype: object"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_skull.obs.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "widespread-cache",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_skull.obs['predicted_doublets'] = meninges_skull.obs['predicted_doublets'].astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "greater-edwards",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "doublet_scores         float64\n",
       "predicted_doublets    category\n",
       "sample_id             category\n",
       "barcode               category\n",
       "dtype: object"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_skull.obs.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "underlying-handle",
   "metadata": {},
   "source": [
    "## save object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "unauthorized-tractor",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_skull.write_h5ad(\"/lustre/scratch126/cellgen/team298/sm54/Data_Integration/Integrated_Object/data/meninges_skull_cellbender_filtered_raw_counts_with_doublet_scores_20231019.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "armed-sight",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "doublet_scores         float64\n",
       "predicted_doublets    category\n",
       "sample_id             category\n",
       "barcode               category\n",
       "dtype: object"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_skull.obs.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "terminal-parent",
   "metadata": {},
   "source": [
    "# Spine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "authorized-monthly",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/lustre/scratch126/cellgen/team298/sm54/Data_Integration/Spine/data/cellranger_outputs_all_samples\n"
     ]
    }
   ],
   "source": [
    "cd /lustre/scratch126/cellgen/team298/sm54/Data_Integration/Spine/data/cellranger_outputs_all_samples/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "modular-spider",
   "metadata": {},
   "source": [
    "# Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "theoretical-reset",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata= pd.read_csv(\"/home/jovyan/mount/gdrive/Spine/metadata_spine_DRteam.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "further-abraham",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCW</th>\n",
       "      <th>region</th>\n",
       "      <th>subregion</th>\n",
       "      <th>dissociation</th>\n",
       "      <th>batch</th>\n",
       "      <th>irods_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7.0</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>cervical</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841321_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7.0</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841322_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841323_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>cervical</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15868</td>\n",
       "      <td>cellranger600_count_41025_SB_200532_10621989_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>cervical</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15868</td>\n",
       "      <td>cellranger600_count_41025_SB_200532_10621992_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    PCW    region subregion dissociation      batch  \\\n",
       "0   7.0    lumbar  cervical   no_trypsin  HDBR15918   \n",
       "1   7.0  thoracic  thoracic   no_trypsin  HDBR15918   \n",
       "2   7.0  cervical    lumbar   no_trypsin  HDBR15918   \n",
       "3   5.0  cervical  cervical   no_trypsin  HDBR15868   \n",
       "4   5.0  cervical  cervical      trypsin  HDBR15868   \n",
       "..  ...       ...       ...          ...        ...   \n",
       "63  NaN       NaN       NaN          NaN        NaN   \n",
       "64  NaN       NaN       NaN          NaN        NaN   \n",
       "65  NaN       NaN       NaN          NaN        NaN   \n",
       "66  NaN       NaN       NaN          NaN        NaN   \n",
       "67  NaN       NaN       NaN          NaN        NaN   \n",
       "\n",
       "                                           irods_path  \n",
       "0   cellranger600_count_40813_SB_200532_10841321_G...  \n",
       "1   cellranger600_count_40813_SB_200532_10841322_G...  \n",
       "2   cellranger600_count_40813_SB_200532_10841323_G...  \n",
       "3   cellranger600_count_41025_SB_200532_10621989_G...  \n",
       "4   cellranger600_count_41025_SB_200532_10621992_G...  \n",
       "..                                                ...  \n",
       "63                                                NaN  \n",
       "64                                                NaN  \n",
       "65                                                NaN  \n",
       "66                                                NaN  \n",
       "67                                                NaN  \n",
       "\n",
       "[68 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "separated-illustration",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata=metadata.dropna(how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "experimental-movement",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCW</th>\n",
       "      <th>region</th>\n",
       "      <th>subregion</th>\n",
       "      <th>dissociation</th>\n",
       "      <th>batch</th>\n",
       "      <th>irods_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7.0</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>cervical</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841321_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7.0</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841322_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841323_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>cervical</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15868</td>\n",
       "      <td>cellranger600_count_41025_SB_200532_10621989_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>cervical</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15868</td>\n",
       "      <td>cellranger600_count_41025_SB_200532_10621992_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>17.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018320_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>17.0</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018321_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>17.0</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018322_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>17.0</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018323_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>17.0</td>\n",
       "      <td>sacral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018324_G...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     PCW     region subregion dissociation      batch  \\\n",
       "0    7.0     lumbar  cervical   no_trypsin  HDBR15918   \n",
       "1    7.0   thoracic  thoracic   no_trypsin  HDBR15918   \n",
       "2    7.0   cervical    lumbar   no_trypsin  HDBR15918   \n",
       "3    5.0   cervical  cervical   no_trypsin  HDBR15868   \n",
       "4    5.0   cervical  cervical      trypsin  HDBR15868   \n",
       "..   ...        ...       ...          ...        ...   \n",
       "57  17.0   cervical       NaN      trypsin  HDBR15948   \n",
       "58  17.0  thoracic        NaN      trypsin  HDBR15948   \n",
       "59  17.0   thoracic       NaN      trypsin  HDBR15948   \n",
       "60  17.0     lumbar       NaN      trypsin  HDBR15948   \n",
       "61  17.0     sacral       NaN      trypsin  HDBR15948   \n",
       "\n",
       "                                           irods_path  \n",
       "0   cellranger600_count_40813_SB_200532_10841321_G...  \n",
       "1   cellranger600_count_40813_SB_200532_10841322_G...  \n",
       "2   cellranger600_count_40813_SB_200532_10841323_G...  \n",
       "3   cellranger600_count_41025_SB_200532_10621989_G...  \n",
       "4   cellranger600_count_41025_SB_200532_10621992_G...  \n",
       "..                                                ...  \n",
       "57  cellranger600_count_41865_SB_200532_11018320_G...  \n",
       "58  cellranger600_count_41865_SB_200532_11018321_G...  \n",
       "59  cellranger600_count_41865_SB_200532_11018322_G...  \n",
       "60  cellranger600_count_41865_SB_200532_11018323_G...  \n",
       "61  cellranger600_count_41865_SB_200532_11018324_G...  \n",
       "\n",
       "[62 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "negative-spelling",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_276/933390425.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  metadata[\"irods_path\"]= metadata[\"irods_path\"].astype(str)\n"
     ]
    }
   ],
   "source": [
    "metadata[\"irods_path\"]= metadata[\"irods_path\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "confirmed-square",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCW</th>\n",
       "      <th>region</th>\n",
       "      <th>subregion</th>\n",
       "      <th>dissociation</th>\n",
       "      <th>batch</th>\n",
       "      <th>irods_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7.0</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>cervical</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841321_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7.0</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841322_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15918</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841323_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>cervical</td>\n",
       "      <td>no_trypsin</td>\n",
       "      <td>HDBR15868</td>\n",
       "      <td>cellranger600_count_41025_SB_200532_10621989_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>cervical</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15868</td>\n",
       "      <td>cellranger600_count_41025_SB_200532_10621992_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>17.0</td>\n",
       "      <td>cervical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018320_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>17.0</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018321_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>17.0</td>\n",
       "      <td>thoracic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018322_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>17.0</td>\n",
       "      <td>lumbar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018323_G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>17.0</td>\n",
       "      <td>sacral</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>HDBR15948</td>\n",
       "      <td>cellranger600_count_41865_SB_200532_11018324_G...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     PCW     region subregion dissociation      batch  \\\n",
       "0    7.0     lumbar  cervical   no_trypsin  HDBR15918   \n",
       "1    7.0   thoracic  thoracic   no_trypsin  HDBR15918   \n",
       "2    7.0   cervical    lumbar   no_trypsin  HDBR15918   \n",
       "3    5.0   cervical  cervical   no_trypsin  HDBR15868   \n",
       "4    5.0   cervical  cervical      trypsin  HDBR15868   \n",
       "..   ...        ...       ...          ...        ...   \n",
       "57  17.0   cervical       NaN      trypsin  HDBR15948   \n",
       "58  17.0  thoracic        NaN      trypsin  HDBR15948   \n",
       "59  17.0   thoracic       NaN      trypsin  HDBR15948   \n",
       "60  17.0     lumbar       NaN      trypsin  HDBR15948   \n",
       "61  17.0     sacral       NaN      trypsin  HDBR15948   \n",
       "\n",
       "                                           irods_path  \n",
       "0   cellranger600_count_40813_SB_200532_10841321_G...  \n",
       "1   cellranger600_count_40813_SB_200532_10841322_G...  \n",
       "2   cellranger600_count_40813_SB_200532_10841323_G...  \n",
       "3   cellranger600_count_41025_SB_200532_10621989_G...  \n",
       "4   cellranger600_count_41025_SB_200532_10621992_G...  \n",
       "..                                                ...  \n",
       "57  cellranger600_count_41865_SB_200532_11018320_G...  \n",
       "58  cellranger600_count_41865_SB_200532_11018321_G...  \n",
       "59  cellranger600_count_41865_SB_200532_11018322_G...  \n",
       "60  cellranger600_count_41865_SB_200532_11018323_G...  \n",
       "61  cellranger600_count_41865_SB_200532_11018324_G...  \n",
       "\n",
       "[62 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "embedded-spring",
   "metadata": {},
   "outputs": [],
   "source": [
    "arr = os.listdir()\n",
    "\n",
    "sample_list= metadata[\"irods_path\"].tolist()\n",
    "\n",
    "sample_list = [x for x in sample_list if str(x) != 'nan']\n",
    "\n",
    "\n",
    "filter_data = [x for x in arr if\n",
    "              any(y in x for y in sample_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "strategic-avenue",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cellranger600_count_39102_SB_200532_10297929_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254458_GRCh38-2020-A',\n",
       " 'cellranger600_count_41025_SB_200532_10621989_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297927_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254464_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863985_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254461_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863987_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018320_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841321_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018319_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297930_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254463_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254465_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863988_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863995_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841317_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254460_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018313_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841322_GRCh38-2020-A',\n",
       " 'cellranger600_count_41025_SB_200532_10621990_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018316_GRCh38-2020-A',\n",
       " 'cellranger600_count_37295_SB_200532_10032967_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863992_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841319_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254462_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297931_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297933_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297926_GRCh38-2020-A',\n",
       " 'cellranger600_count_41025_SB_200532_10621992_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254466_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018322_GRCh38-2020-A',\n",
       " 'cellranger600_count_37295_SB_200532_10032964_GRCh38-2020-A',\n",
       " 'cellranger600_count_37295_SB_200532_10032966_GRCh38-2020-A',\n",
       " 'cellranger600_count_37295_SB_200532_10032969_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018324_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863994_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018314_GRCh38-2020-A',\n",
       " 'cellranger600_count_41025_SB_200532_10621991_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863993_GRCh38-2020-A',\n",
       " 'cellranger600_count_37295_SB_200532_10032965_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297924_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018321_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863986_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297928_GRCh38-2020-A',\n",
       " 'cellranger600_count_42152_SB_200532_11254459_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297923_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841320_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018315_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297932_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841323_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297922_GRCh38-2020-A',\n",
       " 'cellranger600_count_39102_SB_200532_10297925_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863991_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018318_GRCh38-2020-A',\n",
       " 'cellranger600_count_41456_SB_200532_10863989_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018317_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841318_GRCh38-2020-A',\n",
       " 'cellranger600_count_41865_SB_200532_11018323_GRCh38-2020-A',\n",
       " 'cellranger600_count_37295_SB_200532_10032968_GRCh38-2020-A',\n",
       " 'cellranger600_count_40813_SB_200532_10841316_GRCh38-2020-A']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "mechanical-chick",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(filter_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "senior-boulder",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "super-present",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2302"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "viral-macintosh",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scrublet as scr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "massive-market",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def read_10x_output_spine(smp_list, type = 'filtered'):\n",
    "    import os\n",
    "    \n",
    "    #Writing output from separate samples, processed using CellRanger, into a dictionary of Scanpy objects:\n",
    "    ad = {}\n",
    "\n",
    "    #Generate AnnData for each sample\n",
    "    for sample_name in smp_list:\n",
    "        path = sample_name\n",
    "        for i in os.listdir(path):\n",
    "            if type in i and 'h5' in i:\n",
    "                file = i\n",
    "        ad[sample_name] = sc.read_10x_h5(sample_name +'/'+file)\n",
    "        scrub = scr.Scrublet(ad[sample_name].X.copy())\n",
    "        doublet_scores, predicted_doublets = scrub.scrub_doublets()\n",
    "        ad[sample_name].obs['doublet_scores'] = doublet_scores\n",
    "        ad[sample_name].obs['predicted_doublets'] = predicted_doublets\n",
    "        ad[sample_name].var['SYMBOL'] = ad[sample_name].var.index\n",
    "        #ad[sample_name].var.drop(columns=['gene_ids',\"feature_types\",\"genome\"], inplace=True)\n",
    "        ad[sample_name].var_names_make_unique() \n",
    "        ad[sample_name].obs['sample_id'] = sample_name\n",
    "        ad[sample_name].obs['barcode'] = ad[sample_name].obs_names\n",
    "\n",
    "    #Merge AnnData objects from all the samples together    \n",
    "    from scipy.sparse import vstack\n",
    "    stack = vstack([ad[x].X for x in smp_list]) # stack data\n",
    "    adata = sc.AnnData(stack, var = ad[smp_list[0]].var)\n",
    "    adata.obs = pd.concat([ad[x].obs for x in smp_list], axis = 0)\n",
    "\n",
    "    for c in adata.obs.columns:\n",
    "        adata.obs[c] = adata.obs[c].astype('str')\n",
    "    adata.obs =  adata.obs.copy()\n",
    "\n",
    "    return adata, ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "hollow-begin",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.59\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 2.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.0%\n",
      "Elapsed time: 3.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.31\n",
      "Detected doublet rate = 3.1%\n",
      "Estimated detectable doublet fraction = 26.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.7%\n",
      "Elapsed time: 13.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 13.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.65\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 4.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 26.7%\n",
      "Elapsed time: 12.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 2.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 15.4%\n",
      "Elapsed time: 2.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.5%\n",
      "Elapsed time: 5.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 1.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.9%\n",
      "Elapsed time: 2.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.51\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 35.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.70\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 26.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.16\n",
      "Detected doublet rate = 25.3%\n",
      "Estimated detectable doublet fraction = 58.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 43.1%\n",
      "Elapsed time: 34.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.74\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 26.1%\n",
      "Elapsed time: 12.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Warning: failed to automatically identify doublet score threshold. Run `call_doublets` with user-specified threshold.\n",
      "Elapsed time: 55.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 15.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 2.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.6%\n",
      "Elapsed time: 2.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.6%\n",
      "Estimated detectable doublet fraction = 6.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.4%\n",
      "Elapsed time: 1.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 9.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.3%\n",
      "Elapsed time: 8.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.38\n",
      "Detected doublet rate = 1.4%\n",
      "Estimated detectable doublet fraction = 14.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.8%\n",
      "Elapsed time: 6.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 17.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 16.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.66\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 20.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 14.7%\n",
      "Estimated detectable doublet fraction = 45.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.1%\n",
      "Elapsed time: 27.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.34\n",
      "Detected doublet rate = 2.3%\n",
      "Estimated detectable doublet fraction = 17.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.9%\n",
      "Elapsed time: 19.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 2.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.3%\n",
      "Elapsed time: 1.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 2.2%\n",
      "Estimated detectable doublet fraction = 19.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.0%\n",
      "Elapsed time: 1.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 6.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.8%\n",
      "Elapsed time: 2.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 1.5%\n",
      "Estimated detectable doublet fraction = 18.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.2%\n",
      "Elapsed time: 10.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 1.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.8%\n",
      "Elapsed time: 5.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 12.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.4%\n",
      "Elapsed time: 12.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 3.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.5%\n",
      "Elapsed time: 7.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.26\n",
      "Detected doublet rate = 5.6%\n",
      "Estimated detectable doublet fraction = 27.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 20.3%\n",
      "Elapsed time: 31.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 50.0%\n",
      "Elapsed time: 33.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 19.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.62\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 100.0%\n",
      "Elapsed time: 20.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.58\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.8%\n",
      "Elapsed time: 20.0 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.20\n",
      "Detected doublet rate = 14.8%\n",
      "Estimated detectable doublet fraction = 46.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.8%\n",
      "Elapsed time: 54.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 0.6%\n",
      "Estimated detectable doublet fraction = 4.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 13.0%\n",
      "Elapsed time: 1.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 21.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.9%\n",
      "Elapsed time: 19.8 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.6%\n",
      "Estimated detectable doublet fraction = 4.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 13.5%\n",
      "Elapsed time: 2.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.64\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 66.7%\n",
      "Elapsed time: 15.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.77\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 12.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.87\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 53.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.3%\n",
      "Estimated detectable doublet fraction = 2.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.8%\n",
      "Elapsed time: 2.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.59\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 6.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.0%\n",
      "Elapsed time: 6.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.35\n",
      "Detected doublet rate = 2.3%\n",
      "Estimated detectable doublet fraction = 22.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.2%\n",
      "Elapsed time: 8.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.35\n",
      "Detected doublet rate = 2.0%\n",
      "Estimated detectable doublet fraction = 22.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.1%\n",
      "Elapsed time: 9.6 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 8.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.7%\n",
      "Elapsed time: 3.3 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.16\n",
      "Detected doublet rate = 22.5%\n",
      "Estimated detectable doublet fraction = 61.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.5%\n",
      "Elapsed time: 42.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.8%\n",
      "Estimated detectable doublet fraction = 2.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 36.4%\n",
      "Elapsed time: 2.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.73\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 14.3%\n",
      "Elapsed time: 20.5 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.38\n",
      "Detected doublet rate = 1.9%\n",
      "Estimated detectable doublet fraction = 22.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.5%\n",
      "Elapsed time: 10.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 1.3%\n",
      "Estimated detectable doublet fraction = 17.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.5%\n",
      "Elapsed time: 10.1 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 1.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.8%\n",
      "Elapsed time: 2.9 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.23\n",
      "Detected doublet rate = 12.0%\n",
      "Estimated detectable doublet fraction = 38.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 31.0%\n",
      "Elapsed time: 57.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 1.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.7%\n",
      "Elapsed time: 1.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.18\n",
      "Detected doublet rate = 17.3%\n",
      "Estimated detectable doublet fraction = 52.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 32.9%\n",
      "Elapsed time: 32.2 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.68\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 1.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.5%\n",
      "Elapsed time: 9.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 20.7 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.67\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 12.4 seconds\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.38\n",
      "Detected doublet rate = 1.7%\n",
      "Estimated detectable doublet fraction = 34.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.0%\n",
      "Elapsed time: 2.8 seconds\n"
     ]
    }
   ],
   "source": [
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "spine, ad_list = read_10x_output_spine(\n",
    "    smp_list=filter_data,\n",
    "    \n",
    "    type='filtered'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "acquired-premises",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 547321 × 36601\n",
       "    obs: 'doublet_scores', 'predicted_doublets', 'sample_id', 'barcode'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'SYMBOL'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "developmental-genome",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTATTGAGA-1</th>\n",
       "      <td>0.057432432432432436</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACCCAGTATTGAGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGACCATAA-1</th>\n",
       "      <td>0.15625</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAAAGACCATAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAACACGCTGCA-1</th>\n",
       "      <td>0.0979498861047836</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAACACGCTGCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTATACCCA-1</th>\n",
       "      <td>0.0816326530612245</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAAGTATACCCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAATCACGACTA-1</th>\n",
       "      <td>0.14540059347181006</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAATCACGACTA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTCACTACACA-1</th>\n",
       "      <td>0.10358565737051792</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTCACTACACA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCAGGGTAG-1</th>\n",
       "      <td>0.0488771466314399</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTTCAGGGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCGCAACG-1</th>\n",
       "      <td>0.08006814310051105</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTTCCGCAACG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTAACGTTC-1</th>\n",
       "      <td>0.05135135135135135</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGTTGGTAACGTTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTTCGTA-1</th>\n",
       "      <td>0.09826589595375722</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGTTGGTGTTCGTA-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>547321 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          doublet_scores predicted_doublets  \\\n",
       "AAACCCAGTATTGAGA-1  0.057432432432432436              False   \n",
       "AAACGAAAGACCATAA-1               0.15625              False   \n",
       "AAACGAACACGCTGCA-1    0.0979498861047836              False   \n",
       "AAACGAAGTATACCCA-1    0.0816326530612245              False   \n",
       "AAACGAATCACGACTA-1   0.14540059347181006              False   \n",
       "...                                  ...                ...   \n",
       "TTTGGTTCACTACACA-1   0.10358565737051792              False   \n",
       "TTTGGTTTCAGGGTAG-1    0.0488771466314399              False   \n",
       "TTTGGTTTCCGCAACG-1   0.08006814310051105              False   \n",
       "TTTGTTGGTAACGTTC-1   0.05135135135135135              False   \n",
       "TTTGTTGGTGTTCGTA-1   0.09826589595375722              False   \n",
       "\n",
       "                                                            sample_id  \\\n",
       "AAACCCAGTATTGAGA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAAAGACCATAA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAACACGCTGCA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAAGTATACCCA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAATCACGACTA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "...                                                               ...   \n",
       "TTTGGTTCACTACACA-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGGTTTCAGGGTAG-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGGTTTCCGCAACG-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGTTGGTAACGTTC-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGTTGGTGTTCGTA-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "\n",
       "                               barcode  \n",
       "AAACCCAGTATTGAGA-1  AAACCCAGTATTGAGA-1  \n",
       "AAACGAAAGACCATAA-1  AAACGAAAGACCATAA-1  \n",
       "AAACGAACACGCTGCA-1  AAACGAACACGCTGCA-1  \n",
       "AAACGAAGTATACCCA-1  AAACGAAGTATACCCA-1  \n",
       "AAACGAATCACGACTA-1  AAACGAATCACGACTA-1  \n",
       "...                                ...  \n",
       "TTTGGTTCACTACACA-1  TTTGGTTCACTACACA-1  \n",
       "TTTGGTTTCAGGGTAG-1  TTTGGTTTCAGGGTAG-1  \n",
       "TTTGGTTTCCGCAACG-1  TTTGGTTTCCGCAACG-1  \n",
       "TTTGTTGGTAACGTTC-1  TTTGTTGGTAACGTTC-1  \n",
       "TTTGTTGGTGTTCGTA-1  TTTGTTGGTGTTCGTA-1  \n",
       "\n",
       "[547321 rows x 4 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spine.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "dominant-audio",
   "metadata": {},
   "outputs": [],
   "source": [
    "spine.var.drop(columns=['gene_ids',\"feature_types\",\"genome\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "front-cookie",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SYMBOL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>AL627309.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>AL627309.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC141272.1</th>\n",
       "      <td>AC141272.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "      <td>AC023491.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "      <td>AC007325.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>AC007325.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>AC007325.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  SYMBOL\n",
       "MIR1302-2HG  MIR1302-2HG\n",
       "FAM138A          FAM138A\n",
       "OR4F5              OR4F5\n",
       "AL627309.1    AL627309.1\n",
       "AL627309.3    AL627309.3\n",
       "...                  ...\n",
       "AC141272.1    AC141272.1\n",
       "AC023491.2    AC023491.2\n",
       "AC007325.1    AC007325.1\n",
       "AC007325.4    AC007325.4\n",
       "AC007325.2    AC007325.2\n",
       "\n",
       "[36601 rows x 1 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spine.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sufficient-supervisor",
   "metadata": {},
   "source": [
    "## save object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "polish-playlist",
   "metadata": {},
   "outputs": [],
   "source": [
    "spine.write_h5ad(\"/lustre/scratch126/cellgen/team298/sm54/Data_Integration/Integrated_Object/data/spine_no_cellbender_filtered_raw_counts_with_doublet_scores_20231018.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "second-burner",
   "metadata": {},
   "source": [
    "# Combined anndata "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "rotary-shareware",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_skull= sc.read_h5ad(\"/lustre/scratch126/cellgen/team298/sm54/Data_Integration/Integrated_Object/data/meninges_skull_cellbender_filtered_raw_counts_with_doublet_scores_20231019.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "alternative-samoa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.099505</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>TAAGCCTAGGGACCTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.201810</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CGCACAATCCAGGTTG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.186701</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CGCTTCTAGGATTGCT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.309463</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CTTTAGTTCCCAGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.260821</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CTGACCAAGGCTGTGC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.012195</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>CTTCTCAAGCTATTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.081081</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>TCGCGAGGTGCAACTA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.014573</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>GGTTCTTGTCGCGCAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.086651</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>TTTGTCCCAGCCTGCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BRA_F13768422</th>\n",
       "      <td>0.119847</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BRA_F13774552_and_HCA_BRA_F13768422</td>\n",
       "      <td>CAAGGCTGTGAGCGAA-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>938582 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    doublet_scores  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...        0.099505   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...        0.201810   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...        0.186701   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...        0.309463   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...        0.260821   \n",
       "...                                                            ...   \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...        0.012195   \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...        0.081081   \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...        0.014573   \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...        0.086651   \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...        0.119847   \n",
       "\n",
       "                                                   predicted_doublets  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "...                                                               ...   \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...              False   \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...              False   \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...              False   \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...              False   \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...              False   \n",
       "\n",
       "                                                                                  sample_id  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...    HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...    HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...    HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...    HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...    HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "...                                                                                     ...   \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...  HCA_BRA_F13774552_and_HCA_BRA_F13768422   \n",
       "\n",
       "                                                               barcode  \n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...  TAAGCCTAGGGACCTC-1  \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...  CGCACAATCCAGGTTG-1  \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...  CGCTTCTAGGATTGCT-1  \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...  CTTTAGTTCCCAGTAG-1  \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...  CTGACCAAGGCTGTGC-1  \n",
       "...                                                                ...  \n",
       "CTTCTCAAGCTATTAG-1_HCA_BRA_F13774552_and_HCA_BR...  CTTCTCAAGCTATTAG-1  \n",
       "TCGCGAGGTGCAACTA-1_HCA_BRA_F13774552_and_HCA_BR...  TCGCGAGGTGCAACTA-1  \n",
       "GGTTCTTGTCGCGCAA-1_HCA_BRA_F13774552_and_HCA_BR...  GGTTCTTGTCGCGCAA-1  \n",
       "TTTGTCCCAGCCTGCA-1_HCA_BRA_F13774552_and_HCA_BR...  TTTGTCCCAGCCTGCA-1  \n",
       "CAAGGCTGTGAGCGAA-1_HCA_BRA_F13774552_and_HCA_BR...  CAAGGCTGTGAGCGAA-1  \n",
       "\n",
       "[938582 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_skull.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "pursuant-mercury",
   "metadata": {},
   "outputs": [],
   "source": [
    "meninges_skull.var[\"SYMBOL\"]= meninges_skull.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "certified-manual",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 938582 × 36601\n",
       "    obs: 'doublet_scores', 'predicted_doublets', 'sample_id', 'barcode'\n",
       "    var: 'SYMBOL'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meninges_skull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "through-prize",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cns_haem/lib/python3.10/site-packages/anndata/_core/anndata.py:1898: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "spine= sc.read_h5ad(\"/lustre/scratch126/cellgen/team298/sm54/Data_Integration/Integrated_Object/data/spine_no_cellbender_filtered_raw_counts_with_doublet_scores_20231018.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "german-senator",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTATTGAGA-1</th>\n",
       "      <td>0.057432432432432436</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACCCAGTATTGAGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGACCATAA-1</th>\n",
       "      <td>0.15625</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAAAGACCATAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAACACGCTGCA-1</th>\n",
       "      <td>0.0979498861047836</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAACACGCTGCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTATACCCA-1</th>\n",
       "      <td>0.0816326530612245</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAAGTATACCCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAATCACGACTA-1</th>\n",
       "      <td>0.14540059347181006</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_39102_SB_200532_10297929_G...</td>\n",
       "      <td>AAACGAATCACGACTA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTCACTACACA-1</th>\n",
       "      <td>0.10358565737051792</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTCACTACACA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCAGGGTAG-1</th>\n",
       "      <td>0.0488771466314399</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTTCAGGGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCGCAACG-1</th>\n",
       "      <td>0.08006814310051105</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTTCCGCAACG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTAACGTTC-1</th>\n",
       "      <td>0.05135135135135135</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGTTGGTAACGTTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTTCGTA-1</th>\n",
       "      <td>0.09826589595375722</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGTTGGTGTTCGTA-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>547321 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          doublet_scores predicted_doublets  \\\n",
       "AAACCCAGTATTGAGA-1  0.057432432432432436              False   \n",
       "AAACGAAAGACCATAA-1               0.15625              False   \n",
       "AAACGAACACGCTGCA-1    0.0979498861047836              False   \n",
       "AAACGAAGTATACCCA-1    0.0816326530612245              False   \n",
       "AAACGAATCACGACTA-1   0.14540059347181006              False   \n",
       "...                                  ...                ...   \n",
       "TTTGGTTCACTACACA-1   0.10358565737051792              False   \n",
       "TTTGGTTTCAGGGTAG-1    0.0488771466314399              False   \n",
       "TTTGGTTTCCGCAACG-1   0.08006814310051105              False   \n",
       "TTTGTTGGTAACGTTC-1   0.05135135135135135              False   \n",
       "TTTGTTGGTGTTCGTA-1   0.09826589595375722              False   \n",
       "\n",
       "                                                            sample_id  \\\n",
       "AAACCCAGTATTGAGA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAAAGACCATAA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAACACGCTGCA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAAGTATACCCA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "AAACGAATCACGACTA-1  cellranger600_count_39102_SB_200532_10297929_G...   \n",
       "...                                                               ...   \n",
       "TTTGGTTCACTACACA-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGGTTTCAGGGTAG-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGGTTTCCGCAACG-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGTTGGTAACGTTC-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGTTGGTGTTCGTA-1  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "\n",
       "                               barcode  \n",
       "AAACCCAGTATTGAGA-1  AAACCCAGTATTGAGA-1  \n",
       "AAACGAAAGACCATAA-1  AAACGAAAGACCATAA-1  \n",
       "AAACGAACACGCTGCA-1  AAACGAACACGCTGCA-1  \n",
       "AAACGAAGTATACCCA-1  AAACGAAGTATACCCA-1  \n",
       "AAACGAATCACGACTA-1  AAACGAATCACGACTA-1  \n",
       "...                                ...  \n",
       "TTTGGTTCACTACACA-1  TTTGGTTCACTACACA-1  \n",
       "TTTGGTTTCAGGGTAG-1  TTTGGTTTCAGGGTAG-1  \n",
       "TTTGGTTTCCGCAACG-1  TTTGGTTTCCGCAACG-1  \n",
       "TTTGTTGGTAACGTTC-1  TTTGTTGGTAACGTTC-1  \n",
       "TTTGTTGGTGTTCGTA-1  TTTGTTGGTGTTCGTA-1  \n",
       "\n",
       "[547321 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spine.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "associate-brighton",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "doublet_scores        category\n",
       "predicted_doublets    category\n",
       "sample_id             category\n",
       "barcode               category\n",
       "dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spine.obs.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "timely-material",
   "metadata": {},
   "outputs": [],
   "source": [
    "spine.obs[\"doublet_scores\"]=spine.obs[\"doublet_scores\"].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "imposed-discretion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "doublet_scores         float64\n",
       "predicted_doublets    category\n",
       "sample_id             category\n",
       "barcode               category\n",
       "dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spine.obs.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "excessive-picture",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cns_haem/lib/python3.10/site-packages/anndata/_core/anndata.py:1898: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 59791 × 36601\n",
       "    obs: 'doublet_scores', 'predicted_doublets', 'sample_id', 'barcode'\n",
       "    var: 'SYMBOL'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doublet_cutoff = 0.2  # Adjust this value as needed\n",
    "\n",
    "# Filter out cells with a doublet score below the cutoff\n",
    "spine[spine.obs['doublet_scores'] >= doublet_cutoff].copy()\n",
    "\n",
    "# The 'filtered_adata' now contains cells with doublet scores above or equal to the cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "solved-vinyl",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cns_haem/lib/python3.10/site-packages/anndata/_core/anndata.py:1898: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "combined_object= anndata.concat([meninges_skull,spine])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "liked-arena",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doublet_scores</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.099505</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>TAAGCCTAGGGACCTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.201810</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CGCACAATCCAGGTTG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.186701</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CGCTTCTAGGATTGCT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.309463</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CTTTAGTTCCCAGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_F13101383</th>\n",
       "      <td>0.260821</td>\n",
       "      <td>False</td>\n",
       "      <td>HCA_BN_F13109713_and_HCA_BN_F13101383</td>\n",
       "      <td>CTGACCAAGGCTGTGC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTCACTACACA-1</th>\n",
       "      <td>0.103586</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTCACTACACA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCAGGGTAG-1</th>\n",
       "      <td>0.048877</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTTCAGGGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCGCAACG-1</th>\n",
       "      <td>0.080068</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGGTTTCCGCAACG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTAACGTTC-1</th>\n",
       "      <td>0.051351</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGTTGGTAACGTTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTTCGTA-1</th>\n",
       "      <td>0.098266</td>\n",
       "      <td>False</td>\n",
       "      <td>cellranger600_count_40813_SB_200532_10841316_G...</td>\n",
       "      <td>TTTGTTGGTGTTCGTA-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1485903 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    doublet_scores  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...        0.099505   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...        0.201810   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...        0.186701   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...        0.309463   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...        0.260821   \n",
       "...                                                            ...   \n",
       "TTTGGTTCACTACACA-1                                        0.103586   \n",
       "TTTGGTTTCAGGGTAG-1                                        0.048877   \n",
       "TTTGGTTTCCGCAACG-1                                        0.080068   \n",
       "TTTGTTGGTAACGTTC-1                                        0.051351   \n",
       "TTTGTTGGTGTTCGTA-1                                        0.098266   \n",
       "\n",
       "                                                   predicted_doublets  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...              False   \n",
       "...                                                               ...   \n",
       "TTTGGTTCACTACACA-1                                              False   \n",
       "TTTGGTTTCAGGGTAG-1                                              False   \n",
       "TTTGGTTTCCGCAACG-1                                              False   \n",
       "TTTGTTGGTAACGTTC-1                                              False   \n",
       "TTTGTTGGTGTTCGTA-1                                              False   \n",
       "\n",
       "                                                                                            sample_id  \\\n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...              HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...              HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...              HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...              HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...              HCA_BN_F13109713_and_HCA_BN_F13101383   \n",
       "...                                                                                               ...   \n",
       "TTTGGTTCACTACACA-1                                  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGGTTTCAGGGTAG-1                                  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGGTTTCCGCAACG-1                                  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGTTGGTAACGTTC-1                                  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "TTTGTTGGTGTTCGTA-1                                  cellranger600_count_40813_SB_200532_10841316_G...   \n",
       "\n",
       "                                                               barcode  \n",
       "TAAGCCTAGGGACCTC-1_HCA_BN_F13109713_and_HCA_BN_...  TAAGCCTAGGGACCTC-1  \n",
       "CGCACAATCCAGGTTG-1_HCA_BN_F13109713_and_HCA_BN_...  CGCACAATCCAGGTTG-1  \n",
       "CGCTTCTAGGATTGCT-1_HCA_BN_F13109713_and_HCA_BN_...  CGCTTCTAGGATTGCT-1  \n",
       "CTTTAGTTCCCAGTAG-1_HCA_BN_F13109713_and_HCA_BN_...  CTTTAGTTCCCAGTAG-1  \n",
       "CTGACCAAGGCTGTGC-1_HCA_BN_F13109713_and_HCA_BN_...  CTGACCAAGGCTGTGC-1  \n",
       "...                                                                ...  \n",
       "TTTGGTTCACTACACA-1                                  TTTGGTTCACTACACA-1  \n",
       "TTTGGTTTCAGGGTAG-1                                  TTTGGTTTCAGGGTAG-1  \n",
       "TTTGGTTTCCGCAACG-1                                  TTTGGTTTCCGCAACG-1  \n",
       "TTTGTTGGTAACGTTC-1                                  TTTGTTGGTAACGTTC-1  \n",
       "TTTGTTGGTGTTCGTA-1                                  TTTGTTGGTGTTCGTA-1  \n",
       "\n",
       "[1485903 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_object.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "responsible-farmer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1485903 × 36601\n",
       "    obs: 'sample_id', 'barcode'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "authentic-plasma",
   "metadata": {},
   "outputs": [],
   "source": [
    "del spine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "nearby-barbados",
   "metadata": {},
   "outputs": [],
   "source": [
    "del meninges_skull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "therapeutic-entrepreneur",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "517"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "swiss-carolina",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_object.write_h5ad(\"/lustre/scratch126/cellgen/team298/sm54/Data_Integration/Integrated_Object/data/combined_spine_skull_meninges_raw_counts_with_doublet_scores_20231019.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "casual-authority",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1485903 × 36601\n",
       "    obs: 'doublet_scores', 'predicted_doublets', 'sample_id', 'barcode'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adjustable-english",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cns_haem",
   "language": "python",
   "name": "cns_haem"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
